\documentclass[10pt]{llncs}
\usepackage{fullpage}
\usepackage{graphicx}
\usepackage{amsfonts}
\usepackage{float}
\usepackage{amsmath}







\title{Tractability results for the Double-Cut-and-Join circular
  median problem}

\author{
  Ahmad Mahmoody-Ghaidary \inst{1,2}
  \and
  Cedric Chauve \inst{1}
  \and
  Ladislav Stacho \inst{1}
}

\institute{Department of Mathematics, Simon Fraser University, Burnaby
  (BC), Canada
\and Department of Computer Science, Brown University, Providence (RI), USA}

\begin{document}

\maketitle



\begin{abstract}
  The circular median problem in the Dou\-ble-Cut-and-Join (DCJ)
  distance asks to find, for three given genomes, a fourth circular
  genome that minimizes the sum of the mutual distances with the three
  other ones. This problem has been shown to be NP-complete. We show
  here that, if the number of vertices of degree  in the breakpoint
  graph of the three input genomes is fixed, then the problem is
  tractable\footnote{Version of \today. This paper is currently under
    peer-review. The results appeared in {\em Ahmad Mahmoody-Ghaidary,
      Tractability Results for the Double-Cut and Join
      Multichromosomal Problem, MSc thesis, Department of Mathematics,
      Simon Fraser University, 2011}.}.
\end{abstract}



\section{Introduction} \label{sec:intro}

Comparative genomics has been an important source of combinatorial and
algorithmic questions during the last 20 years, especially the
computation of genomic distances and ancestral genomes, as illustrated
by the recent book of Fertin {\em et al.}  \cite{Fertin2009}. Among
these problems, the {\em median} problem is of particular interest:
while the distance problem is tractable in many models, the median
problem is its simplest natural extension (a distance is a function of
two genomes, while the median score is a function of three genomes)
and is computationally intractable in most models. Computing median is
at the heart of inferring gene order phylogenies and ancestral gene
orders~\cite{Murphy2005,Lin2010,Xu2011}. This motivated research on
tractability issues of genomic median problems, well summarized in the
recent paper~\cite{Tannier2009}, as well as on practical algorithms to
address it (see~\cite{Xu2008,Zhang2009,Xu2009a,Xu2009b} and references
there).

Roughly speaking, the median problem is as follows: given three
genomes ,  and  and a genomic distance model , find
a genome  that minimizes the \emph{cost} of  over  defined by . It is in fact an
ancestral genome reconstruction problem, as  can be seen as the
last-common ancestor of  and , with  acting as {\em
  outgroup} (i.e. a genome whose last common ancestor with  and
 is an ancestor of ). In~\cite{Tannier2009}, Tannier {\em et
  al.} explored several variants, based on different models of genomes
(linear, circular or mixed, see Section~\ref{sec:prelim}) and of
genomic distances (Breakpoint, Double-Cut-and-Join, Reversals,
\dots). In particular, they showed that if  is the
Double-Cut-and-Join (DCJ) distance, which is currently the most widely
used genomic distance, then computing a circular or mixed median is
NP-complete. In fact, the only known tractable median problem is the
mixed breakpoint median:  is the breakpoint distance and the median
can contain both linear and circular chromosomes. From a combinatorial
point of view, the central object in the DCJ model is the {\em
  breakpoint graph}: the DCJ distance between two genomes with 
genes is indeed easily obtained from the number of cycles and paths
containing odd number of vertices (odd paths) in this
graph~\cite{Yancopoulos2005,Bergeron2006}. Recent progress in
understanding properties of this graph, and especially of the family
of {\em adequat subgraphs}, lead Xu to introduce algorithms to compute
DCJ median genomes which are efficient on real data, but do not define
well characterized classes of tractable
instances~\cite{Xu2008,Xu2009a,Xu2009b,Xu2011}.

In the present work, we show the following result: if the breakpoint
graph of three genomes contains a constant number of vertices of
degree 3, then computing a DCJ circular median is tractable. To the
best of our knowledge, this is the first result defining an explicit
non-trivial class of tractable instances related to the DCJ median
problem. In Section~\ref{sec:prelim}, we define precisely
combinatorial representations of genomes, the DCJ distance, breakpoint
graphs and the problem we addressed here. In
Section~\ref{sec:results}, we state and prove our main result.





\section{Preliminaries} \label{sec:prelim}

\paragraph{Genes, genomes and breakpoint graph}
Let  represent a set of  {\em
  genes}\footnote{The term {\em gene} is used here in a generic way,
  and might include other genomic markers such as synteny/orthology
  blocks for example.}. Each gene  has a {\em head}  and a
{\em tail} . From now, we assume  always contains  genes.

A {\em genome} , with gene set , is encoded by the order and
orientation of its genes along its chromosomes (i.e. its {\em gene
  order}), or equivalently by the set of the adjacencies between its
gene extremities, that can naturally be represented by a {\em
  matching} on the set of vertices  (Fig.~\ref{genomebreak} (a)). The connected components of the
graph whose vertices are  and edges are the disjoint union of
the edges of  and the edges  (forcing gene
extremities for a given gene to be contiguous) form the {\em
  chromosomes} of  (Fig.~\ref{genomebreak} (b)). A chromosome is
\emph{linear} if it is a path and \emph{circular} if it is a
cycle.  is {\em circular} if it contains only circular chromosomes
(perfect matching), {\em linear} if it contains only linear
chromosomes, and {\em mixed} otherwise.  Fig.~\ref{genomebreak}(a,b)
illustrates this view of genomes as matchings.

The {\em breakpoint graph}  of  genomes  on  is the disjoint union of these genomes, i.e. the
graph with vertex set  and edges given by the matchings defining
these  genomes. Following the usual convention, we consider that
edges in this graph are colored, with color  assigned to genome
 (); it results that  can
have multiple edges of different colors (see Fig.~\ref{genomebreak}
(c)). 

\begin{figure}
\begin{center}
  \includegraphics[scale=0.45]{breakpoint.pdf}
 \vspace*{-3mm}\caption{(a) A genome on 4 genes, with two chromosomes,
   one circular chromosome with gene order  and one linear
   chromosome with gene order , where the sign  indicates
   a reverse orientation. (b) The same genomes with added dashed edges
   connecting gene extremities: every connected component of the
   resulting graph is a chromosome. (c) The breakpoint graph of three
   genomes (whose edges are respectively light gray, thin black and
   thick black) on 4 genes.}
    \label{genomebreak}
\end{center}

\end{figure}

\paragraph{DCJ distance and median}
Given two genomes  and  on , with  being a circular
genome, the {\em DCJ distance}  is given by

where  is the number of cycles in . So larger  implies smaller distance . The general definition
of the DCJ distance (when both  and  are mixed genomes) also
requires to consider the odd paths\footnote{ An \emph{odd}
  (resp.\emph{even}) subgraph is a subgraph with an odd (resp. even)
  number of vertices.} of the breakpoint graph~\cite{Bergeron2006},
but it is easy to see that the breakpoint graph does not contain odd
path if at least one genome is circular. Note that the edges on a
cycle in  are alternatively from  and . An \emph{-alternating cycle} is an even cycle with edges in  and 
alternatively. For simplicity, we may sometimes only call such cycles
\emph{alternating cycles}.

A {\em DCJ circular median} for three genomes , , and ,
or alternatively for their breakpoint graph , is a
circular genome  which minimizes 
. So a circular genome  which maximizes the the
total number of -alternating cycles (for an  ) is a DCJ circular median.

\paragraph{Terminology}
From now, by \emph{median} we always mean \emph{DCJ circular median}.
We denote also by  the sum  for a median .

Let  be a breakpoint graph, and let  be a
median of .  The graph  (also denoted
by  when the context is clear) is called the \emph{median graph}
of  with the DCJ circular median genome  (using disjoint union).
The edges in  are called {\em colored
  edges}\index{colored edge}, and edges in  are called {\em median
  edges}\index{median!edge}. 

A -cycle in  is an -alternating cycle of length ,
for some , in . We denote the total number of alternating
cycles for a median graph  of a breakpoint graph  by
\footnote{Note that  does not depend
  only on the topology of , but also on the colors of its
  edges. Moreover, for different medians  of ,  has the
  same number of alternating cycles, so  does not
  depend of a particular median.}.  If  is a subgraph of , then
 is the maximum number of alternating cycles composed
of edges in , taken over all matchings in .

A terminal vertex in a graph is a vertex of degree . A subgraph of
 is said to be {\em isomorphic to  (resp. )} if it is a
cycle (resp. path) on  vertices.

\begin{remark}
The problem we consider in the present work is to compute a DCJ
circular median of three given genomes, or equivalently to find a
matching in  that maximizes the number of alternating cycles. From
this point of view this is a purely graph theoretical problem that can
be extended naturally to any edge-colored graph, with the convention
that if the graph has an odd number of vertices, then exactly one
vertex does not belong to the matching.
\end{remark}

\paragraph{Shrinking in a breakpoint graph}
{\em Shrinking} a pair of vertices  or an edge with end
vertices  and  was defined in~\cite{Xu2008}. It consists of
three steps: {\em (1)} removing all edges between  and  (if
there is any), {\em (2)} identifying the remaining edges incident to
both  and  and with same color, {\em (3)} removing  and
. We denote the resulting graph by 
(Fig.~\ref{shrink}).

\begin{figure}{
    \begin{center}
      \includegraphics[scale=0.5]{shrinking.pdf}
      \caption{Illustration of the shrinking of a pair  of
        vertices of a breakpoint graph.}
      \label{shrink}
    \end{center}
}\end{figure}



\begin{proposition} \label{Shrink}
  Let  be the breakpoint graph of genomes , and
  . Suppose that there are  colored edges between
   and . If there exists a median  containing the edge ,
  then .
\end{proposition}

\begin{proof} 
 Consider a median  which contains the edge  (which implies
 that both  and  are in the same alternating cycle in
 ). Let ,  (the graph
 obtained from  by removing ,  and the edge ).

 Let  be an alternating cycle in . If  does not contain 
 and , then, obviously,  does not contain any of the  edges
 between  and . Thus,  remains unchanged in . Assume
 now that  contains . If the length of  is larger than 2,
 shrinking  results in a cycle with smaller length in
  (the length decreases by ). Otherwise, if  has length
 2, it disappears in . Thus the number of alternating cycles
 which disappear in  is , since there are  edges
 between  and . Therefore, .
  
  Now suppose  is a median of .  By a similar argument, if , then  has 
  alternating cycles. So, , and, as , we
  have that  (resp. ) is a median of  (resp. ).  \end{proof}



\section{A class of tractable instances}\label{sec:results}

Our main theoretical result is the definition of a large class of
tractable instances for the median problem, namely the ones whose
breakpoint graph contains few vertices of degree . Obviously, the
median problem for three genomes involves a breakpoint graph with
maximum degree . We show here that the hardness of the problem is due
to these vertices of degree .

\begin{theorem}\label{thm:main}
  Let , and  be three genomes. If there exists a median
  of  with at most  edges whose both
  end-vertices are of degree  in , then computing such a median
  can be done in time , where  is the number of vertices of degree , and
   is the number of genes in .
\end{theorem}

\begin{remark}\label{rem:corollaries}
  Note that, as corollaries of this theorem, we have in particular that, 
  \begin{enumerate}
  \item if  is bounded, then computing a median is tractable,
  \item if  is bounded, then computing a median is Fixed-Parameter
    Tractable (FPT) (see~\cite{niedermeier-2008} for a reference on FPT
    algorithms) with parameter .  
  \end{enumerate}
  Moreover, if  is not bounded, we can remove some edges incident
  to vertices of degree , so that in the new instance the number of
  vertices of degree  is bounded. Now, by point 1 above, there is a
  polynomial time algorithm which computes the median of the new
  instance. \end{remark} 

Informally, to prove Theorem~\ref{thm:main}, we first consider the
case where  is a collection of cycles and paths (i.e. has maximum
degree ) and show that a median can be computed in polynomial
time. Next, we consider all possibilities (configurations) for
matching vertices of degree  as median edges. For each configuration,
we reduce the breakpoint graph by shrinking and removing some edges to
obtain a graph whose connected components are paths or cycles. Having
computed all possible configurations for vertices of degree  and
being able to compute a median for all resulting graphs lead to
Theorem~\ref{thm:main}.



From now, , , and  are mixed genomes on  genes, and
 is a median of these genomes, unless otherwise specified. We
denote their breakpoint graph by , and the median graph by .

\subsection{Preliminary results}

We first introduce two useful lemmas that give lower bounds on the
function  in various cases.

\begin{lemma}\label{subpc}
  If  is isomorphic to  or , for , then for
  every subgraph , .
\end{lemma}

\begin{proof}
  Consider the path . Let  be the matching
  consisting of the edges , and ,
  where . Obviously, the number of
  alternating cycles in  is , so
  .
  Similarly .  See
  Fig.~\ref{pc}.

  Any proper subgraph  or  is a union of
  disjoint paths. If we take the union of matchings described above
  for each of these paths and call it , there are at least
   alternating cycles in . Therefore for any
  subgraph , . \end{proof}

\begin{figure}[h]{
    \begin{center}
      \includegraphics[scale=0.75]{pc.pdf}
      \caption{Median edges (dashed) for cycles and union of disjoint paths.}
      \label{pc}
    \end{center}
}\end{figure}

\begin{definition}
  Let  and  be two subgraphs of .  is an {\em
    alternating-subdivision} of  if we can obtain an isomorphic
  copy of  from  as follows: subdivide each edge 
  by an even (possibly zero) number of vertices resulting in a path
  , then remove every second edge, i.e.,
  . We call the removed edges
  a {\em completing matching} for  respective to .
\end{definition}

In the previous definition, note that there might be more than one way
to obtain an isomorphic copy of  from , and consequently,
completing matching is not necessarily unique.




\begin{lemma}\label{alsub}
  If  is an alternating-subdivision of , then .
\end{lemma}

\begin{proof}
  Let  be a median of ,  an arbitrary completing matching
  for  respective to  and .  is a perfect
  matching of , and each alternating cycle in  defines a
  unique alternating cycle in  which implies that
   (see Fig.~\ref{proofalsub}).
\end{proof} 

\begin{figure}[h]
  \begin{center}(a)
    \includegraphics[scale=0.55]{as.pdf} \\ (b)
    \includegraphics[scale=0.55]{asm.pdf}
    \caption{(a) Obtaining  as an alternating-subdivision of
      . (b) Obtaining a matching of  from a median of  (the
      dashed edges are the median edges and the edges of a completing
      matching for  respective to ).}
     \label{sub}
     \label{proofalsub}
  \end{center}
 \end{figure}








\subsection{Independence of arbitrary paths and even cycles}


In this section we introduce the fundamental notion of {\em
  independence} of connected components of cycles and paths in a
breakpoint graph. 

\begin{definition}\label{def:crossing-indep}
  Let  be a subgraph of . An {\em -crossing
    edge}\index{crossing edge} in a median graph  is a median
  edge which connects a vertex in  to a vertex in . An {\em -crossing cycle}\index{crossing cycle} is an
  alternating cycle which contains at least one -crossing edge.
  The subgraph  is {\em -independent} if there is a median 
  for  such that the number of -crossing edges in  is at
  most .
\end{definition}


\begin{proposition}\label{indep}
  Let  be a connected component of . If  is isomorphic to
   or , for , then  is 0-independent.
\end{proposition}

\begin{proof}
  Let  be a median of . Suppose  has  -crossing
  edges in . If , then we are done, so assume that
  . Since  has an even number of vertices,  is even
  and . Because  is a connected component in , each
  -crossing cycle contains an even number of -crossing edges.
  
  Let  be the set of all -crossing cycles in , and
   be the set of all -crossing edges in .  Let
   be the set of colored edges in all cycles of
  , and  be the set of all -crossing edges in
  all cycles of .

  {\smallskip\noindent\em Case 1.} If there is no -crossing cycle,
  i.e., , we modify  by
  removing all -crossing edges, and re-matching the vertices inside
  of  together and outside of  together. Since  is even,
  this is always possible and we get a median with no -crossing
  edge.  

  {\smallskip\noindent\em Case 2.} From now, we assume that there
  exists at least one -crossing cycle. The remainder of the proof
  relies on a transformation on  that reduces the number of edges
  in -crossing cycles, leading to a median with no -crossing
  edge.

  {\smallskip\noindent\em Step 1.} The first step consists of
  choosing, for each -crossing cycle, an arbitrary colored edge in
   incident to an -crossing edge from this cycle. Let  be the
  subgraph of  induced by these chosen colored edges and
  .



  {\smallskip\noindent\em Claim 1.  is an alternating-subdivision
    of .} For a vertex  let  be the neighbor of
   in . If  and  then, by
  definition,  is a colored edge of an -crossing cycle which is
  incident to an -crossing edge. Therefore, there is an alternating
  path from  to , with alternating colored and median edges
  from that cycle. If this path has  colored edges, we subdivide
  the edge  using  vertices and remove every second
  edge. Proceeding in this way for every edge  we obtain
  an alternating-subdivision  of .

  {\smallskip\noindent\em Claim 2. .} First, as every colored edge is in at most one
  alternating cycle and two edges of the same color are not incident
  to each other, . Also , and
  by Lemma~\ref{subpc}, . Finally, from
  Lemma~\ref{alsub}, .

  {\smallskip\noindent\em Step 2.} Now we remove all the edges in
  . Let  be an arbitrary median of ,  the
  matching for  defined by the union of  and an arbitrary
  completing matching for  respective to , and .

  {\smallskip\noindent\em Claim 3.  is a median of .}  First,
  by removing the edges in , the total number of
  alternating cycles decreases by . Next,  and 
  contain at least  alternating cycles each (Claim 2
  above). Hence, the new matching  contains at least the same
  number of alternating cycles than . By definition of a median,
   can not contain more alternating cycles than , so it
  contains the same number of alternating cycles, and is a median of
  . Note that this also implies that .
  
{\smallskip\noindent\em Claim 4.  and .}  If there exists 
  then there would be at least one -crossing cycle induced by 
  which is not induced by  or , this implies  would contain
  more alternating cycles than , which contradicts the fact
  that  and  have the same number of alternating cycles.
  Next, , as  and  (the vertices
  in  are matched to themselves). Therefore, .

  By iterating the above steps we obtain a median with no crossing
  cycle. Then, by case 1, we can modify this median to a median
  without -crossing edge. \end{proof}





\begin{proposition}\label{1dep}
  Let  be a connected component of . If  is isomorphic to
  , for , then  is 1-independent. 
\end{proposition}

\begin{proof}
  We follow the same proof strategy than for
  Proposition~\ref{indep}. The number of -crossing edges is odd. If
  there is no -crossing cycle, we can remove an even number of them
  as in case 1 of the proof of Proposition~\ref{indep}, leaving only
  one -crossing edge. Otherwise, if we assume that there are
  -crossing cycles, we can apply the transformation defined in case
  2 of the proof of Proposition~\ref{indep}. It has similar
  properties, as, from Lemma~\ref{subpc}, for every subgraph , , which implies
  again that .
\end{proof}



\begin{proposition}\label{themedian}
  If  contains only cycles and paths, there exists a median of 
  in which even components have no crossing edge, and each odd path
  has exactly one crossing edge.
\end{proposition}

\begin{proof}
  This result follows from applying, on an arbitrary median graph, the
  transformation introduced in the proof of in Proposition~\ref{indep}
  to each even/odd path or even cycle of the breakpoint graph,
  reducing then the number of crossing edges for each of them, without
  increasing the number of crossing edges in other components.  \end{proof}


\subsection{Alternating cycles for arbitrary paths and even cycles}

The results of the previous section open the way to computing a median
of a breakpoint graph with maximum degree  by considering each path
or even cycle independently, and matching odd paths into pairs (each
defined by a single crossing edge).  The main point of the current
section is to show that paths and even cycles are easy to consider
when computing a median.

\begin{proposition} \label{cycP}
  If  is isomorphic to , for some , then
  . Moreover, there exists
  a median whose edges in  define 
  alternating 2-cycles, and one crossing edge incident to a terminal
  vertex of  if  is odd.
\end{proposition}

\begin{proof}
  From Lemma~\ref{subpc}, . We use induction on  to show that
  . This obviously
  holds for . So we assume that , and consider a
  median  for . If there is no 2-cycle (an alternating cycle
  consisting of two parallel edges) in , each alternating cycle
  has length at least , and hence at least 2 colored edges. So
  .

  Now assume that the median  contains a 2-cycle, with vertices 
  and . Shrinking  results in  that is either a
  single path with  vertices or two paths with  and 
  vertices such that .  In both cases, using induction and
  the fact that all paths are 0-independent or 1-independent, we can
  conclude that,
  \begin{itemize}
  \item if  contains one path, ,
  \item if  contains two paths, .
  \end{itemize}
  To obtain a median with exactly 
  alternating cycles in , we can simply define median edges by
  linking successive vertices in  (as in the proof of
  Lemma~\ref{subpc}). If  is odd this forces the unique
  -crossing edge (Proposition~\ref{1dep}) to contain the last end
  vertex of  (one of its two end vertices), which has no impact on
  the number of alternating cycles as, by definition, this crossing
  edge will not belong to any alternating cycle.  \end{proof}



\begin{lemma} \label{cycC}
  If  is isomorphic to , for some , then either
   or .
\end{lemma}

\begin{proof}
  Obviously, . So, we assume that
  . Let  be an arbitrary median of
  . Following the proof of Proposition~\ref{cycP}, if all
  alternating cycles in  have length at least 4, then the number
  of alternating cycles is at most , so there must exist at least
  one 2-cycle in .  Let  be a colored edge in a 2-cycle:
  
  (Proposition~\ref{Shrink}).  Moreover  is a path, or
  a cycle, and it is a cycle if and only if the two edges incident to
  the ends of  have the same color. If it is a path,
  Proposition~\ref{cycP} implies that  and , which contradicts the assumption
  that . So  is a cycle and the
  edges incident to  have same color. By induction on  (note
  that  and ) we can find a
  median of  with  or , alternating
  cycles. Hence,  or . \end{proof}



Some definitions below assume that cycles of  are oriented, so we
assume from now that edges of every cycle of  are consistently
oriented, clockwise or counterclockwise. Fig.~\ref{cyckind} provides
an illustration.

\begin{definition}\label{def:kind-cyc}
  A cycle  of  is of the {\it first kind} if
  , and it is of the {\it second kind} if
  .
\end{definition}

\begin{definition}\label{def:signature}
  Let  be a cycle of . The {\it signature} of a vertex of  is
  an ordered pair  such that  and  are the colors of the
  edges incident to that vertex:  is the color of the incoming edge
  and  the color of the outgoing edge.  Two vertices  and 
  are {\it diagonal} if their signatures are of the form  and
  .
\end{definition}

\begin{definition}\label{def:cross} 
  Let  be a median of an even cycle , and  and  be
  edges in :  and  {\it cross} if  appear
  in this order along . A {\it cross-free diagonal} matching for
   is a matching whose edges connect pairs of diagonal vertices and
  no two edges cross.
\end{definition}

\begin{figure}{
  \begin{center}
    \includegraphics[scale=0.85]{evencycles.pdf}
  \caption{Dashed edges are median edges: (Left) a cycle  of the
    first kind --- (Right) a cycle  of the second kind; the
    matched vertices are diagonal}
       \label{cyckind}
  \end{center}
 }
\end{figure}




\begin{lemma} \label{2ndKind}
  Let  be isomorphic to an even cycle of the second kind, and 
  be a median of . (1) Each edge in  joins two diagonal
  vertices, and (2) the edges in  do not cross.
\end{lemma}

\begin{proof}
  Let .  We first prove point (1) by contradiction. Assume
  that  and that  and  are not diagonal. Let 
  and  be the respective signatures of  and . Our
  assumption implies that , and we can distinguish
  two cases:  and .
  \begin{itemize}
  \item If , by shrinking the pair  we
    obtain a smaller cycle , and by Proposition~\ref{Shrink},
     which
    is a contradiction, since  is of the second kind. Note that in
    this case  and  cannot be consecutive vertices on .
  \item If , by shrinking the pair , the
    resulting graph can be either a path with  vertices, or a
    cycle and a path, together with  vertices.  In the first
    case, vertices  and  must be consecutive on . But now
    , which is
    a contradiction, since  is of the second kind. In the second
    case , since paths are
    either 0- or 1-independent and  (note that in the
    latter case  and  cannot be consecutive). This is again a
    contradiction, as  is of the second kind.
  \end{itemize}

  We now prove point (2).  is of the second kind, as shown in the
  proof of Lemma~\ref{cycC}, there is a 2-cycle containing a colored
  edge . Moreover, by point (1), vertices  and  are
  diagonal. So  is isomorphic to  and it
  must be of the second kind, as otherwise . Obviously,  does not cross with any median edge of . By
  shrinking this pair and, by induction on the length of the cycle,
  applied to , the proof is complete. \end{proof}



\begin{lemma} \label{2ndKindcyc}
  Let  be isomorphic to  .  is of the second kind if
  and only if there exists a matching  of  that is cross-free
  diagonal.
\end{lemma}

\begin{proof}
  The necessity follows from Lemma~\ref{2ndKind}.  Now assume that
  there exists a cross-free diagonal matching  on vertices of .
  It is easy to see that  contains at least one edge  where 
  and  are consecutive on  (note that  is a perfect matching,
  since  has even number of vertices). If we shrink the pair , the resulting graph is  and the remaining edges of
   are a cross-free diagonal matching for .  We can
  complete the proof by induction on , since , and the
  statement of the lemma is obviously true for  and .
\end{proof}

\begin{lemma} \label{WhatKind}
  Let  be isomorphic to . Deciding if  admits a
  cross-free diagonal matching can be done in time .
\end{lemma}

\begin{proof}
  Let . We rely on a simple greedy algorithm,
  which is in fact a classical algorithm for deciding if a circular
  parenthesis word is balanced; we present it for the sake of
  completeness.

  The key point was given in the proof of Lemma~\ref{2ndKindcyc}: any
  cross-free diagonal matching contains at least one pair of
  consecutive vertices that are matched. Given the circular nature of
  , we can extend this property as follows: if  and  are
  consecutive diagonal vertices and  admits a diagonal cross-free
  matching, then there exists a matching where  and  are
  matched. This leads immediately to a greedy algorithm that matches
  such vertices as soon as they are visited, using a simple stack data
  structure:
 




  \begin{center}
    \scalebox{1}{ \fbox{\begin{minipage}{0.9\linewidth}
          \begin{enumerate}
          \item Let  be an empty matching.
          \item Let  be an empty stack.
          \item For  to 
            \begin{enumerate}
            \item if the top element  of  is diagonal with
              , pop it from the stack  and add 
              to .
            \item else, push  on .
            \end{enumerate}
          \item If  is empty,  admits a cross-free diagonal
            matching, given by , otherwise it does not admit one.
          \end{enumerate}
        \end{minipage}
    } }
  \end{center}

  The time complexity of this algorithm is obviously linear in .
\end{proof}


\begin{proposition}\label{cycles}
  If  is isomorphic to an even cycle of size  (), then
  computing  can be done in time .
\end{proposition} 

\begin{proof}
  Immediate consequence of Lemma~\ref{cycC}, Lemma~\ref{2ndKindcyc},
  and Lemma~\ref{WhatKind}.\end{proof}


\subsection{Proof of Theorem~\ref{thm:main}}

We now have all the elements to prove our main result,
Theorem~\ref{thm:main}. We first prove that computing a median of a
breakpoint graph of maximum degree two is tractable.

\begin{lemma}\label{linkage}
  If  has maximum degree , then there exists a median of  such
  that every odd connected component of  is connected by median
  edges to exactly one other odd connected component.
\end{lemma}

\begin{proof}
  Let  be a median as described in the proof of
  Proposition~\ref{themedian}: every even connected component has no
  crossing edge and each odd path has exactly one crossing
  edge. Moreover, odd cycles have at least one crossing edge.

  Let  be an odd connected component and  one of its crossing
  edges, connecting  to another odd component . Shrinking 
  results into  which is a set of even components
  and it is then -independent. Moreover, as  and  were
  distinct connected components of , from Proposition~\ref{Shrink}
  (with ), . 

  Repeating this argument for other odd components and the fact that
  the number of odd components is even (because the number of vertices
  in the breakpoint graph is even) completes the proof. \end{proof}


\begin{lemma}\label{pair}
  If  has maximum degree  and consists of two odd connected
  components  and , of respective sizes  and ,
  then computing a median of  can be done in time .
\end{lemma}

\begin{proof}
    For parity reasons, a median  contains at least one edge 
    between  and  ( is a -crossing edge).By shrinking  we obtain either one even connected component or
    two even connected components, and, from
    Proposition~\ref{themedian}, we can compute a median for each
    connected component independently. This computation requires
    linear time (Propositions~\ref{cycP} and~\ref{cycles}). There are
    at most  possible candidates for
    . Hence computing a median of  is tractable in time
    . \end{proof}




\begin{proposition}\label{odds}
  If  is a breakpoint graph with  vertices with maximum degree
  , then computing a median of  can be done in .
\end{proposition}

\begin{proof}
    We first consider the case where  contains only odd connected
    components. We define a complete edge-weighted graph  as
    follows:
   \begin{enumerate}
    \item each connected component  defines a vertex ;
    \item each edge  has weight  
    \end{enumerate}
    By Lemma~\ref{pair},  is computable in polynomial time. We
    claim it is computable in .  Suppose  has 
    components and  are the number of vertices in
    each component. So we have . The time to
    construct  is of order
    
    
    
    Finally, by Lemma~\ref{linkage} we only need to find a maximum
    weight matching for , which can be done in  by
    using Edmonds's algorithm~\cite{Edmonds}.  

    If the breakpoint graph  has maximum degree , its connected
    components are paths or cycles. From Proposition~\ref{themedian}
    and Proposition \ref{cycles} we can find the median edges for even
    components independently. Finally for odd components we find the
    median edges as described in the first part of the proof. \end{proof}
    



\noindent{\em Proof of Theorem~\ref{thm:main}.}  We now assume that
 has maximum degree .  

The main idea is to consider all possibilities for matching the
vertices of degree  of . A vertex  of degree  can be matched
in two ways.
\begin{itemize}
\item If it is matched to another vertex of degree , by shrinking
  these two vertices we obtain a smaller graph with fewer vertices of
  degree , and, from Proposition~\ref{Shrink}, we know precisely
  the number of alternating cycles (here -cycles) lost in the
  shrinking process, given by the number of genome edges between the
  two shrinked vertices.
\item If it is matched to a vertex of degree less than 3, then one of
  the edges incident to  is not in any alternating cycle, and we
  can remove this edge and transform  into a vertex of degree 
  (Fig.~\ref{3edge}).
\end{itemize}

\begin{figure} 
  \begin{center}
    \includegraphics[scale=.35]{unused.pdf}
    \caption{The dashed edge is a median edge. The gray edge cannot be
      in any alternating cycle.}
    \label{3edge}
  \end{center}
\end{figure}

Now for each , , we can select  vertices
among all  vertices of degree  (there are 
possibilities), compute an arbitrary perfect matching on these 
vertices, and, for each each remaining vertex of degree , remove an
edge incident to this vertex (there are 
possibilities). The resulting breakpoint graph  is of maximum
degree  and a median can be computed in time , whose number
of alternating cycles needs only to be augmented by the number of
edges between matched vertices of degree  in . 

The number of all such configurations is in
 (the term  is needed to
account for the case ), which leads to the stated complexity.




























 










\section{Conclusion}

In this work, we characterized a large class of tractable instances
for the DCJ median problem (with circular median and mixed
genomes). In fact, we showed that only the vertices of degree  make
the problem intractable. Also, by removing  edges from the
breakpoint graph and decreasing its maximum degree, cost of its median
is not bigger than  plus the cost of the main median (i.e. the
current cost  is a lower bound for the cost of the main median).
Finally, we showed there is an FPT algorithm for the DCJ median
problem, if there exists a median such that the number of its edges
connecting two vertices of degree  is bounded.

Our work also shows that the multiplicity of solutions (i.e. medians)
is likely to happen when dealing with breakpoint graphs with long
paths or even cycles, as we showed that such components can admit
several optimal medians. Hence, our results, as they stand now, are of
interest more for computing the score of a median than for computing
actual medians that can be seen as realistic ancestral
genomes. However, the problem of uniform sampling of optimal median is
worth being explored, even in the simpler setting of breakpoint graphs
of maximum degree  in a first time.

From a theoretical point of view, our work raises several
questions. First, it leaves open the possibility that the DCJ median
problem is FPT. Using the number of vertices of degree  as a
parameter is a a natural approach, although this seems to be a
difficult question to address. The next obvious problem is to extend
our approach to the case of a mixed or linear median. This would
require to better understand the combinatorics of odd paths in the
breakpoint graphs in relation to medians. The simpler problem to find
an optimal way to remove exactly one edge from each circular
chromosome of a circular median while minimizing the number of
destroyed alternating cycles is also open. Extending our results to
the related \emph{DCJ halving problem}~\cite{Tannier2009} is also a
natural question.

Another interesting question is about expanding the breakpoint
distance toward the DCJ distance: for two genomes  and  on
 genes, their breakpoint distance is equal to
 The parameters  and  are also equal
the number of 2-cycles and 1-paths () in the breakpoint graph
, respectively.  The DCJ distance of these genomes is:
 where  and  are the number of
(even) cycles and odd paths in the , respectively. This
motivates us to define a dissimilarity function as follows:
 where  is the number of (even) cycles with at
most  vertices, and  is the number of odd paths
with at most  vertices.  By considering this dissimilarity
measure, the median problem is tractable when , since
. By taking  we have
, and the median problem would
be intractable. A natural question is then to understand for which
values of  and/or  the median problem is tractable, or FPT.

\bigskip{\noindent\bf Acknowledgments.}
C.C. and L.S. are supported by NSERC Discovery Grants.



\begin{thebibliography}{10} \label{bibliography}

\bibitem{Bergeron2006}
  A.~Bergeron, J.~Mixtacki, J.~Stoye.
  \newblock A Unifying View of Genome Rearrangements.
  \newblock In {\em Algorithms in Bioinformatics (WABI 2006)}, vol. 4175 of {\em Lecture Notes Comput. Sci.}, pp. 163--173.
  \newblock Springer.
  \newblock 2006.

\bibitem{Edmonds}  
  J. Edmonds.
  \newblock Paths, trees, and flowers.
  \newblock {\em Canadian Journal of Mathematics}, 17:449--467.
  \newblock 1965.

\bibitem{Fertin2009}
  G.~Fertin, A.~Labarre, I.~Rusu, E.~Tannier, S.~Vialette.
  \newblock {\em The combinatorics of Genome rearrangements}.
  \newblock MIT Press.
  \newblock 2009. 
  
\bibitem{Lin2010}
  C.H.~Lin, H.~Zhao, S.H.~Lowcay, A.~Shahab, G.~Bourque.
  \newblock webMGR: an online tool for the multiple genome rearrangement problem.
  \newblock {\em Bioinformatics}, 26:408--410.
  \newblock 2010.
    
\bibitem{Mucha}
  M. Mucha, P. Sankowski.
  \newblock Maximum Matchings via Gaussian Elimination.
  \newblock In {\em FOCS 2004}, pp. 248--255.
  \newblock IEEE Computer Society Press.
  \newblock 2004.
  
\bibitem{Murphy2005}
  W.J.~Murphy {\em et al.}
  \newblock Dynamics of mammalian chromosome evolution inferred from multispecies comparative maps.
  \newblock {\em Science}, 309:613--617. 
  \newblock 2005.

\bibitem{niedermeier-2008}
  R.~Niedermeier.
  \newblock {\em Invitation to Fixed-Parameter Algorithms.}
  \newblock Volume 31 of {\em Oxford Lecture Series in Mathematics and its Applications}.
  \newblock Oxford University Press.
  \newblock 2008.



\bibitem{Tannier2009}
  E.~Tannier, C.~Zheng, D.~Sankoff.
  \newblock Multichromosomal median and halving problems under different genomic distances.
  \newblock {\em BMC Bioinformatics} 10:120. 
  \newblock 2009.

\bibitem{Xu2008}
  A.W.~Xu, D.~Sankoff.
  \newblock Decompositions of Multiple Breakpoint Graphs and Rapid Exact Solutions to the Median Problem.
  \newblock In {\em Algorithms in Bioinformatics (WABI 2008)}, vol. 5251 of {\em Lecture Notes Comput. Sci.}, pp. 25--37.
  \newblock Springer.
  \newblock 2008.

\bibitem{Xu2009a}
  A.W.~Xu.
  \newblock DCJ Median Problems on Linear Multichromosomal Genomes: Graph Representation and Fast Exact Solutions.
  \newblock In {\em Comparative Genomics (RECOMB-CG 2009)}, vol. 5817 of {\em Lecture Notes Comput. Sci.}, pp. 70--83.
  \newblock Springer.
  \newblock 2009.

\bibitem{Xu2009b}
  A.W.~Xu.
  \newblock A Fast and Exact algorithm for the Median of three Problem: a Graph Decomposition Approach.
  \newblock {\em Journal of Computational Biology}, 16:1--13.
  \newblock 2009.

\bibitem{Xu2011}
  A.W.~Xu, B.M.E.~Moret.
  \newblock GASTS: Parsimony Scoring under Rearrangements.
  \newblock In {\em Algorithms in Bioinformatics (WABI 2011)}, vol. 6833 of {\em Lecture Notes Comput. Sci.}, pp. 351--363.
  \newblock Springer.
  \newblock 20111.

\bibitem{Yancopoulos2005}
  S.~Yancopoulos, O.~Attie, R.~Friedberg
 \newblock Efficient sorting of genomic permutations by translocation, inversion and block interchange. 
  \newblock {\em Bioinformatics}, 21:3340-3346.
  \newblock 2005.
 
\bibitem{Zhang2009}
  M.~Zhang, W.~Arndt, J.~Tang.
  \newblock An Exact Median Solver for the DCJ Distance.
  \newblock In {\em Pacific Symposium on Biocomputing (PSB 2009)}, pp. 138--149.
  \newblockï¿½2009.


\bibitem{Zhao2009}
  H.~Zhao, G.~Bourque.
  \newblock Recovering genome rearrangements in the mammalian phylogeny. 
  \newblock {\em Genome Research} 19:934--942.
  \newblock 2009.
\end{thebibliography}


\end{document}
