\documentclass{article}

\usepackage{etoolbox}
\newtoggle{todo}
\newtoggle{arxiv}
\toggletrue{todo}
\toggletrue{arxiv}


\usepackage{microtype}
\usepackage{graphicx}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{booktabs} 

\usepackage{hyperref}
\usepackage[inline]{enumitem}
\usepackage[table]{xcolor}
\usepackage{hhline}
\usepackage{comment}
\usepackage{bbm}

\newcommand{\theHalgorithm}{\arabic{algorithm}}

\iftoggle{arxiv}{
\usepackage[numbers]{natbib}
  \setlength{\textwidth}{6.5in}
  \setlength{\textheight}{9in}
  \setlength{\oddsidemargin}{0in}
  \setlength{\evensidemargin}{0in}
  \setlength{\topmargin}{-0.5in}
  \newlength{\defbaselineskip}
  \setlength{\defbaselineskip}{\baselineskip}
  \setlength{\marginparwidth}{0.8in}
}{




  \usepackage{natbib}
}


\renewcommand{\bibsection}{}



\usepackage{parskip}
\usepackage{xspace}
\usepackage{amsopn,amsmath,amsthm,amssymb}
\usepackage{algorithm}
\usepackage{algorithmic}
\usepackage{breqn}
\usepackage{tabu}
\usepackage{url}
\usepackage{color}
\usepackage{mathtools}
\usepackage{pbox}
\usepackage{scalerel}

\DeclareMathOperator*{\argmax}{argmax} 


\addtolength\textwidth{0.15in}
\addtolength\textheight{0.15in}
\addtolength\textfloatsep{-0.5em}
\addtolength\intextsep{-0.2em}



\usepackage{listings}
\definecolor{codegreen}{rgb}{0,0.6,0}
\definecolor{codegray}{rgb}{0.5,0.5,0.5}
\definecolor{codepurple}{rgb}{0.58,0,0.82}
\definecolor{backcolour}{rgb}{1.0,1.0,1.0}

\lstdefinestyle{mystyle}{
    backgroundcolor=\color{backcolour},   
    commentstyle=\color{codegreen},
    keywordstyle=\color{magenta},
    numberstyle=\tiny\color{codegray},
    stringstyle=\color{codepurple},
    basicstyle=\ttfamily\footnotesize,
    breakatwhitespace=false,         
    breaklines=true,                 
    captionpos=b,                    
    keepspaces=true,                                  
    numbersep=5pt,                  
    showspaces=false,                
    showstringspaces=false,
    showtabs=false,                  
    tabsize=2
}

\lstset{style=mystyle}


\usepackage{upgreek}

\newcommand\todo[1]{\iftoggle{todo}{\textcolor{red}{#1}}{}}

\newcolumntype{N}{>{\centering\arraybackslash}m{.5in}}
\newcolumntype{G}{>{\centering\arraybackslash}m{2in}}
\usepackage[normalem]{ulem}
\useunder{\uline}{\ul}{}

\newcolumntype{C}[1]{>{\centering}m{#1}}




\iftoggle{arxiv}{
  \title{Ontology-driven weak supervision for clinical \\ entity classification in electronic health records}
  \usepackage{authblk}
  \author[1]{Jason A. Fries}
  \author[1,2]{Ethan Steinberg}
  \author[2]{Saelig Khattar}
  \author[1]{Scott L. Fleming}
  \author[1]{\\ Jose Posada}
  \author[1]{Alison Callahan}
  \author[1]{Nigam H. Shah}
  \affil[1]{Center for Biomedical Informatics Research, Stanford University}
  \affil[2]{Department of Computer Science, Stanford University}
}{

}

\date{\today}

\begin{document}

\maketitle


\begin{center}
 Corresponding author: jason-fries@stanford.edu 
\end{center}

\section*{Abstract}

\noindent In the electronic health record, using clinical notes to identify entities such as disorders and their temporality (e.g. the order of an event relative to a time index) can inform many important analyses. 
However, creating training data for clinical entity tasks is time consuming and sharing labeled data is challenging due to privacy concerns. 
The information needs of the COVID-19 pandemic highlight the need for agile methods of training machine learning models for clinical notes. 
We present Trove, a framework for weakly supervised entity classification using medical ontologies and expert-generated rules. 
Our approach, unlike hand-labeled notes, is easy to share and modify, while offering performance comparable to learning from manually labeled training data. 
In this work, we validate our framework on six benchmark tasks and demonstrate Trove's ability to analyze the records of patients visiting the emergency department at Stanford Health Care for COVID-19 presenting symptoms and risk factors. 


\section*{Introduction}
\label{sec:introduction}


Analyzing text to identify concepts such as disease names and their associated attributes like negation are foundational tasks in medical natural language processing (NLP). 
Traditionally, training classifiers for named entity recognition (NER) and cue-based entity classification have relied on hand-labeled training data. 
However annotating medical corpora requires considerable domain expertise and money, creating barriers to using machine learning in critical applications \cite{Ravi2017-zz, Esteva2019-nn}.
Moreover, hand-labeled datasets are static artifacts that are expensive to change.   
The recent COVID-19 pandemic highlights the need for machine learning tools that enable faster, more flexible analysis of clinical and scientific documents in response to rapidly unfolding events \cite{Lu_Wang2020-up}.

To address the scarcity of hand-labeled training data, machine learning practitioners increasingly turn to lower cost, less accurate label sources to rapidly build classifiers. 
Instead of requiring hand-labeled training data, \emph{weakly supervised learning} relies on task-specific rules and other imperfect labeling strategies to programmatically generate training data. 
This approach combines the benefits of rule-based systems, which are easily shared, inspected and modified, with machine learning which typically improves performance and generalization properties. 
Weakly supervised methods have demonstrated success across a range of NLP and other settings \cite{Kuleshov2019-cj,Fries2019-eb,Khattar_undated-zp,DBLP:conf/nips/VarmaSSFFKRXFPR19,Dunnmon2020-jf} .

Knowledge bases and ontologies provide a compelling foundation for building weakly supervised entity classifiers. 
Ontologies codify a vast amount of medical knowledge via taxonomies and example instances for millions of medical concepts. 
However, repurposing ontologies for weak supervision creates challenges when combining label information from multiple sources without access to ground truth labels. 
The hundreds of terminologies found in the Unified Medical Language System (UMLS) Metathesaurus \cite{Bodenreider2004-wh} and other sources \cite{Jonquet2009-jx} typify the highly redundant, conflicting, and imperfect entity definitions found across medical ontologies. 
Naively combining such conflicting label assignments can cause substantial performance drops in weakly supervised classification \cite{Ratner2017-rc}; therefore, a key challenge is correcting for labeling errors made by individual ontologies when combining label information. 

Rule-based systems for NER and cue detection \cite{Chapman2001-or,Peng2018-nl} are common in clinical text processing, where labeled corpora are difficult to share due to privacy concerns. 
Generating imperfect training labels from indirect sources (e.g., patient notes) is often used in analyzing medical images \cite{Wang2017-js,Rajpurkar2018-rk,Draelos2020-dx}. 
Recent work has explored learning the accuracies of sources to correct for label noise when using rule-based systems {to generate training data for text classification \cite{Kuleshov2019-cj,Ratner2019-xo}. 
Weakly supervised clinical applications have explored document and relation classification using task-specific rules \cite{Wang2019-kp,Callahan2019-mz} or leveraging dependency parsing and compositional grammars to automate relation classification for standardizing clinical concepts \cite{peterson2020corpus}.}
However these largely focus on {relation and document classification} via task-specific labeling rules {or sourcing supervision from a single ontology} and do not explore NER or automating labeling via multiple ontologies. 

Prior research on weakly supervised NER has required complex preprocessing to identify possible entity spans \cite{Fries2017-bh}, generated labels from a single source rather than combining multiple sources \cite{Shang2018-sb}, or relied on ad hoc rule engineering \cite{Safranchik2020-no}.
High impact application areas, such as clinical NER using weak supervision, are largely unstudied. 
Recent weak supervision frameworks such as Snorkel \cite{Ratner2017-rc} are domain and task-agnostic, introducing barriers to quickly developing and deploying labeling heuristics in complex domains such as medicine.
Key questions remain about the extent to which we can automate weak supervision using existing medical ontologies and how much additional task-specific rule engineering is required for state-of-the-art performance.
It is also unclear whether, and by how much, pre-trained language models such as BioBERT \cite{Lee2019-xt} improve the ability to generalize from weakly labeled data and reduce the need for task-specific labeling rules. 

We present a Trove, a framework for training weakly supervised medical entity classifiers using off-the-shelf ontologies as a source of reusable, easily automated labeling heuristics.
Doing so transforms the work of using weak supervision from that of coding task-specific labeling rules to defining a target entity type and selecting ontologies with sufficient coverage for a target dataset, which is a common interface for popular biomedical annotation tools such as NCBO BioPortal and MetaMap \cite{Jonquet2009-jx,aronson2010overview}. 
We examine whether ontology-based weak supervision, coupled with recent pre-trained language models such as BioBERT, reduces the engineering cost of creating entity classifiers while matching performance of prior, more expensive, weakly supervised approaches. 
We further investigate how ontology-based labeling functions can be extended when we need to incorporate additional, task-specific rules.
The overall pipeline is shown in Fig. \ref{fig:trove}.

In this work, we demonstrate the utility of Trove through six benchmark tasks for clinical and scientific text, reporting state-of-the-art weakly supervised performance (i.e., using no hand-labeled training data) on NER datasets for chemical/disease and drug tagging. 
We further present weakly supervised baselines for two tasks in clinical text: disorder tagging and event temporality classification.
Using ablation analyses, we characterize the performance trade-offs of training models with labels generated from easily automated ontology-based weak supervision vs. more expensive, task-specific rules.
Finally, we present a case study deploying Trove for COVID-19 symptom tagging and risk factor monitoring using a daily data feed of Stanford Health Care emergency department notes. 



Weakly supervised learning is an umbrella term referring to methods for training classifiers using imperfect, indirect, or limited labeled data and includes techniques such as distant supervision \cite{Craven1999-sl,Mintz2009-zs}, co-training \cite{Blum1998-te} and others \cite{Ma2016-rb}.
Prior approaches for weakly supervised NER such as co-training use a small set of labeled seed examples \cite{Collins1999-js} which are iteratively expanded through bootstrapping or self-training \cite{Medlock2007-qt}. Semi-supervised methods also use some amount of labeled training data and incorporate unlabeled data by imposing constraints on properties such as expected label distributions \cite{Mann2010-yd}.
Distant supervision requires no labeled training data, but typically focuses on a single source for labels such as AutoNER \cite{Shang2018-sb}, which used phrase mining and a tailored dictionary of canonical entity names to construct a more precise labeler, rather than unifying labels assigned using heterogeneous sources of unknown quality.
Crowdsourcing methods combine labels from multiple human annotators with unknown accuracy \cite{Khetan2017-vq}.
However compared to human labelers, programmatic label assignment has different correlation and scaling properties which create technical challenges when combining sources. Data programming \cite{Ratner2016-so,Ratner2017-rc,Ratner2019-xo} formalizes theory for combining multiple label sources with different coverage and unknown accuracy as well as correlation structure to correct for labeling errors. 

In the setting of weakly supervised NER and sequence labeling, SwellShark \cite{Fries2017-bh} uses a variant of data programming to train a generative model using labels from multiple dictionary and rule-based sources.
However this approach required task-specific preprocessing to identify candidate entities \emph{a priori} to achieve competitive performance. 
Safranchik et al. \cite{Safranchik2020-no} presented WISER, a linked hidden Markov model where weak supervision was defined separately over tags and tag transitions using linking rules derived from language models, ngram statistics, mined phrases and custom heuristics to train a BiLSTM-CRF. 
SwellShark and WISER both focused on hand-coded, task-specific labeling function design.

Trove advances weakly supervised medical entity classification by: (1) eliminating the requirement for identifying probable entity spans \emph{a priori} by combining word-level weak supervision with contextualized word embeddings; (2) developing general purpose, more easily automated ontology-based labeling functions which reduce the need for engineering hand-coded rules; (3) quantifying the relative contributions of sources of label assignment -- such as pre-existing ontologies from the UMLS (low cost) and task-specific rule engineering (high cost) -- to the achieved performance for a task; and (4) evaluating Trove in a deployed medical setting, tagging symptoms and risk factors of COVID-19.



\begin{figure}[ht]
\centering
\includegraphics[width=1.0\textwidth]{fig_1}
\caption{\label{fig:trove}
\textbf{Trove pipeline for ontology-driven weak supervision for medical entity classification.}
Users specify: (I) a mapping of an ontology's class categories to entity classes; (II) a set of label sources (e.g., ontologies, task-specific rules) for weak supervision; and (III) a collection of unlabeled document sentences with which to build a training set. 
Ontologies instantiate labeling function templates which are applied to sentences to generate a label matrix.
This matrix is used to train the label model which learns source accuracies and corrects for label noise to predict a consensus probability per word.
Consensus labels are transformed into the probabilistic sequence label dataset  which is used as training data for an end model (e.g., BioBERT). Alternatively, the label model can also be used as the final classifier.}
\end{figure}


\section*{Results}

\subsection*{Experiment overview}
After quantifying the performance of ontology-driven weak supervision in all our tasks, we performed four experiments. 
First, we examined performance differences by label source ablations, which compared ontology-based labeling functions against those incorporating task-specific rules. 
Second, we compared Trove to existing weakly supervised tagging methods. 
Third, we examined learning source accuracies for UMLS terminologies. 
Finally we report on a case study that used Trove to monitor emergency department notes for symptoms and risk factors associated with patients tested for COVID-19. 

We evaluated four methods of combining labeling functions to train entity classifiers.
{
(1) Majority vote (MV) is the majority class for each word predicted by all labeling functions. In cases of abstain or ties, predictions default to the majority class. 
(2) Label model (LM) is the default data programming model. Abstain and ties default to the majority class.
(3) Weakly Supervised (WS) is BioBERT trained on the probabilistic dataset generated by the label model.
(4) Fully supervised (FS) is BioBERT trained on the original expert-labeled training set, tuned to match current published state-of-the-art performance, and using the validation set for early stopping.}

For reference, we also included published F1 metrics for state-of-the-art (SOTA) supervised performance for each task, as determined to the best of our knowledge. 
Note some published SOTA benchmarks (e.g., BC5CDR in Lee et al. \cite{Lee2019-xt}) use both the hand-labeled train and validation sets for training, so they are not directly comparable to our experimental setup. 


\subsection*{Performance of Trove in medical entity classification tasks}
Table \ref{tbl:results} reports F1 performance for weak supervision using ontology-based labeling functions and those incorporating additional, task-specific rules. 
For NER tasks, adding task-specific rules performed within {1.3 - 4.9} F1 points (4.1\%) of models trained on hand-labeled data and for span tasks within 3.4 - 13.3 F1 points. 
The total number of task-specific labeling functions used ranged from 9 to 27. 
For ontology-based supervision, the label model improved performance over MV by {4.1} F1 points on average and BioBERT provided an additional average increase of {0.3} F1 points.  



\begin{table}[H]
\footnotesize
\noindent
\begin{tabular}{|@{}l|C{0.5cm}|C{0.7cm}|C{1.7cm}|C{1.55cm}|C{0.5cm}|C{0.7cm}|C{1.7cm}|C{1.55cm}|cc@{}|}\hline
\multicolumn{1}{|N|}{} & \multicolumn{4}{c|}{Ontologies} & \multicolumn{4}{c|}{+ Task-specific Rules} & \multicolumn{2}{c|}{Hand-labeled} \\ 
\multicolumn{1}{|c|}{\textbf{}} & \multicolumn{4}{c|}{(Guidelines+UMLS+Other)} & \multicolumn{4}{c|}{} & \multicolumn{2}{c|}{} \\  
\hline
Task & LFs & MV & LM       & WS        & LFs & MV & LM        & WS        & FS & SOTA  \\ 
\hline
Chemical     & 22           & 79.8         & 88.0 0.1          & \textbf{88.5 0.2} & +9     & 81.1         & 89.2 0.2          & \textbf{91.1 0.1} & 92.4 0.2  & 93.5 \cite{Lee2019-xt} \\ \hline
Disease      & 16           & 74.7         & \textbf{78.9 0.1} & 78.3 0.2          & +6            & 76.4         & 79.8 0.3          & \textbf{79.9 0.2}  & 84.5 0.2  & 87.2 \cite{Lee2019-xt} \\ \hline
Disorder     & 25           & 67.8         & 68.3 0.3 & \textbf{69.1 0.2} & +11           & 71.2         & 75.0 0.2          & \textbf{76.3 0.1} & 79.6 0.3  & 80.1 \cite{dai2017medication} \\ \hline
Drug         & 16           & 75.3         & 78.6 0.1          & \textbf{79.2 0.2} & +11           & 82.2         & 85.8 0.4          & \textbf{88.3 0.3} & 93.2 0.3  & 91.4 \cite{si2019enhancing} \\ \hline
Negation     & -            & -            & -                  & -                  & 17           & 92.5         & \textbf{93.0 0.0} & 92.7 0.6          & 96.1 0.2  &  \\ \hline
DocTimeRel   & -            & -            & -                  & -                  & 27           & 67.8         & 69.2 0.0          & \textbf{72.9 0.5} & 86.2 0.1  & 83.4  \cite{lin2016multilayered} \\ 
\hline
\end{tabular}

\caption{\label{tbl:results} \textbf{F1 scores for ontology and task-specific rule-based weak supervision}.  Models are majority vote (MV); label model (LM); weakly supervised BioBERT (WS); our fully supervised BioBERT (FS); and published state-of-the-art (SOTA). LFs denote labeling function counts or total added task-specific rules. Bold indicates the best score for each approach and task. Scores are the mean and 1 SD of {n=10} random weight initializations. {A two-sided} Wilcoxon signed-rank test was used to compute statistical significance. 
 denotes p  {0.05} for difference between weakly supervised BioBERT (WS) and the label model (LM). {For (chemical, disease, disorder, drug) exact p-values for ontologies were (0.0039, 0.0020, 0.0020, 0.0020) and for task-specific rules (0.0020, 0.3223, 0.0020, 0.0020). For Negation p=0.0273 and for DocTimeRel p=0.0020.}
 denotes p  {0.05} for difference between the label model (LM) and majority vote (MV). {Here all task p-values were 0.0020.}  
 Mowery et al. \cite{Mowery2014-yx} only reported accuracy for the negation task. 
}
\end{table}



\subsection*{Labeling source ablations}

For NER tasks, we examined five ablations, ordered by increasing cost of labeling effort. 
{
(1) Guidelines, a dictionary of all positive and negative examples explicitly provided in annotation guidelines, including dictionaries for punctuation, numbers, and English stopwords. 
(2) +UMLS, all terminologies available in the UMLS. 
(3) +Other, additional ontologies or existing dictionaries not included in the UMLS.
(4) +Rules, task-specific rules including regular expressions, small dictionaries, and other heuristics.
(5) Hand-labeled, supervised learning using the expert-labeled training split. 
}




Tiers 1-4 are additive and include all prior levels. We initialized labeling function templates as follows: 

{For ontology-based labeling functions,} we used the UMLS Semantic Network and corresponding Semantic Groups as our entity categories and defined a mapping of semantic types (STYs) to target class labels . Non-UMLS ontologies that did not provide semantic type assignments (e.g., ChEBI) were mapped to a single class label. 
All UMLS terminologies  were ranked by term coverage on the unlabeled training set, defined as each term's document frequency summed by terminology, and the top  terminologies were used to initialize templates, where  was tuned with a validation set. 
The remaining  UMLS terminologies were merged into a single labeling function to ensure all terms in the UMLS were included. 
UMLS synsets were constructed using concept unique identifiers (CUIs) and templates were initialized with the union of all terminologies and fixed across all NER tasks.

{For task-specific labeling functions,} we evaluated our ability to supplement ontology-based supervision with hand-coded labeling functions and estimated the relative performance contribution of adding these task-specific rules. 
All training set documents were preprocessed to tag entities using the ontology-based labeling functions outlined above and indexed to support search queries for efficient data exploration. 
The design of task-specific labeling functions is a mix of data exploration, i.e., looking at entities identified by ontology labeling functions to identify errors, and similarity search to identify common, out-of-ontology concept patterns. 
Only the training set was examined during this process and the test set was held out during all labeling function development and model tuning.

For NER, we used two rule types to label concepts: (1) pattern matching via regular expressions and small dictionaries of related terms (e.g., illegal drugs); and (2) bigram word co-occurrence graphs from ontologies to support fuzzy span matching. 
Pattern matching comprised the majority of our task-specific labeling functions. 
While task-specific labeling functions codify generalized patterns not captured by ontologies, we also note that a number of our task-specific labeling functions were necessary due to the idiosyncratic nature of {ground truth} labels in benchmark tasks. 
For example, in the i2b2/n2c2 drug tagging task, annotation guidelines included more complex, conditional entity definitions, such as not labeling negated or historical drug mentions. 
We incorporated these guidelines using the Negation and DocTimeRel labeling functions described below.
See Supplementary Fig. 1 and Supplementary {Note} for a more detailed example of designing task-specific labeling functions.


For span tasks, which classify Negation and DocTimeRel for pre-identified entities, we do not use ontology-based labeling functions directly for supervision.
Instead, ontology-tagged entities were used to guide development of labeling functions that search left and right context windows around a target entity for cue phrases.
Designing search patterns for left and right context windows is the same strategy used by NegEx/ConText \cite{Chapman2001-or,harkema2009context} to assign negation and temporal status.
For Negation, we built on NegEx by adding additional patterns found via exploration of the training documents.
For DocTimeRel we used a heuristic based on the nearest explicit datetime mention (in token distance) to an event mention \cite{Fries2016-ei}.  
Additional contextual pattern matching rules were added to detect other cues of event temporality, e.g., using section headers such as {past medical history} to identify events occurring before the note creation time.


\begin{figure}[H]
  \centering
  \includegraphics[width=1.0\textwidth]{fig_2}
  \caption{\label{fig:ablations} \textbf{Ablation study of F1 performance by labeling source.} Majority vote (MV) vs. weakly supervised BioBERT (WS) vs. fully supervised (FS) for all labeling source ablations showing the absolute F1-score for all labeling tiers.  
{The colored region of each bar indicates MV performance and the white regions denote performance improvements of WS over MV. The mean performance of FS is indicated by the green lines and square points. WS and FS consist of n= experiment replicates using different random initialization seeds, presented as the mean with error bars  SD. MV is deterministic and does not include replicates.}}
\end{figure}



Fig. \ref{fig:ablations} reports F1 scores across all ablation tiers. 
In all settings, the weakly supervised BioBERT models outperformed MV. 
Gains of {8.0 to 34.7} F1 points are seen in the guideline-only tier and {1.3 to 8.2} points in other tiers. 
Incorporating source accuracies into BioBERT training provided significant benefits when combining high precision sources with low precision/high recall sources. 
In the case of chemical tagging with MV, the UMLS tier {(red)} outperformed UMLS+Other {(orange)} by {1.8} F1 points {(81.6 vs. 79.8)}. 
This was due to adding the ChEBI ontology which increased recall but only had 65\% word-level precision. 
Majority vote cannot learn or utilize this information, so naively adding ChEBI labels hurt performance. 
However the label model learned ChEBI's accuracy to take advantage of the noisier, but higher coverage signal, thus the WS UMLS+Other {(orange+white)} outperformed UMLS ({(red+white)}) by {2.5} F1 points ({88.0 vs 85.5}).
See Supplementary Tables 1-4 for complete performance metrics across all ablation tiers.


\subsection*{Comparing Trove with existing weakly supervised methods}

We compared Trove to three existing weakly supervised methods for NER and sequence labeling: SwellShark \cite{Fries2017-bh}, AutoNER \cite{Shang2018-sb}, and WISER \cite{Safranchik2020-no}. 
We compared performance on BC5CDR (the combination of disease and chemical tasks) against all methods and on the i2b2/n2c2 drug task for SwellShark. 
All performance numbers are for models trained on the original training set split, with the exception of SwellShark which is trained on an additional 25,000 weakly labeled documents. 
All weakly supervised methods use the labeling functions, preprocessing, and dictionary curation methods as described in the original manuscripts.
Table \ref{tbl:others} compares Trove with these existing weakly supervised methods. 
Our ontology-based approach outperformed AutoNER by 1.7 F1 points. 
For models incorporating task-specific rules, we outperformed the best weakly supervised model SwellShark by 1.9 F1 points. 
SwellShark reported F1 scores on the i2b2/n2c2 drug task of 78.3 for dictionaries and 83.4 for task-specific rules. 
Our best models achieved 79.2 and 88.4 F1 respectively. 



\begin{table}[H]
\centering
\noindent
\begin{tabular}{|l|l|l|l|c|c|c|} 
\hline
Supervision Method & Label Source & \#Train Docs & End Model & P    & R    & F1  \\
\hline
Fully Supervised            & Hand-labeled          & 500                    & BioBERT            & 87.6          & 89.3          & 88.7          \\
Fully Supervised    & Hand-labeled          & 500                    & BiLSTM-CRF         & 87.2          & 87.9          & 87.5          \\
\hline
SwellShark          & Dictionaries          & 25,500                 & BiLSTM-CRF         & {\ul 84.6}    & 74.1          & 79.0          \\
AutoNER            & Dictionaries          & 500                    & BiLSTM-CRF         & 83.2          & 81.1          & 82.1          \\
Ours (Trove+Snorkel)        & Dictionaries          & 500                    & BioBERT            & 81.6          & {\ul 86.1}    & {\ul 83.7}    \\
\hline
SwellShark         & Custom Rules          & 25,500                 & BiLSTM-CRF         & \textbf{86.1} & 82.4          & 84.2          \\
WISER              & Custom Rules          & 500                    & BiLSTM-CRF         & 82.7          & 83.3          & 83.0          \\
Ours (Trove+Snorkel)        & Custom Rules          & 500                    & BioBERT            & 85.5          & \textbf{86.8} & \textbf{86.1} \\
\hline
\end{tabular}
\caption{\label{tbl:others} \textbf{Comparison of Trove against existing weakly supervised NER methods.}  Precision (P), recall (R), and F1 scores for the BC5CDR task. Underlined numbers indicates the best weakly supervised score using only dictionaries/ontologies and bold indicates the best score using custom rules.
For this task, ontology-based supervision alone outperformed existing weakly supervised methods except for SwellShark which required custom rules and candidate generation. Incorporating task-specific rules into Trove further improved performance.
}
\end{table}



\subsection*{UMLS terminologies as {plug-and-play} weak supervision}

Biomedical annotators such as NCBO BioPortal require selecting a set of target ontologies/terminologies to use for labeling.
Since Trove is capable of automatically combining noisy terminologies, given a shared semantic type definition, we tested the ability to avoid selecting specific UMLS terminologies for use as supervision sources. 
This is challenging because estimating accuracies with the label model requires observing agreement and disagreement among multiple label sources, 
however it is non-obvious how to partition the UMLS, which contains many terminologies, into labeling functions. 
The naive extremes are to either create a single labeling function from the union of all terminologies or include all terminologies as individual labeling functions. 

To explore how partitioning choices impact label model performance, we held all non-UMLS labeling functions fixed across all ablation tiers and computed performance across  partitions of the UMLS by terminology. 
All scores were normalized to the best global majority vote score per tier, selected using the best  choice evaluated on the validation set, to assess the impact of correcting for label noise.

Fig. \ref{fig:umls} shows the impact of partitioning the UMLS into  different labeling functions. 
Modeling source accuracy consistently outperformed MV across all tiers, in some cases by 2-8 F1 points. The best performing partition size  ranged from 1-10 by task. 
The naive baseline approaches -- collapsing the UMLS into a single labeling function or treating all terminologies as individual labeling functions -- generally did not perform best overall. 


\begin{figure}[H]
  \centering
  \includegraphics[width=1.0\textwidth]{fig_3}
  \caption{\label{fig:umls} \textbf{The relationship between the number of UMLS partitions and label model performance.}
  \textbf{a} BC5CDR chemical entities. \textbf{b} BC5CDR disease entities. \textbf{c} ShARe/CLEF 2014 disorder entities. \textbf{d} i2b2/n2c2 2009 drug entities.
  The UMLS is partitioned into  terminologies (x-axis, log-scale) ordered by term coverage on the unlabeled training set. 
  Red (MV) and blue (LM) lines are the mean difference in F1 performance (y-axis) of {n=5} random weight initializations. {Error bars are represented using the solid colored line to denote the mean value of data points and the shaded regions corresponding to SD.}
  The grey region indicates performance worse than the best possible MV, discovered via the validation set. Across virtually all partitioning choices, modeling source accuracies outperformed MV, with 1 to 10 performing best overall.}
\end{figure}



\subsection*{Case study in rapidly building clinical classifiers}
We deployed Trove to monitor emergency departments for patients undergoing COVID-19 testing, analyzing clinical notes for presenting symptoms/disorders and risk factors \cite{Callahan2020-gf}. 
This required identifying disorders and defining a novel classification task for exposure to a confirmed COVID-19 positive individual, a risk factor informing patient contact tracing. 
The dataset consisted of daily dumps of emergency department notes from Stanford Health Care (SHC), beginning in March 2020. 
Our study was approved by the Stanford University Administrative Panel on Human Subjects Research, protocol \#24883
{and included a waiver of consent. All included patients from SHC signed a privacy notice which informs them that their records may be used for research purposes given approval by the IRB, with study procedures in place to protect patient confidentiality.}

We manually annotated a gold test set of 20 notes for all mentions of disorders and 776 notes for mentions of a positive COVID exposure.  
Two clinical experts generated gold annotations which were adjudicated for disagreements by authors AC and JAF. 
As a baseline for disorder tagging, we used the fully supervised ShARe/CLEF disorder tagger.
This reflects a readily available, but out-of-distribution training set (MIMIC-II \cite{saeed2011multiparameter} vs. SHC).
We used the same disorder labeling function set as our prior experiments, adding one additional dictionary of COVID terms \cite{emerse}.  
BioBERT was trained using 2482 weakly-labeled documents.
Custom labeling functions were written for the exposure task and models were trained on 14k sentences.


Table \ref{tbl:covid} contains our COVID case study results. 
The label model provided up to 5.2 F1 points improvement over majority vote and performed best overall for disorder tagging.
Our best weakly supervised model outperformed the disorder tagger trained on hand-labeled MIMIC-II data by 2.3 F1 points.
For exposure classification, the label model provided no benefit, but the weakly supervised end model provided a 6.9\% improvement (+5.2 F1 points) over the rules alone. 


\begin{table}[H]
\centering
\begin{tabular}{|l|c|c|c|c|c|c|c|c|c|c|c|c|c|} 
\hline
\multicolumn{1}{|c|}{\textbf{}} & \multicolumn{1}{c|}{ } & \multicolumn{3}{c|}{MV} & \multicolumn{3}{c|}{LM} & \multicolumn{3}{c|}{{WS}} & \multicolumn{3}{c|}{{FS}} \\ 
\hline
{Supervision} & {Task} & {P}  & {R} & {F1} & {P} & {R} & {F1} & {P} & {R} & {F1} & {P} & {R} & {F1}\\
\hline
Hand-labeled      &  Disorder &  & - & & & - &  & &  - & & 68.0 & 74.5 & 71.1           \\ 
\hline
Ontologies &  Disorder & 64.4 & 66.4 & 65.3 & 69.3 & 71.7 & 70.5 & 67.1 & 72.3 & 69.6 &  & - &  \\ 
+Task-specific &  Disorder & 69.1 & 70.4 & 69.8 & \textbf{73.0} & 73.9 & \textbf{73.4} & 70.5 & \textbf{74.8} & 72.6 & & - &  \\ 
\hline
Task-specific & Exposure & 82.6  & 69.1  & 75.2 & 82.6  & 69.1  & 75.2 & {\ul 87.2} & {\ul 74.5} & {\ul 80.4}  & & - & \\ 
\hline
\end{tabular}
\caption{\label{tbl:covid} \textbf{COVID-19 presenting symptoms/disorders and risk factors evaluated on Stanford Health Care emergency department notes}. Bold and underlined scores indicate the best score in symptom/disorder tagging and COVID exposure classification respectively. Ontology-based weak supervision performed almost as well as the out-of-distribution, hand-labeled MIMIC-II data used for FS. Adding task-specific rules, even though they were developed without seeing Stanford data, outperformed the hand-labeled FS model by 2.3 F1 points. }
\end{table}


\section*{Discussion}

Our experiments demonstrate the effectiveness of using weakly supervised methods to train entity classifiers using off-the-shelf ontologies and without requiring hand-labeled training data. 
Medical ontologies are freely available sources of weak supervision for NLP applications \cite{Rubin2008-vw} and in several NER tasks, our ontology-only weakly supervised models matched or outperformed more complex weak supervision methods in the literature. 
Our work also highlights how domain-aware language models, such as BioBERT, can be combined with weak supervision to build low-cost and highly performant medical NLP classifiers. 

Rule-based approaches are common tools in scientific literature analysis and clinical text processing \cite{Fu2020-hh}. 
Our results suggest that engineering task-specific rules in addition to labels provided by ontologies provides strong performance for several NER tasks -- in some cases approaching the performance of systems built using hand-labeled data. 
We further demonstrated how leveraging the structure inherent in knowledge bases such as the UMLS to estimate source accuracies and correct for label noise provides substantial performance benefits. 
We find that the classification performance of the label model alone is strong, with BioBERT providing modest gains of {1.0} F1 points on average. 
Since the label model is orders of magnitude more computationally efficient to train than BERT-based models, in many settings (e.g., limited access to high-end GPU hardware) the label model alone may suffice.

Our tasks reflect a wide range of difficulty. 
Clinical tasks required more task-specific rules to address the increased complexity of entity definitions and other non-grammatical, sub-language phenomena \cite{Friedman2002-gs}. 
Here custom rules improved clinical tasks an average of {8.1} F1 points vs. {2.1} points for scientific literature. 
Moreover, adding non-UMLS ontologies to PubMed tasks consistently improved overall performance while providing little-to-no benefit for our clinical tasks. 
Annotation guidelines for our clinical tasks also increased complexity. 
The i2b2/n2c2 drug task combines several underlying classification problems (e.g., filtering out negated medications, patient allergies, and historical medications) into a single tagging formulation. This extends beyond entity typing and requires more complex, cue-driven rule design. 

Manually labeling training data is time consuming and expensive, creating barriers to using machine learning for new medical classification tasks. 
Sometimes, there is a critical need to rapidly analyze both scientific literature and unstructured electronic health record data -- as in the case of the COVID-19 pandemic when we need to understand the full repertoire of symptoms, outcomes, and risk factors at short notice \cite{Callahan2020-gf,Shweta2020-wg,Wang2020-dp}. 
However, sharing patient notes and constructing labeled training sets presents logistical challenges, both in terms of patient privacy and in developing infrastructure to aggregate patient records \cite{noauthor_2020-ea}. 
In contrast, labeling functions can be easily shared, edited, and applied to data across sites in a privacy preserving manner to rapidly construct classifiers for symptom tagging and risk factor monitoring. 



This work has several limitations.
Our task-specific labeling functions were not exhaustive and only reflect low-cost rules easily generated by domain experts. 
Additional rule development could lead to improved performance. 
In addition, we did not explore data augmentation or multi-task learning in the BioBERT model, which may further mitigate the need to engineer task-specific rules. 
There is considerable prior work developing machine learning models for tagging disease, drug, and chemical entities that could be incorporated as labeling functions.
However, our goal was to explore performance tradeoffs in settings where existing machine learning models are not available. 
Our framework leverages the wide range of medical ontologies available for English language settings, which provides considerable advantages for weakly supervised methods. 
Additional work is needed to characterize the extent to which the framework can benefit tasks in non-English settings. 

Combining labels from multiple ontology sources violates an independence assumption of data programming as used in this work, because for any pair of source ontologies we may have correlated noise.
This restriction applies to all label sources, but is more prevalent in cases with extremely similar label sources, as can occur with ontologies. 
In our experiments, for a small number of sources, the impact was minor, however performance tended to decrease after including more than 20 ontologies. 
Additional research into unsupervised methods for structure learning \cite{Bach2017-yt,Varma2019-lx}, i.e., learning dependencies among sources from unlabeled data, could further improve performance or mitigate the need to limit the number of included ontologies.  

Identifying named entities and attributes such as negation are critical tasks in medical natural language processing. 
Manually labeling training data for these tasks is time consuming and expensive, creating a barrier to building classifiers for new tasks. 
The Trove framework provides ontology-driven weak supervision for medical entity classification and achieves state-of-the-art weakly supervised performance in the NER tasks of recognizing chemicals, diseases, and drugs.
We further establish new weakly supervised baselines for disorder tagging and classifying the temporal order of an event entity relative to its document timestamp.
The weakly supervised NER classifiers perform within {1.3 - 4.9} F1 points of classifiers trained with hand-labeled data. 
Modeling the accuracies of individual ontologies and rules to correct for label noise improved performance in all of our entity classification tasks. 
Combining pre-trained language models such as BioBERT with weak supervision results in an additional improvement in most tasks.

The Trove framework demonstrates how classifiers for a wide range of medical NLP tasks can be quickly constructed by leveraging medical ontologies and weak supervision without requiring manually labeled training data. 
Weakly supervised learning provides a mechanism for combining the generalization capabilities of state-of-the-art machine learning with the flexibility and inspectability of rule-based approaches.








\section*{Methods}
\label{sec:methods}

\subsection*{Datasets and tasks}
We analyze two categories of medical tasks using six datasets: (1) NER; and (2) span classification where entities are identified \emph{a priori} and classified for cue-driven attributes such as negation or document relative time i.e., the order of an event entity relative to the parent document's timestamp. 
Both categories of tasks are formalized as token classification problems, either tagging all words in a sequence (NER) or just the head words for an entity set (span classification). 
Table \ref{tbl:tasks} contains summary statistics for all six datasets.
All documents were preprocessed using a spaCy \cite{Honnibal2017-si} pipeline optimized for biomedical tokenization and sentence boundary detection \cite{Callahan2019-mz}.


\begin{table}[]
\begin{tabular}{|l|l|l|l|l|l|l|}
\hline
{Task} & {Domain} & {Name}              & {Type} & {\emph{k}} & {Documents} & {Entities} \\ \hline
Disease       & Literature      & BC5CDR \cite{Wei2015-eh}          & NER           & 2                & 500/500/500        & 4182/4244/4424         \\ \hline
Chemical      & Literature      & BC5CDR   \cite{Wei2015-eh}          & NER           & 2                & 500/500/500        & 5203/5347/5385         \\ \hline
Disorder      & Clinical        & ShARe/CLEF 2014 \cite{Mowery2014-yx} & NER           & 2                & 166/133/133        & 5619/4449/7367         \\ \hline
Drug          & Clinical        & i2b2/n2c2 2009 \cite{Uzuner2010-il}       & NER           & 2                & 100/75/75          & 3157/2504/2819         \\ \hline
Negation      & Clinical        & ShARe/CLEF 2014 \cite{Mowery2014-yx} & Span          & 2                & 166/133/133        & 5619/4449/7367         \\  \hline
DocTimeRel   & Clinical        & THYME 2016 \cite{Bethard2016-pe}      & Span          & 4                & 293/147/151        & 38937/20974/18990      \\ \hline
\end{tabular}
\caption{\label{tbl:tasks} \textbf{Dataset summary statistics.} There are () classes per task. The (Documents) and (Entities) columns indicate counts for train/validation/test splits. }
\end{table}


Our COVID-19 case study used a daily feed of emergency department notes from Stanford Health Care (SHC), beginning in March 2020. 
Our study was approved by the Stanford University Administrative Panel on Human Subjects Research, protocol \#24883 and included a waiver of consent. All included patients from SHC signed a privacy notice which informs them that their records may be used for research purposes given approval by the IRB, with study procedures in place to protect patient confidentiality.



We used 99 label sources covering a broad range of medical ontologies. 
We used the 2018AA release of the UMLS Metathesaurus, removing non-English and zoonotic source terminologies as well as sources containing fewer than 500 terms, resulting in 92 sources. 
Additional sources included the 2019 SPECIALIST abbreviations \cite{noauthor_2009-sz}; Disease Ontology \cite{Schriml2012-ne}; Chemical Entities of Biological Interest (ChEBI) \cite{Degtyarenko2008-ni}; Comparative Toxicogenomics Database (CTD) \cite{Davis2015-px}; the seed vocabulary used in AutoNER \cite{Shang2018-sb}; ADAM abbreviations database \cite{Zhou2006-kh}; and word sense abbreviation dictionaries used by the clinical abbreviation system CARD \cite{Wu2017-ji}. 

We applied minimal preprocessing to all source ontologies, filtering out English stopwords and numbers, applying a letter case normalization heuristic to preserve abbreviations, and removing all single character terms. 
We did not incorporate UMLS term type information, such as filtering out terms explicitly denoted as {suppressible} within a terminology, since this information is not typically available in non-UMLS ontologies.    
Our overall goal was to impose as few assumptions as possible when importing terminologies, evaluating their ability to function as {plug-and-play} sources for weak supervision. 


\subsection*{Formulation of the labeling problem}
\label{sec:algo}

We assume a sequence labeling problem formulation, where we are given a dataset  of  sequences  consisting of words  from a fixed vocabulary.
Each sequence is mapped to a corresponding sequence of latent class variables , where  for  tag classes.
Since  is not observable, our primary technical challenge is estimating  from multiple, potentially conflicting label sources of unknown quality to construct a probabilistically labeled dataset . 
This dataset can then be used for training classification models such as deep neural networks.
Such a labeling regimen is typically low-cost, but less accurate than the hand-curated labels used in traditional supervised learning, hence this paradigm is referred to as weakly supervised learning. 

\subsection*{Unifying and denoising sources with a label model}

When using biomedical annotators such as MetaMap or NCBO BioPortal, users specify a target set of entity classes and a set of terminology sources with which to generate labeled concepts.
Consider the example outlined in Fig. \ref{fig:words}, where we want to train an entity tagger for disease names using labels generated from four terminologies. 
Here we are interested in generating a consensus set of entities using each terminology's labeled output.
A straightforward unification method is \emph{majority vote} 

where our  terminologies are represented as individual \emph{labeling functions} .
Labeling functions encode an underlying heuristic such as matching strings against a dictionary and given an input instance (e.g., a document or entity span) assign a label in the domain  where -1 denotes {abstain}, i.e., not assigning any class label. 
Majority vote simply takes the mode of all labeling function outputs for each word, emitting the majority class in the case of ties or abstains. 

Majority vote weights sources equally when combining labels, an assumption that does not hold in practice, which introduces noise into the labeling process. 
Sources have unknown, task-dependent accuracies and often make systematic labeling errors.
Failing to account for these accuracies can negatively impact classification performance.
To correct for such {label noise}, we use data programming \cite{Ratner2016-so} to estimate accuracies of each source and ensemble the sources via a \emph{label model} which assigns a consensus probabilistic label per word. 

To learn the label model,  label sources are parameterized as labeling functions . 
The vector of  labeling functions applied to  instances forms the label matrix . 
A key finding of data programming is that we can use  to recover the latent class-conditional accuracy of each label source without ground truth labels by observing the rates of agreement and disagreement across all pairs of labeling functions  \cite{Ratner2016-so}.
This leverages the fact that while the accuracy  (the expectation of the labeling function output  multiplied by the true label) is not directly observable, the product of   is the rate at which labeling functions vote together, which is observable via .
Assuming independent noise among labeling functions, accuracies are then recoverable up to a sign by solving accuracies for disjoint sets of triplets. 
We refer readers to Ratner et al. (2019) \cite{Ratner2019-xo} for more details. 

We use the weak supervision framework Snorkel \cite{Ratner2017-rc} to train a probabilistic label model which captures the relationship between the true label and label sources . 
Here the training input is the label matrix , generated by applying labeling functions  to the unlabeled dataset . 
Formally,  can be encoded as a factor graph-based model with  accuracy factors between  and our true (unobserved) label  (Fig. \ref{fig:trove}, step 3).

Snorkel implements a matrix completion formulation of data programming which enables faster estimation of model parameters  using stochastic gradient descent rather than relying on Gibbs sampling-based approaches \cite{Ratner2019-xo}. 
The label model estimates  to provide {denoised} consensus label predictions  and generates our probabilistically labeled dataset .   

Fig. \ref{fig:words} shows how data programming provides a principled way to synthesize a label when there is disagreement across label sources about what constitutes an entity span. 
The disease mention {diabetes type 2} is not found in Metathesaurus Names (MTH) or SNOMED Clinical Terms (SNOMEDCT) which leads to disagreement and label errors. 
Using a majority vote of labeling functions misses the complete entity span, while the label model learns to account for systematic errors made by each ontology to generate a more accurate consensus label prediction. 


\begin{figure}[ht]
\centering
\includegraphics[width=1.0\textwidth]{fig_4}
\caption{\label{fig:words}\textbf{An example of combining ontology-based labeling functions}. Here four ontology labeling functions {(MTH, CHV, LNC, SNOMEDCT)} are used to label a sequence of words  containing the entity {diabetes type 2}. Majority vote estimates  as a word-level sum of positive class labels, weighing each equally (). The label model learns a latent class-conditional accuracy () for each ontology, which is used to reweight labels to generate a more accurate consensus prediction of . }
\end{figure}



\subsection*{Labeling function templates}

In this work, a labeling function  accepts an unlabeled sequence  as input and emits a vector of predicted labels , i.e., a label  for each word in . 
A typical labeling function serves as a wrapper for an underlying, potentially task-specific labeling heuristic such as pattern matching with a regular expression or a more complex rule system. 
Since these labeling functions are not easily automated and require hand coding, we refer to them as task-specific labeling functions. 
These are analogous to the rule-based approaches used in 48\% of recent medical concept recognition publications \cite{Fu2020-hh}.

In contrast, medical ontologies can be automatically
transformed into labeling functions with little-to-no custom coding by defining reusable labeling function templates. 
Templates only require specifying a set of target entity categories
and providing a collection of terminologies mapped to those categories. 
These categories are easily derived from knowledge bases such as the UMLS Metathesaurus (where the UMLS Semantic Network \cite{mccray2003upper} provides a consistent categorization of UMLS concepts) or other domain-specific taxonomies.
In this work, we use UMLS Semantic Groups \cite{mccray2001aggregating} (mappings of semantic types into simpler, non-hierarchical categories such as disorders) as the basis for our concept categories.

We explore two types of ontology-based labeling functions, which leverage knowledge codified in medical ontologies for term semantic types and synonymy.

\emph{Semantic type} labeling functions require a set of terms (single or multi-word entities)  mapped to semantic types, where a term may be mapped to multiple entity classes.
This mapping is converted to a -dimensional probability vector where  is the number of entity classes .
Given input sequence , use string matching to find all longest term matches (in token length) and assign each match to its most probable entity class , abstaining on ties.
Using the longest match is a heuristic which helps disambiguates nested terms ({lung} as anatomy vs {lung cancer} as disease).
Matching optionally includes a set of slot-filled patterns to capture simple compositional mentions (e.g., { \{*\} (\{*\})  Tylenol (Acetaminophen)}).

\emph{Synonym (synset) labeling functions} require synsets (collections of synonymous terms)  and terms  mapped to a semantic types.
Given input sequence  and it's parent context (e.g., document) search for 1 unique synonym matches from a target synset and label all matches . 
This is useful for disambiguating abbreviations (e.g, {Duchenne muscular dystrophy  DMD}) , where a long form of an abbreviated term appears elsewhere in a document. 
Matches can be unconstrained, e.g., any tuple found anywhere in a context, or subject to matching rules e.g., using Schwartz-Hearst abbreviation disambiguation \cite{Schwartz2003-mf} to identify out-of-dictionary abbreviations. 

\subsection*{Training the BioBERT end model}
The output of the label model is a set of probabilistically labeled words, which we transform back into sequences .  While probabilistic labels may be used directly for classification, this suffers from a key limitation: the label model cannot generalize beyond the direct output of labeling functions. Rules alone can miss common error cases such as out-of-dictionary synonyms or misspellings. Therefore, to improve coverage we train a discriminative end model, in this case a deep neural network, to transform the output of labeling functions into learned feature representations. Doing so leverages the inductive bias of pre-trained language models \cite{Devlin2019-qr} and provides additional opportunities for injecting domain knowledge via data augmentation \cite{Wei2019-fv} and multi-task learning \cite{Zhang2017-zd} to improve classification performance. 

We use the transformer-based BioBERT \cite{Lee2019-xt}, a language model fine-tuned on biomedical text.
We also evaluated ClinicalBERT \cite{Alsentzer2019-fg} for clinical tasks, and found its performance to be the same as BioBERT.
BioBERT is trained as a token-level classifier with a max sequence length of 512 tokens. 
We follow Devlin et al. \cite{Devlin2019-qr} for sequence labeling formulation, using the last BERT layer of each word's head wordpiece token as the contextualized embedding.
Since sequence labels may be incomplete (i.e., cases where all labeling functions abstain on a word), we mask all abstained tokens when computing the loss during training.
We modified BioBERT to support a noise-aware binary cross entropy loss function \cite{Ratner2016-so} which minimizes the expected value with respect to  to take advantage of the more informative probabilistic labels.

\subsection*{Hyperparameter tuning for the label and end models}
All models were trained using weakly-labeled versions of the original training splits, i.e., no hand-labeled instances. 
We used a hand-labeled validation and test set for hyperparameter tuning and model evaluation, respectively. Result metrics are reported using the test set. 
The label model was tuned for learning rate, training epochs, L2 regularization, and a uniform accuracy prior used to initialize labeling function accuracies. 
BioBERT weights were fine-tuned, and end models were tuned for learning rate and training epochs. 
We used a linear decay learning rate schedule with a 10\% warmup period.
See Supplementary Tables 5-6 for hyperparameter grids.


\subsection*{Metrics}
We report precision, recall, and F1-score for all tasks. 
DocTimeRela is reported using micro-averaging. 
NER metrics are computed using exact span matching \cite{Tjong_Kim2000-cm}. Each NER task is trained separately as a binary classifier using IO (inside, outside) tagging to simplify labeling function design, with predicted tags converted to BIO (beginning, inside, outside) to properly count errors detecting head words. 
Span task metrics are calculated assuming access to gold test set spans, as per the evaluation protocol of the original challenges. 
Label model and BioBERT scores are reported as the mean and standard deviation of 10 runs with different random seeds.
{A two-sided} Wilcoxon signed-rank test with an alpha level of 0.05 was used to calculate statistical significance.

\section*{Data availability}
{
All primary data that support the findings of this study are available via public benchmark datasets (BC5CDR, \url{https://biocreative.bioinformatics.udel.edu/tasks/biocreative-v/track-3-cdr/}) or are otherwise available per data use agreements with the respective data owners (ShARe/CLEF 2014, \url{https://physionet.org/content/shareclefehealth2014task2/1.0/}; THYME, \url{https://healthnlp.hms.harvard.edu/center/pages/data-sets.html}; i2b2/n2c2 2009, \url{https://portal.dbmi.hms.harvard.edu/projects/n2c2-nlp/}).
The data that support the findings of the clinical case study are available on request from the corresponding author JAF. 
These data are not publicly available because they contain information that could compromise patient privacy.}

{
Trove requires access to the UMLS, which is available by license from National Library of Medicine, Department of Health and Human Services, \url{https://www.nlm.nih.gov/research/umls/index.html}.
Open source ontologies used in this study are available at: SPECIALIST Lexicon, \url{https://lsg3.nlm.nih.gov/LexSysGroup/Summary/lexicon.html}; 
Disease Ontology, \url{https://bioportal.bioontology.org/ontologies/DOID};
Chemical Entities of Biological Interest (ChEBI), \url{ftp://ftp.ebi.ac.uk/pub/databases/chebi/};
Comparative Toxicogenomics Database (CTD), \url{http://ctdbase.org};
AutoNER core dictionary, \url{https://github.com/shangjingbo1226/AutoNER/blob/master/data/BC5CDR/dict_core.txt};
ADAM abbreviations database, \url{http://arrowsmith.psych.uic.edu/arrowsmith_uic/adam.html}; and the Clinical Abbreviation Recognition and Disambiguation (CARD) framework, \url{https://sbmi.uth.edu/ccb/resources/abbreviation.htm}.}


\section*{Code availability}
Trove is written in Python v3.6, {spaCy 2.3.4 was used for NLP preprocessing, and} Snorkel v0.9.5 was used for training the label model. 
BioBERT-Base v1.1, Transformers v2.8 \cite{Wolf2019-kd}, and PyTorch v1.1.0 were used to train all discriminative models. 
Trove is open source software and publicly available at \url{https://github.com/som-shahlab/trove}; {https://doi.org/10.5281/zenodo.4497214} \cite{fries-trove}


\section*{References}
\bibliographystyle{naturemag}
\bibliography{trove-2020}

\section*{Acknowledgements}
This work was funded under NLM R01-LM011369-05. 
Thank you to Birju Patel and Keith Morse who did our COVID-19 clinical annotations and to Daisy Ding and Adrien Coulet who helped refine experimental hypotheses during the early stages of this project. 
Computational resources were provided by Nero, a shared big data computing platform made possible by the Stanford School of Medicine Research Office and Stanford Research Computing Center.
Additional thanks to reader feedback from Stephen Pfohl, Erin Craig, Conor Corbin, and Jennifer Wilson.

This is a post-peer-review, pre-copyedit version of an article published in Nature Communications. 
The final authenticated version is available online at: https://doi.org/10.1038/s41467-021-22328-4 


\section*{Author contributions}
JAF conceived the initial study. JAF, ES, SK, SF, JP, AC wrote code and conducted experimental analysis of machine learning models. AC, JAF managed and adjudicated clinical text annotations. JAF, ES, NHS contributed ideas and experimental designs. NHS supervised the project. All authors contributed to writing. 


\section*{Competing Interests}
The authors declare no competing interests.



\clearpage


\section*{Title}
Ontology-driven weak supervision for clinical entity classification in electronic health records



\section*{Supplementary Information}
\subsection*{Supplementary Figures}

\begin{figure}[H]
  \centering
  \includegraphics[width=1.0\textwidth]{supplementary_fig_1}
  \caption{\label{fig:rules} \textbf{Example workflow for developing task-specific labeling functions}. 
  \textbf{a} Users examine documents tagged by  our ontology-based labeling functions and search for common, out-of-ontology entity mentions, in this case ``T wave changes" in ECGs. 
  \textbf{b} Users create a search query to identify similar missing mention patterns in other documents. 
  \textbf{c} Based on the set of documents returned via the search query, users refine their entity pattern into a regular expression which can automatically be used as a labeling function when coupled with a target class label (ShARe/CLEF 2014 disorders in this example).}
\end{figure}

\begin{figure}[H]
\centering
\begin{lstlisting}[language=Python]
# load semantic types and specify entity mapping
entity_classes = {
   "Antibiotic"    :1, 
   "Clinical Drug" :1
}

# define entity categories with classes y \in {0,1}
categories = {
   name:0 if name not in entity_classes else entity_classes[name] 
   for name in umls.semantic_types()
}

# build ontology (t \rightarrow [p_1,..., p_k]) and synsets ({\hat{t}_1,...,\hat{t}_n})
ontology = build_entity_map(umls["SNOMEDCT_US"], categories) 
synsets  = build_synset_map(umls["SNOMEDCT_US"], categories)

# labeling functions
lfs = [
   SemanticTypeLabelingFunction(name="LF_SNOMED", ontology),
   SynSetLabelingFunction(name="LF_SNOMED_synsets", synsets)
]
\end{lstlisting}
\caption{\textbf{Example ontology-based labeling functions.}
Semantic type and synset labeling functions do not require that users manually code rules, only that they specify ontologies with sufficient coverage for an entity class of interest. These examples initialize labeling functions for a simple definition of ``drug" using the SNOMEDCT\_US terminology from the UMLS.
}
\end{figure}


\begin{figure}[H]
\centering
\begin{lstlisting}[language=Python]
rgxs = [
    r"(ACEi|ACE inhibitor[s]*)",
    r"([l][- ](glutathione|arginine))",
    r"([A-Z]){2}[0-9]{3,}",
    r"((alpha|beta|gamma)[-][T])"
]
lf = RegexLabelingFunction(name="LF_chemicals_rgx", rgxs=rgxs, label=1))


rgxs = [
    r"\b([A-Za-z0-9]+?[rlntd]ase[s]*)\b",
    r"[A-Za-z0-9]+ factor[s]*",
    r"\b(anti[a-z]+)\b"
]
lf = RegexLabelingFunction(name="LF_not_chemicals_rgx", rgxs=rgxs, label=0))
\end{lstlisting}
\caption{\textbf{Example task-specific labeling functions.}
Regular expression labeling functions are developed by manually inspecting unlabeled data and identifying common patterns for the entity of interest.
These examples are for chemical tagging in B5CDR.}
\end{figure}


\subsection*{Supplementary Tables}

\begin{table}[H]
\centering
\begin{tabular}{lclccc} \toprule
Task & Method & Ablation Tier & Precision & Recall & F1 \\
\cmidrule(lr){1-6}
Chemical   &   MV   &   Guidelines   &   90.7 0.0   &   3.1 0.0   &   6.0 0.0  \\
Chemical   &   MV   &   Guidelines+UMLS   &   87.0 0.0   &   76.8 0.0   &   81.6 0.0  \\
Chemical   &   MV   &   Guidelines+UMLS+Other   &   74.6 0.0   &   85.7 0.0   &   79.8 0.0  \\
Chemical   &   MV   &   Guidelines+UMLS+Other+Rules   &   78.3 0.0   &   84.2 0.0   &   81.1 0.0  \\
\cmidrule(lr){1-6}
Chemical   &   LM   &   Guidelines   &   90.7 0.0   &   3.1 0.0   &   6.0 0.0  \\
Chemical   &   LM   &   Guidelines+UMLS   &   89.0 0.2   &   82.3 0.2   &   85.5 0.1  \\
Chemical   &   LM   &   Guidelines+UMLS+Other   &   91.0 0.2   &   85.2 0.2   &   88.0 0.1  \\
Chemical   &   LM   &   Guidelines+UMLS+Other+Rules   &   90.8 0.4   &   87.7 0.4   &   89.2 0.2  \\
\cmidrule(lr){1-6}
Chemical   &   WS  &   Guidelines   &   76.0 6.7   &   7.8 3.1   &   14.0 5.0  \\
Chemical   &   WS   &   Guidelines+UMLS   &   87.0 0.1   &   84.6 0.2   &   85.8 0.1  \\
Chemical   &   WS   &   Guidelines+UMLS+Other   &   85.7 0.3   &   91.5 0.2   &   88.5 0.2  \\
Chemical   &   WS   &   Guidelines+UMLS+Other+Rules   &   91.0 0.4   &   91.2 0.3   &   91.1 0.1  \\
\cmidrule(lr){1-6}
Chemical   &   FS   &   Supervised   &   92.1 0.4   &   92.6 0.7   &   92.4 0.2  \\
\bottomrule  
\end{tabular}
\caption{ \label{tbl:chemicals} Complete performance metrics for BC5CDR chemical tagging for all supervision tiers.
Scores are the mean and 1 SD of 5 random weight initializations.
}
\end{table}



\begin{table}[H]
\centering
\begin{tabular}{lclccc} \toprule
Task & Method & Ablation Tier & Precision & Recall & F1 \\
\cmidrule(lr){1-6}
Disease   &   MV   &   Guidelines   &   58.5 0.0   &   6.8 0.0   &   12.3 0.0  \\
Disease   &   MV   &   Guidelines+UMLS   &   67.8 0.0   &   65.2 0.0   &   66.5 0.0  \\
Disease   &   MV   &   Guidelines+UMLS+Other   &   71.9 0.0   &   77.8 0.0   &   74.7 0.0  \\
Disease   &   MV   &   Guidelines+UMLS+Other+Rules   &   74.1 0.0   &   78.7 0.0   &   76.4 0.0  \\
\cmidrule(lr){1-6}
Disease   &   LM   &   Guidelines   &   58.5 0.0   &   6.8 0.0   &   12.3 0.0  \\
Disease   &   LM   &   Guidelines+UMLS   &   70.8 0.9   &   71.3 0.1   &   71.0 0.4  \\
Disease   &   LM   &   Guidelines+UMLS+Other   &   80.9 0.9   &   77.0 0.7   &   78.9 0.1  \\
Disease   &   LM   &   Guidelines+UMLS+Other+Rules   &   81.8 1.1   &   78.0 0.7   &   79.8 0.3  \\
\cmidrule(lr){1-6}
Disease   &   WS   &   Guidelines   &   40.9 6.8   &   51.9 4.8   &   45.1 3.1  \\
Disease   &   WS   &   Guidelines+UMLS   &   69.4 0.4   &   75.2 0.4   &   72.1 0.4  \\
Disease   &   WS   &   Guidelines+UMLS+Other   &   76.9 0.4   &   79.7 0.3   &   78.3 0.2  \\
Disease   &   WS   &   Guidelines+UMLS+Other+Rules   &   78.0 0.4   &   81.9 0.1   &   79.9 0.2  \\
\cmidrule(lr){1-6}
Disease   &   FS   &   Supervised   &   82.6 0.4   &   86.5 0.2   &   84.5 0.2  \\
\bottomrule  
\end{tabular}
\caption{ \label{tbl:diseases} Complete performance metrics for BC5CDR disease tagging for all supervision tiers.
Scores are the mean and 1 SD of 5 random weight initializations.
}
\end{table}



\begin{table}[H]
\centering
\begin{tabular}{lclccc} \toprule
Task & Method & Ablation Tier & Precision & Recall & F1 \\
\cmidrule(lr){1-6}
Disorder   &   MV   &   Guidelines   &   69.2 0.0   &   3.8 0.0   &   7.2 0.0  \\
Disorder   &   MV   &   Guidelines+UMLS   &   76.1 0.0   &   57.8 0.0   &   65.7 0.0  \\
Disorder   &   MV   &   Guidelines+UMLS+Other   &   74.2 0.0   &   62.4 0.0   &   67.8 0.0  \\
Disorder   &   MV   &   Guidelines+UMLS+Other+Rules   &   77.0 0.0   &   66.3 0.0   &   71.2 0.0  \\
\cmidrule(lr){1-6}
Disorder   &   LM   &   Guidelines   &   69.2 0.0   &   3.8 0.0   &   7.2 0.0  \\
Disorder   &   LM   &   Guidelines+UMLS   &   73.2 0.0   &   61.6 0.0   &   66.9 0.0  \\
Disorder   &   LM   &   Guidelines+UMLS+Other   &   74.1 1.4   &   63.3 0.5   &   68.3 0.3  \\
Disorder   &   LM   &   Guidelines+UMLS+Other+Rules   &   79.4 0.8   &   71.1 0.4   &   75.0 0.2  \\
\cmidrule(lr){1-6}
Disorder   &   WS   &   Guidelines   &   35.0 5.0   &   53.9 5.5   &   41.9 2.7  \\
Disorder   &   WS   &   Guidelines+UMLS   &   74.1 0.3   &   64.8 0.5   &   69.1 0.3  \\
Disorder   &   WS   &   Guidelines+UMLS+Other   &   70.8 0.2   &   67.5 0.3   &   69.1 0.2  \\
Disorder   &   WS   &   Guidelines+UMLS+Other+Rules   &   79.4 0.2   &   73.4 0.3   &   76.3 0.1  \\
\cmidrule(lr){1-6}
Disorder   &   FS   &   Supervised   &   77.7 0.5   &   81.7 0.1   &   79.6 0.3  \\
\bottomrule  
\end{tabular}
\caption{ \label{tbl:diseases} Complete performance metrics for ShARe/CLEF 2014 disorder tagging for all supervision tiers.
Scores are the mean and 1 SD of 5 random weight initializations.
}
\end{table}



\begin{table}[H]
\centering
\begin{tabular}{lclccc} \toprule
Task & Method & Ablation Tier & Precision & Recall & F1 \\
\cmidrule(lr){1-6}
Drug   &   MV   &   Guidelines   &   76.2 0.0   &   14.8 0.0   &   24.8 0.0  \\
Drug   &   MV   &   Guidelines+UMLS   &   70.1 0.0   &   81.9 0.0   &   75.5 0.0  \\
Drug   &   MV   &   Guidelines+UMLS+Other   &   69.5 0.0   &   82.0 0.0   &   75.3 0.0  \\
Drug   &   MV   &   Guidelines+UMLS+Other+Rules   &   81.6 0.0   &   82.9 0.0   &   82.2 0.0  \\
\cmidrule(lr){1-6}
Drug   &   LM   &   Guidelines   &   77.5 0.0   &   15.0 0.0   &   25.2 0.0  \\
Drug   &   LM   &   Guidelines+UMLS   &   75.5 0.1   &   79.7 0.0   &   77.5 0.1  \\
Drug   &   LM   &   Guidelines+UMLS+Other   &   75.9 0.1   &   81.5 0.2   &   78.6 0.1  \\
Drug   &   LM   &   Guidelines+UMLS+Other+Rules   &   86.2 0.3   &   85.4 0.7   &   85.8 0.4  \\
\cmidrule(lr){1-6}
Drug   &   WS   &   Guidelines   &   30.0 5.9   &   83.0 1.0   &   43.7 6.2  \\
Drug   &   WS   &   Guidelines+UMLS   &   72.6 0.3   &   83.5 0.1   &   77.7 0.2  \\
Drug   &   WS   &   Guidelines+UMLS+Other   &   75.7 0.2   &   83.0 0.3   &   79.2 0.2  \\
Drug   &   WS   &   Guidelines+UMLS+Other+Rules   &   88.1 0.2   &   88.5 0.3   &   88.3 0.3  \\
\cmidrule(lr){1-6}
Drug   &   FS   &   Supervised   &   93.7 0.3   &   92.7 0.4   &   93.2 0.3  \\
\bottomrule  
\end{tabular}
\caption{ \label{tbl:diseases} Complete performance metrics for i2b2/n2c2 2009 drug tagging for all supervision tiers.
Scores are the mean and 1 SD of 5 random weight initializations.}
\end{table}


\begin{table}[H]
\centering
\begin{tabular}{ll} \toprule
Parameter        & Values \\
\cmidrule(lr){1-2}
learning rate    & [0.01, 0.005, 0.001, 0.0001] \\
l2               & [0.001, 0.0001] \\
epochs           & [50, 100, 200, 600, 700, 1000] \\
precision init   & [0.6, 0.7, 0.8, 0.9] \\
\bottomrule  
\end{tabular}
\caption{\label{tbl:lmgrid} Label model hyperparameter grid.}
\end{table}


\begin{table}[H]
\centering
\begin{tabular}{ll} \toprule
Parameter        & Values \\
\cmidrule(lr){1-2}
learning rate    & [5e-5, 1e-5, 1e-3] \\
epochs           & [5, 25, 50, 100] \\
\bottomrule  
\end{tabular}
\caption{\label{tbl:lmgrid} BioBERT hyperparameter grid.}
\end{table}





\subsection*{Supplementary Note}

Task-specific Rule Design: After using Trove to combine multiple ontologies to label entities, we often want to incorporate additional supervision signal to capture more out-of-ontology entities and further improve classification performance. 
While any existing rule-based system can be used as a labeling functions, either treated as a gestalt, black box labeler or broken down into more modular rules, in this work  we largely focus on regular expression labeling functions.
Regular expressions are flexible, map to a simple supervision paradigm where users are writing search queries, and correspond to how many rule-based systems are designed in practice \cite{Fu2020-hh}.

In Supplementary Fig. \ref{fig:rules} we illustrate an example workflow for developing a labeling pattern which relies on a mix of data exploration and writing search queries.  
We assume all documents are queryable via a search index backend such as Elasticsearch \cite{gormley2015elasticsearch}. First, a user browses a random sample of notes to identify common missing or incorrect entity spans, as generated by our initial ontology-based labeling functions. 
Second, once a target set of missing entities is identified, the user creates a search query to find similar entity mentions, e.g., ``ST-T wave changes" in the example below. 
Finally, the set of retrieved results is used to expand upon a set of regular expressions, which is then mapped to a class label for use as a labeling function.

Since labeling functions consisting of a single pattern generally have low coverage and often low conflict among other labelers, we typically bundle multiple, related regular expressions into a single labeling function to increase coverage. 
This process is repeated until the overall label model performance reaches a target performance threshold. 
























Additional dataset preprocessing: For the DocRelaTime and Negation tasks, labeling functions assume access to explicit datetime mentions (TIMEX3) and clinical event entities (e.g. disorders, drugs, procedures). 
However, our experiments assume machine-learning based entity taggers are not available for these subtasks. 
Instead, we use a dictionary of clinical events derived from the UMLS to tag possible event entities, which are used to generate noisy candidate entities for both Negation and DocRelaTime tasks.
TIMEX3 entities are tagged using regular expressions and normalized into abstractions supporting datetime math.
Labeling functions are applied to these candidates to train the label model, with the resulting probabilistic labels used to train our BioBERT models.
For the ShARe/CLEF tasks we report scores on a subset of the overall disorder entity set, removing non-contiguous, relational-style disorders spans, which comprised 7.9\% (628) of test set mentions.


Guideline annotation examples: These examples are provided directly in annotation guideline documents. 

\begin{itemize}
\item Chemical 
(\href{https://biocreative.bioinformatics.udel.edu/media/store/files/2015/bc5_CDR_data_guidelines.pdf}{BioCreative V CDR Task - Data Annotation Guidelines})
\begin{itemize}
\item Positive
[ATP, Ca, DCE, Fe, K, Li, NO, O2, amino acid, angiotensin II, angiotensin ii, antidepressant, antidepressant drug, antidepressive agent, cAMP, carbidopa, estrogen, estrogen receptor agonist, estrogenic agent, estrogenic compound, estrogenic effect, ethanolic extract of daucus carota seed, fatty acid, glucose, grape seed proanthocyanidin extract, levodopa, low-dose oral contraceptive, nitric oxide, oral contraceptive, phasic oral contraceptive, polyethylene glycol, saturated fatty acid, steroid, sucrose, thymoanaleptics, thymoleptics]
\item Negative
[DNA, adrenergic, anti-HIV agent, anticholinesterase drug, anticoagulant, anticonvulsant, antipsychotic, atom, cellulose, collagen, glucagon, glucocorticoid, glycogen, gold standard, insulin, ion, juice, lipid, lipopolysaccharide, mRNA, molecular, muscarinic, nucleic acid polymer, oligosaccharide, opiate, opioid, opioid alkaloids, opium poppy plant, papaver somniferum, polypeptide, polysaccharide, prolactin, protein, purinergic, saline, starch, water]
\end{itemize}
\end{itemize}

\begin{itemize}
\item Disease
(\href{https://biocreative.bioinformatics.udel.edu/media/store/files/2015/bc5_CDR_data_guidelines.pdf}{BioCreative V CDR Task - Data Annotation Guidelines})
\begin{itemize}
\item {Positive}
[akathisis, auditory toxicity, bone marrow oedema, cancer, cardiac toxicity, death, dyskinesia, erythroblastocytopenia, hepatitis, hypertension, hypertensive, liver toxicity, ototoxicity, ovarian and peritoneal cancer, pain, partial seizures, peritoneal cancer, toxicity, tumor, visual toxicity]
\item {Negative}
[cancerogenesis, complication, deficiencies, deficiency, disease, syndrome, tumorigenesis]
\end{itemize}
\end{itemize}


\begin{itemize}
\item Disorder (\href{https://drive.google.com/file/d/0B7oJZ-fwZvH5VmhyY3lHRFJhWkk/edit}{ShARe/CLEF eHealth 2013 Shared Task: Guidelines for the Annotation of Disorders in Clinical Notes})
\begin{itemize}
\item {Positive}
[bowel obstruction, chest pain, chronic gingivitis, colon cancer, crohn, facial droop, lower extremity DVT, lupus, numbness, pain, rash, schizophrenia, severe pre-eclampsia, small bowel obstruction, stroke, tumor, tumor of the skin, watering of the eye]
\item {Negative} {NONE}
\end{itemize}
\end{itemize}


\begin{itemize}
\item Drug (\href{http://faculty.washington.edu/fxia/mpapers/2009/uzuner2009_i2b2_annot.pdf}{i2b2 Medication Extraction Challenge Preliminary Annotation Guidelines})
\begin{itemize}
\item {Positive}
[CITALOPRAM HYDROBROMIDE, CZI, ECASA, ECASA ( ASPIRIN ENTERIC COATED ), IV fluid, KCL IMMEDIATE REL, LISINOPRIL, NIFEREX TABLET, NITROGLYCERIN 1/150, NTG, POTASSIUM CHLORIDE, TPN, TYLENOL ( ACETAMINOPHEN ), TYLENOL ( ACETAMINOPHEN ), acetaminophen, asa, aspirin, atenolol, avapro, bb, caltrate plus D, caltrate plus D, novolog, diuretic, diuretics, fasting lipids sent, fluocinonide 0.5\% cream, furosemide, glucophage, lasix, lasix, lasix, long acting nitrate, nephrotoxic meds, plavix, red blood cells, saline, saline solution, this medication, total parenteral nutrition, tylenol, tylenol 3, nitroglycerin 1/150, vitamin A, vitamin C, vitamin D, vitamin E, vitamin E, vitamins, vitamins A, vitamins C, vitamins D, vitamins E]
\item {Negative} {NONE}
\end{itemize}
\end{itemize}





\end{document}
