\documentclass[runningheads]{llncs}
\usepackage{amssymb}
\setcounter{tocdepth}{3}
\usepackage{graphicx}
\usepackage{amsmath}
\usepackage[ruled,linesnumbered,vlined]{algorithm2e}
\usepackage{epsfig}
\usepackage{url}

\urldef{\mailsa}\path|{vamsik,rajasek,hieu}@engr.uconn.edu|
\newcommand{\keywords}[1]{\par\addvspace\baselineskip
\noindent\keywordname\enspace\ignorespaces#1}

\begin{document}

\title{An Efficient Algorithm For Chinese Postman Walk on Bi-directed de Bruijn Graphs} 

\titlerunning{Fast CPP Algorithms}

\author{Vamsi Kundeti \and Sanguthevar Rajasekaran \and Heiu Dinh}
\institute{Department of Computer Science and Engineering\\
University of Connecticut\\
Storrs, CT 06269, USA\\
\mailsa\\
} \maketitle

\begin{abstract}
Sequence assembly from short reads is an important problem in biology. It is known that solving the sequence assembly problem exactly on a bi-directed de Bruijn
graph or a string graph is intractable. 
However finding a Shortest Double stranded DNA string (SDDNA) containing all the -long 
words in the reads seems to be a good heuristic to get close to the original genome. This problem is equivalent to finding a cyclic Chinese Postman (CP) walk on the underlying un-weighted 
bi-directed de Bruijn graph built from the reads. The Chinese Postman walk Problem (CPP) is solved by reducing it to a general bi-directed flow on this graph which runs in  time.  

\vspace{0.1in}

In this paper we show that the cyclic CPP on bi-directed graphs can be solved without reducing
it to bi-directed flow. We present a  time
algorithm to solve the cyclic CPP on a weighted bi-directed de Bruijn graph, where 

and . Our algorithm performs asymptotically
better than the bi-directed flow algorithm when the number of {\em imbalanced} nodes 
 is much less than the nodes in the bi-directed graph. From our experimental
results on various datasets, we have noticed that the value of  lies between  and  with  probability.

\vspace{0.1in}

Many practical bi-directed de Bruijn graphs do not have cyclic CP walks. In such cases
it is not clear how the bi-directed flow can be useful in identifying contigs. Our
algorithm can handle such situations and identify maximal bi-directed sub-graphs
that have CP walks. A  time heuristic algorithm based on these ideas
has been implemented for the SDDNA problem. This algorithm was tested on short reads
from a plant genome and achieves an approximation ratio of at most . We also present a  time algorithm for the single 
source shortest path problem on bi-directed de Bruijn graphs, which may be of 
independent interest. 
\end{abstract}

\section{Introduction}

Sequencing the human genome was one of the major scientific breakthroughs in the 
last seven years. Analysis of the sequenced genome can give us vital information about the
expression of genes, which in turn can help scientists to develop drugs for diseases. Thus
sequencing the genome of an organism is of fundamental importance in both medicine and biology.
Unfortunately the technology used in major human genome sequencing projects -- Human Genome
Project (HGP)~\cite{hgp_project} and Celera~\cite{celera_project}, was too expensive to be 
adopted in a large scale. This led to the research on {\em next-generation sequencing} 
methods. {\em Pyro-sequencing} technologies such as SOLiD, 454 and Solexa generate a large number of
short reads which are very accurate. Directed de Bruijn graph based sequence assembly 
algorithms such as \cite{velvet08} and \cite{pevzner01} seem to handle these short read data efficiently 
compared to the string graph based algorithms (see e.g., \cite{myers05}). Unfortunately solving the sequence 
assembly problem exactly on both these graph models seems intractable~\cite{bidirected_graph}. 
However heuristics such as finding a shortest string which includes all the -mers (sub strings 
of length ) seem to yield results close to the original genome. In the case of directed de Bruijn graphs finding an Eulerian tour seems to yield good results. If the graph is not Eulerian then a Chinese Postman (CP) tour has been suggested in~\cite{pevzner01}.
 To account for the double strandedness of the DNA molecule we need to simultaneously search for two complimentary CP tours. 
In~\cite{bidirected_graph} the directed de Bruijn graphs are replaced with bi-directed de Bruijn 
graphs to find two complimentary CP tours simultaneously. A CP tour on the {\em un-weighted} 
bi-directed graph constructed from the reads serves as a solution to the {\em Shortest Double Stranded DNA}  string (SDDNA) problem. The solution presented 
in~\cite{bidirected_graph} solves the SDDNA problem by reducing it to a general weighted bi-directed 
flow problem. This algorithm runs in  time. 

In this paper we present algorithms for SDDNA/CPP  on bi-directed de Bruijn graphs without
using a bi-directed flow algorithm. Our algorithms are based on identifying shortest bi-directed
paths and use of weighted bi-partite matching. Our algorithms perform asymptotically better than the bi-directed flow algorithm when the {\em imbalanced} nodes in the bi-directed graphs are much smaller in number than . This restriction seems to be true in practice from what we have observed in our experiments. On 
the other hand it turns out that in many practical situations these bi-directed de Bruijn graphs 
fail to have {\em cyclic} CP tours. In these cases it is not clear how the bi-directed flow 
algorithm~\cite{bidirected_graph} can help us in identifying a set of {\em contigs} covering every 
-long word at least once. In contrast to this flow algorithm, our algorithm can be useful
in obtaining a {\em minimal} set of contigs when a {\em cyclic} CP tour does no exist. We now
summarize our results as follows. Firstly our deterministic algorithm to solve the {\em cyclic}
CPP on a general bi-directed graph takes  time, 
where , , 
 and .
Secondly we solve the SDDNA problem on an un-weighted bi-directed de Bruijn graph deterministically 
in  time. As a consequence we also present a  
time single source shortest bi-directed path algorithm, which may be of independent interest to 
some assembly algorithms such as Velvet~\cite{velvet08} -- TourBus heuristic.


The organization of the paper is as follows. In Section~\ref{sec:prelim} we provide some preliminaries. 
Section~\ref{sec:prob-def} defines the CPP and SDDNA problems. In Section~\ref{sec:bi-short-path} we
introduce our algorithm for single source shortest bi-directed paths, which is used as a component
in our main algorithm. The main algorithm is introduced in Section~\ref{sec:det_algo} along with
algorithms for several sub-problems. Section~\ref{sec:no-cpp} briefly explains how we can handle situations when
the bi-directed graphs do not have cyclic CP tours. A greedy algorithm that runs in  time
is described in Section~\ref{sec:gdy_algo}. Finally experimental studies are reported in 
Section~\ref{sec:exp}.


\section{Preliminaries}
\label{sec:prelim}
Let  be a string of length . Any substring  (i.e., ) of 
length  is called a mer of . The set of all mer's of a given string  is called the spectrum of 
 and is denoted by . Given a mer ,  denotes the {\em reverse compliment} of  (e.g., if  then ). Let  be the partial ordering among the strings of equal length, 
then  indicates that string  is lexicographically smaller than . Given any mer , 
let  be the lexicographically smaller string between  and . 
We call  the {\em canonical} mer of . More formally, if  then  
else . A molecule of a given mer  is a tuple  
consisting of  and its reverse compliment , the first entry in this tuple is called the 
positive strand and the second entry is called the negative strand.

A {\em bi-directed} graph is a generalized version of a standard directed graph. In a directed graph every 
edge ( or ) has only one arrow head. On the other hand, in a bi-directed graph 
every edge (, , or ) has two arrow 
heads attached to it. Formally, let  be the set of vertices of a bi-directed graph,  is the set of bi-directed edges in a bi-directed graph . 
A {\em walk}  between two nodes  of a bi-directed graph  is a
sequence , such that for 
every intermediate vertex , the orientation of the arrow heads on either side is opposite. To make this more clear let  be the sub-sequence in the walk ,

then for the walk to be valid . If  and  then the walk is called {\em cyclic}. A walk on the bi-directed graph is referred to as
a {\em bi-directed walk}. We define a orientation function  which
gives the orientation of the bi-directed edge between a pair of vertices -- if one exists . 
For instance if  is a bi-directed edge between  and  then 
. An edge which is adjacent on a vertex with a orientation 
 () is called an {\em incoming} ({\em outgoing}) edge. The incoming(outgoing) degree 
of a vertex  is denoted by  (). A vertex  is called {\em balanced}
iff . A vertex is called {\em imbalanced} 
iff . The imbalance of a vertex is called {\em positive} 
iff . Similarly a vertex is {\em negative} imbalanced iff 
. A bi-directed graph is called {\em connected} iff every pair of
vertices have a bi-directed walk between them.


A de Bruijn graph  of the order  on a given string  is defined as follows. The vertex set 
 of  is defined as the spectrum of  (i.e., ). We use the notation 
() to denote the suffix(prefix) of length  in string . The symbol  denotes 
concatenation between two strings. Finally the set of directed edges  of  is defined as follows 
. 
We can further generalize the definition of a de Bruijn graph  on a set  
of strings,  and .

To model the double strandedness of the DNA molecules we should also consider the reverse 
compliments ( while we build the de Bruijn graph. 
To address this a bi-directed de Bruijn graph  has been suggested in 
~\cite{bidirected_graph}. The set of vertices  of  consists of all the 
possible molecules from . For every mer , if  are 
the two mer's of  then an edge is introduced between the molecules () corresponding 
to  and . The orientations of the arrow heads on the edges is chosen as follows. If both  
are the positive strands in  an edge  is introduced. If  is a positive 
strand in  and  is a negative strand in  an edge  is introduced. 
Finally if  is a negative strand in  and  is a positive strand in  an 
edge  is introduced. 

\section{Problem Definitions}
\label{sec:prob-def}
A {\em Chinese Postman walk} in a bi-directed graph is a bi-directed walk which visits every 
edge at least once. A {\em cyclic Chinese Postman walk} of minimum cost on a weighted bi-directed graph is denoted as CPW. The problem of finding a CPW is referred to as CPP. The problem of finding a CPW on an un-weighted 
bi-directed de Bruijn graph (of order ) constructed from a set of reads is called the
{\em Shortest Double stranded DNA} string (SDDNA) problem.  In this paper we give algorithms for the cyclic CPP and SDDNA problems.

\section{Single Source Shortest Path Algorithm on a Bi-directed de Bruijn Graph}
\label{sec:bi-short-path}
We first present an algorithm for the single source shortest path problem on a 
bi-directed de Bruijn graph. The bi-directed de Bruijn graph in the context of 
sequence assembly has non-negative weights on the edges. This makes it possible to 
extend the classic Dijkstra's single source shortest path algorithm to these
graphs. In our algorithm we attach two labels for each vertex in the bi-directed 
graph. Given a source vertex , the algorithm initializes all the labels
similar to Dijkstra's algorithm. In each stage of the algorithm a label with 
the smallest cost is picked and some of labels corresponding to adjacent
nodes are updated. The only major difference between Dijkstra's algorithm 
and our algorithm is the way we update the labels. Dijkstra's algorithm 
updates all the labels/nodes which are adjacent to the smallest label/node
currently picked. However our algorithm updates only those labels/nodes which
are consistent with the bi-directed walk property.  

We now give details of our algorithm and prove its correctness. Let 
 be the bi-directed graph of interest. Also let  be the source 
and  be the destination. We are interested in finding a {\em shortest 
bi-directed} walk from  to . We introduce two labels , 
 for every vertex . The algorithm first initializes
labels corresponding to the source (i.e.  and ) 
to zero. Along with this labels of all the nodes adjacent to  are
also initialized with the corresponding edge weight. The orientation
of the edge determines the label we use for initialization. For instance,
if  is a bi-directed edge with  as the orientation,
the label  is initialized with . On the 
other hand  is left uninitialized. In contrast, if
the orientation of the edge is  then  is initialized
to  and  is left uninitialized. All the uninitialized 
labels contains  by default. 

In each iteration of the algorithm a label with the minimum cost is picked. 
Since we have two types of labels, the minimum label can come from either 
 or . In the first case let  be the node corresponding
to the minimum label during the iteration. This intuitively means that we 
have a path from  to  and the orientation of the edge adjacent to 
 in this path is either  or  -- we are going to 
prove this fact later in the correctness. On the other hand if  is 
different from the destination , then  may possibly appear as an 
internal node in the shortest bi-directed walk between  and . In this 
case the path through  should satisfy {\em bi-directed walk constraint}. 
Thus we should explore only those node(s) adjacent to  with an edge(s) 
orientated as  or . The orientation 
of the edge determines the type of the label we need to update -- similar to 
the label initialization. For instance let  be an edge adjacent on 
 with a orientation of . In this case we should use label 
 to make an update. Similarly if the orientation of the same edge 
is  then  is used in the update process. Consistent
with the classical terminology of the Dijkstra's algorithm, we refer to the 
minimum cost label picked in each iteration as the {\em permanent label}.
For instance if a label  is picked to be the minimum label in an
iteration then we call  as the {\em permanent label} of node .
Now to prove the correctness of the algorithm. It is sufficient to show that the
cost on the permanent label of a node in each iteration is the weight of the shortest
bi-directed path from  to that node. 


\restylealgo{ruled}\linesnumbered
\begin{algorithm}
\SetKwInOut{Input}{INPUT}
\SetKwInOut{Output}{OUTPUT} 
\Input{Bi-directed graph  and two vertices }
\Output{Cost of the shortest bi-directed path between  and }
\\
 \\
 \\
\\
\While{  or }{
	 \\
	 \\
\\
	\If{ or }{
		{\sf return} 
	}
\\
 \\
	\If{}{
		 \\
		 \\
		 \\
		 \\
	}\Else{
		 \\
		 \\
		 \\
		 \\
	}
\\
 \\
	\ForEach{}{
		\\
	}
	\ForEach{}{
		\\
	}
\\
}
 \\
{\sf return} 

\caption{Algorithm to find the shortest bi-directed path from  to  }
\label{algo:shortest_path}
\end{algorithm}

\begin{figure*}
\label{fig:bi-walk-example}
\begin{center}
\includegraphics[scale=0.7]{bi-walk-example.eps}
\end{center}
\caption{{\bf(a)} node  contains two bi-directed walks from node , the green colored path is the shortest.{\bf (b)} the walk starting from node  and ending at node  is a Chinese walk but not a cyclic Chinese walk.}
\end{figure*}

\begin{theorem}
The permanent label of a node  in each iteration of {\bf Algorithm~\ref{algo:shortest_path}} is the weight of the
shortest bi-directed path from  to .
\end{theorem}
\begin{proof}
We prove the statement by induction on the number () of iterations in 
Algorithm~\ref{algo:shortest_path}. We now prove the base case when . Since
we have initialized  and the values of the remaining
both initialized and uninitialized nodes are ; the first iteration picks  and zero
is trivially the cost of shortest bi-directed path form  to . 

Assume that the statement is true for . As per the induction hypothesis the permanent labels 
 correspond to the costs of the shortest 
bi-directed paths between  and . 

Now let 
be the cost of the shortest bi-directed walk from  to . Also let 
 be the path corresponding to the cost .
Note that  cannot be one of the nodes with a permanent label. 
If not, we would have  
(because we should have updated  when the  was given a permanent label) which is a contradiction. 
Now let  be the cost of the shortest path from  to . Clearly, 
 and this means that none of the nodes  haa a permanent label. Since in the iteration  the algorithm updated the labels adjacent to
all the nodes this means that either  or  should have a cost 
and . In each iteration from  we picked the globally 
minimum label  which is a contradiction. 

\end{proof}

We now give a simple example to illustrate the algorithm. Consider the bi-directed graph in 
Figure~\ref{fig:bi-walk-example}(a), with a unit weight on every edge. Let  and  for instance. 
From Figure~\ref{fig:bi-walk-example}(a) we see two bi-directed walks -- {\em red}, {\em green}.
The {\em green} path is the shortest path of length  units. Now let us run our algorithm on this
graph. The algorithm starts with initializing from  and from then on Table~\ref{tab:bi-example}
shows how the  labels are updated until iteration  where we reach the target node  and stop
the algorithm.

\section{Terminal Oriented Shortest Bi-directed Walks}
In the previous section we have seen how to find a shortest bi-directed walk between two nodes in 
a given bi-directed graph. We now define a {\em terminal oriented bi-directed walk} as follows. 
Let  be any 
bi-directed walk between two nodes  and  in a bi-directed graph. Then this bi-directed
walk  is called {\em terminal oriented bi-directed walk} iff  and
. For example in Figure~\ref{fig:bi-walk-example}(a) there are two bi-directed
walks between nodes  and  -- marked with green and red. However only the green bi-directed walk is terminally oriented. A terminal oriented bi-directed walk
 is called the {\em shortest terminal oriented bi-directed walk} iff there is no other terminal
oriented bi-directed walk shorter than . 

\subsection{An algorithm for finding a terminal oriented shortest bi-directed walk}
It is easy to modify Algorithm~\ref{algo:shortest_path} to find a terminal oriented shortest path
between  and . We only have to modify the initialization step and the step which checks
if the target node has been reached. During the initialization at line  of Algorithm~\ref{algo:shortest_path}
we make  and . This avoids the exploration of bi-directed walks which does
not start with . In line , we stop our exploration only if . These changes
ensure that the bi-directed walk at  starts with  and ends with  at . 

\section{A Sufficient Condition for an Eulerian Tour on a Bi-directed Graph} 
The following Lemma~\ref{lem:etour}~\cite{bidirected_graph} is a sufficient condition for a cyclic 
Eulerian tour in a bi-directed graph. A bi-directed graph which has a cyclic Eulerian tour is called
an Eulerian bi-directed graph.
\begin{lemma}
\label{lem:etour}
A connected bi-directed graph is Eulerian if and only if every vertex is balanced.
\end{lemma}
Note that if a bi-directed graph is Eulerian then a cyclic CP walk is the same as a cyclic Eulerian
walk. We emphasize the cyclic adjective for the following reason. Figure~\ref{fig:bi-walk-example}(b) has a 
CP walk starting and ending at vertex . However the CP walk is not cyclic because the walk starts
with  and ends with . The bi-directed graph in Figure~\ref{fig:bi-walk-example}(b)
is not balanced. If the bi-directed graph is not Eulerian, the key strategy to find
a cyclic CP walk is to make it Eulerian by introducing multi-edges into the original graph. The hope is that 
introducing multi-edges would make the bi-directed graph balanced. Thus a cyclic Eulerian walk on a balanced 
multi-edge bi-directed graph would give a cyclic CP walk on the original graph. Since we are interested
in finding a shortest cyclic CP walk, we would like to minimize the number of multi-edges we introduce
in the original graph.

\begin{figure*}
\label{fig:multi-example}
\begin{center}
\includegraphics[scale=0.65]{multi-graph.eps}
\end{center}
\caption{{\bf (a)} a simple bi-directed graph, {\bf (b)} a multi-bi-directed graph. Notice that orientations
of the multi-edges is the same as the orientation of the original edge.}
\end{figure*}

\section{A Deterministic Algorithm to Find a Cyclic CP Walk on a Bi-directed Graph}
\label{sec:det_algo}
We now describe our deterministic algorithm to find a cyclic CP walk on a weighted bi-directed 
graph. First we define a {\em multi-bi-directed graph} as a bi-directed graph in which an edge between two
nodes is overlaid at least once, without changing its orientation. Figure~\ref{fig:multi-example}(a)
shows a bi-directed graph; Figure~\ref{fig:multi-example}(b) shows a valid multi-bi-directed graph. Notice
that while overlaying the edge we did not change its orientation. Since the orientation of the multi-edges
is same as the original edges, any bi-directed walk involving multi-edges is consistent with the bi-directed
walk in the original graph. Another important property of the multi-bi-directed graphs is their ability to make
the nodes balanced. Notice that the vertex  in the original bi-directed graph is positively imbalanced --
. However in the multi-bi-directed graph in Figure~\ref{fig:multi-example}(b)
we are able to balance vertex  by introducing some multi-edges into the original graph. Given a bi-directed
graph , let  be some multi-bi-directed graph corresponding to . The following 
Lemma~\ref{lem:multi-lemma} gives a characterization for  to have a cyclic CP walk. 
\begin{lemma}
\label{lem:multi-lemma}
A non Eulerian bi-directed graph  has a cyclic Chinese Postman walk  a corresponding 
multi-bi-directed graph  which is Eulerian.
\end{lemma}
\begin{proof}
If  has a cyclic Chinese Postman walk, introduce a unique multi-edge in  for every repeated edge in the 
cyclic Chinese Postman walk. This makes the cyclic Chinese Postman walk on the original graph a cyclic Eulerain walk on the
multi-bi-directed graph . Proving the other direction is very similar. 
\end{proof}

Given a multi-bi-directed graph  corresponding to some bi-directed graph , we define the 
{\em multi-bi-directed graph weight} as , where 
is a cost function on the bi-directed graph .  We denote  as the minimum weight Eulerian multi-bi-directed graph 
corresponding to  if at all one exists. The following Lemmas are easy to prove.
\begin{lemma}
\label{lem:reformulate}
Finding a cyclic CP walk on a bi-directed graph  is equivalent to finding a minimum weight Eulerian multi-bi-directed graph  corresponding to . 
\end{lemma}
\begin{lemma}
\label{lem:cost}
If a bi-directed-graph  has a cyclic CP walk then the cost of that walk is equal to the weight of .
\end{lemma}

\subsection{Balancing bi-partite graph}
\label{sec:bal-bi-graph}
Given a bi-directed de Bruijn graph  we define a corresponding {\em Balancing Bi-partite Graph},
 as follows. Let , 
. 
, 
. We now introduce an 
edge between  and  iff  are connected by a {\em terminal oriented
bi-directed walk} from  to . Let  be the weight of
this walk. Then . The weight
of the edge  is the weight of terminal oriented bi-directed walk .

\begin{lemma}
\label{lem:bi-match}
A non Eulerain bi-directed graph  has a cyclic CP walk  the balancing bi-partite graph 
 has a perfect match.
\end{lemma}
\begin{proof}
(Forward direction:) Since  has a cyclic CP walk, every un-balanced node (positive or negative) 
should appear at least  times. Label each occurrence of  in the cyclic CP walk by . Also note that , 
since  has a cyclic CP walk. Now we can pair every  occurrence of a positively imbalanced node  to some
 occurrence of a negatively imbalanced node  since  and  are connected by a terminal oriented bi-directed
walk in the cyclic CP walk. Every such pairing corresponds to a matched edge in .

\vspace{0.1in}

\noindent(Reverse direction:) Consider the perfect match  in . For every edge 
over-lay the underlying oriented bi-directed walk between  on . This makes  a balanced
multi-bi-directed graph. Then by Lemma~\ref{lem:multi-lemma} we can construct a cyclic CP walk in . 
\end{proof}

\subsection{Constructing a family of Eulerian multi-bi-directed graphs}
We now give a construction for generating Eulerian multi-bi-directed graphs corresponding to a given non 
Eulerian bi-directed graph which has a cyclic CP walk. We call this a {\em Balancing Match Family} denoted
by . Lemma~\ref{lem:bi-match} can be used to generate . Assume that  is a non Eulerian bi-directed graph that has a cyclic CP walk. The following construction generates a
family of Eulerian multi-bi-directed graphs corresponding to .
\begin{itemize}
\item {\sf STEP-1:} Create a balancing bi-partite graph  corresponding to  by choosing 
some terminal oriented bi-directed walk between  and .
\item {\sf STEP-2:} Find a perfect match  in . For each edge in  overlay the 
corresponding terminal oriented bi-directed walk on . This generates a Eulerian multi-bi-directed
graph .
\end{itemize}

The following Lemma~\ref{lem:family} is easy to see.
\begin{lemma}
\label{lem:family}
If  is a non Eulerian bi-directed graph that has a cyclic CP walk, then every corresponding
Eulerian multi-bi-directed graph  belongs to the family .
\end{lemma}

The following Lemma gives an expression for the weight of
any .
\begin{lemma}
\label{lem:weight}
Let  be a non Eulerian weighted bi-directed graph which has a cyclic CP walk .
Let  be some Eulerian multi-bi-directed graph. Then, 
, where 
 is a perfect match in .
\end{lemma}
\begin{proof}
Since  is Eulerian it should cover every edge in  - this corresponds to the first term. Secondly,
since  the cost of multi-edges coming from overlaying the terminal bi-directed
walk corresponds to the match  in . This corresponds to the second term. 
\end{proof}

\subsection{An algorithm for finding an optimal cyclic CP walk}
We now put together all the results in the preceding sub-section(s) to give an algorithm to find .
The algorithm is summarized in the following steps.

\begin{itemize}
\item {\sf STEP-1:} We first identify positive and negative imbalanced nodes in . 
Let , 
\item {\sf STEP-2:} Find the cost of a {\em terminal oriented shortest bi-directed walk} between every pair 
. Let this cost be denoted as .
\item {\sf STEP-3:} Create a {\em balancing bi-partite graph}  as follows. Let , , . The cost of an edge . 
\item {\sf STEP-4:} Find a minimum cost perfect match in . Let this match be . If  does not have a
perfect match then  does not have a cyclic CP walk.
\item {\sf STEP-5:} For each edge  , overlay the terminal oriented shortest bi-directed
walk between  and  in the . After overlaying all the terminal oriented bi-directed walks from  on to 
 we obtain . We will prove that it is optimal in Theorem~\ref{thm:thm1}.
\end{itemize}

\begin{theorem}
\label{thm:thm1}
If  is a bi-directed graph that has a cyclic CP walk, then the cost of this cyclic CP walk is equal to 
. Here
 is the min-cost perfect match in the balancing bi-partite graph .
\end{theorem}
\begin{proof}
By Lemma~\ref{lem:family} the multi-bi-directed graph  belongs to . Now by Lemma~\ref{lem:weight},
any optimal solution has to minimize the second term (). To minimize this the
algorithm chooses {\em shortest terminal oriented bi-directed walk} in {\sf STEP-2}. Finally, in {\sf STEP-5} the 
algorithm finds a minimum cost perfect match. Both these steps ensure that  is minimum in the entire
family of multi-bi-directed graphs in .
\end{proof}

\subsection{Runtime analysis of the algorithm to find a cyclic CP walk}
Let  and .
{\sf STEP-2} of the algorithm runs in  time to compute . In 
{\sf STEP-3}  , . For {\sf STEP-4} Hungarian method can be applied
to solve the weighted matching problem in  time. So the total runtime of this
deterministic algorithm is . As mentioned before
if  is much smaller than  this algorithm performs better than the bi-directed flow algorithm.
\subsection{Runtime analysis of the algorithm to find SDDNA}
Since SDDNA runs on a bi-directed de Bruijn graph which is un-weighted, {\sf STEP-2} of the algorithm
runs in  time -- because we don't need to use a Heap, we just do a BFS on the bi-directed
graph. The rest of the analysis for the runtime remains the same and the total run time of the algorithm 
is .


\section{A  Time Heuristic Algorithm for the SDDNA Problem}
\label{sec:gdy_algo}
From the analysis in the previous section, to solve the SDDNA problem deterministically we need to spend
 time. However if we just replace the Hungarian method in {\sf STEP-4}
with a simple greedy algorithm we can get rid of the  term in the asymptotic complexity.
Although we have a constant  approximation algorithm for {\em maximum weighted matching},
we are not aware of any constant approximation algorithms for {\em minimum weight perfect matching}.
As a result this just remains as a heuristic algorithm. On the other hand this algorithm seems to be
performing very close to the optimal (see Section~\ref{sec:gdy_heuristic}).

\section{Dealing with Practical Bi-directed de Bruijn Graphs with no Cyclic CP Walks}
\label{sec:no-cpp}
As we have mentioned earlier most of the bi-directed de Bruijn graphs constructed from the
reads do not satisfy the sufficient condition for cyclic CP walks. In such cases our algorithm
can still be used, by modifying it to find a {\em maximum} match in the balancing bi-partite graph  
rather than perfect match. We can introduce a hypothetical node  and connect all the 
un-matched nodes in the balancing bi-partite graph to  with appropriate bi-directed edges and
thus make all the original nodes balanced. We can now find a cyclic CP walk in this hypothetical 
graph. Every sub-walk in the cyclic CP walk that starts from  and ends at  can be reported
as a {\em contig}. Thus our algorithm is capable of handling cases when the bi-directed graph cannot have a cyclic CP walk.

\section{Experimental Results}
\label{sec:exp}
As we have mentioned in the previous sections the asymptotic complexity of our algorithm depends on 
 -- the maximum of positively and negatively imbalanced nodes. In the case of de Bruijn
graphs , where  is the size the alphabet from which the strings
are drawn. In our case this is exactly four. So we can safely ignore  in the case of de Bruijn
graphs and just concentrate on . In the rest of the discussion we would like to refer to 
as the number of imbalanced nodes.
\subsection{Estimation of the mean of the random variable }
It is clear that  is a random variable with support in . So we would like to 
estimate the expected number of imbalanced nodes in a graph with  bi-directed edges. We
estimated the mean of the random variable  from several samples of bi-directed
de Bruijn graphs constructed from reads from a plant genome. A simple test is applied to 
to estimate the  confidence interval of . See Table~\ref{tab:imbal} for
the details of the samples used. Notice that as we increase the size of  (de Bruijn graph
order) from  to , the number of imbalanced nodes in columns corresponding to 
and  reduces. This is because increasing  reduces the number of edges which may reduce
the number of imbalanced nodes. On the other hand for a fixed value of  the number of
imbalanced nodes increases consistently with the nodes. However the rate of growth is very
slow compared to the rate of growth of the number of nodes. Finally we use this evidence to hypothesize
that the number of imbalanced nodes in practical bi-directed graphs is only between 
to  of the number of nodes in the graph, with a probability of .
\subsection{Performance of the greedy heuristic and handling cases which do not have cyclic CP walks}
\label{sec:gdy_heuristic}
The greedy heuristic described in Section~\ref{sec:gdy_algo} has been compared with the optimal
{\em maximum match} with {\em minimum cost}. As we mentioned in Section~\ref{sec:no-cpp} many 
of these graphs do not contain cyclic CP walks so they do not have a perfect match. To cope with
this situation we treated the balancing bi-directed graph as a complete bi-directed graph, by
introducing a hypothetical edge with large cost whenever there is no edge between two nodes
in the original graph. Thus we just used the size of the match to compare the cost of the greedy
algorithm and the optimal algorithm to solve the matching problem. Table~\ref{tab:appx} gives
the details of the balancing bi-partite graph obtained from several read samples. As we have
mentioned before, to get the approximation ratio we treated a balancing bi-partite graph as a complete
bi-partite graph . If  and  are the sizes of {\em maximum} and {\em maximal}
matches then we treated the cost of hypothetical perfect match as  and 
and their ratio is used as approximation ratio. Finally from the evidence in Table~\ref{tab:appx}
we hypothesize that the approximation ratio for this is between  and  with a probability
of .

\subsection{Implementation and Data}
An implementation of the algorithms discussed is available at \url{http://trinity.engr.uconn.edu/~vamsik/fast_cpp.tgz}.

\section{Conclusion and further research}
In this paper we have given an algorithm for cyclic Chinese Postman walk on a bi-directed de Bruijn graph. Our
algorithm is based on identifying shortest bi-directed walks and weighted matching. This algorithm performs
asymptotically better than the bi-directed flow algorithm when the number of imbalanced nodes are much smaller
than the nodes in the bi-directed graph. On the other hand this algorithm can also handle the instances of
bi-directed graphs which does not have a cyclic CP walk and provide a minimal set of walks, cyclic walks
which cover every edge in the bi-directed graph at least once. 

There are several research directions which can be pursued. Firstly, we need to address how the addition of
paired reads may impose new constraints on the cyclic CPP walk. Secondly, while Eulerization of the bi-directed graph
we have chosen the shortest path bi-directed path, however this may not correspond to the repeating region in 
the genome. Other strategy to make the graph Eulerian is to choose the path with maximum read multiplicity. 
This on other hand may increase the length of the Chinese walk, can we simultaneously optimize these two 
objectives ?.


\begin{table}
\begin{center}
\begin{tabular}{|c|c|c|c|c|c|c|c|c|}
\hline
READS &   & NODES & P-IMBAL & N-IMBAL & \multicolumn{3}{|c|}{BAL-BI-GRAPH} & \\
\hline
 &&&&& &  &  &  \\
\hline
102400 & 21 & 1588569 & 1157 & 1133 & 1186 & 1173 & 1186 & 0.075 \\ 
\hline
153600 & 21 & 2353171 & 2240 & 2141 & 2298 & 2211 & 2298 & 0.098 \\ 
\hline
204800 & 21 & 3097592 & 3509 & 3492 & 3601 & 3590 & 3601 & 0.116 \\ 
\hline
256000 & 21 & 3825101 & 4953 & 5004 & 5074 & 5131 & 5131 & 0.134 \\ 
\hline
307200 & 21 & 4538734 & 6719 & 6748 & 6878 & 6912 & 6912 & 0.152 \\ 
\hline
358400 & 21 & 5235821 & 8586 & 8603 & 8789 & 8802 & 8802 & 0.168 \\ 
\hline
409600 & 21 & 5917489 & 10665 & 10693 & 10914 & 10934 & 10934 & 0.185 \\ 
\hline
102400 & 25 & 1202962 & 569 & 521 & 588 & 540 & 588 & 0.049 \\ 
\hline
153600 & 25 & 1788533 & 1104 & 1026 & 1139 & 1062 & 1139 & 0.064 \\ 
\hline
204800 & 25 & 2362981 & 1744 & 1708 & 1788 & 1759 & 1788 & 0.076 \\ 
\hline
256000 & 25 & 2927656 & 2521 & 2523 & 2579 & 2592 & 2592 & 0.089 \\ 
\hline
307200 & 25 & 3484849 & 3370 & 3414 & 3451 & 3517 & 3517 & 0.101 \\ 
\hline
358400 & 25 & 4032490 & 4333 & 4369 & 4441 & 4485 & 4485 & 0.111 \\ 
\hline
409600 & 25 & 4571554 & 5390 & 5467 & 5518 & 5613 & 5613 & 0.123 \\ 
\hline
\multicolumn{9}{|c|}{}\\ 
 \multicolumn{9}{|c|}{ :  95\% C.I for average  is } \\ 
\hline
\end{tabular}
 \end{center}
\caption{The value of  on short read data from a plant genome sequencing data from CSHL}
\label{tab:imbal}
\end{table}

\begin{table}
\begin{center}
\begin{tabular}{|c|c|c|c|c|c|c|c|c|}
\hline
READS &   & NODES &  &\multicolumn{2}{|c|}{OPT} & \multicolumn{2}{|c|}{GRDY} & APX-RATIO\\ 
\hline
 & & & &SIZE&COST&SIZE&COST \\
\hline
 &&&& &  &  &  &  \\
\hline
102400 & 21 & 1202962 & 1186 & 416 & 770 & 406 & 780 & 1.0130 \\ 
\hline
153600 & 21 & 1788533 & 2298 & 725 & 1573 & 704 & 1594 & 1.0134 \\ 
\hline
204800 & 21 & 2362981 & 3601 & 1092 & 2509 & 1073 & 2528 & 1.0076 \\ 
\hline
256000 & 21 & 3825101 & 5131 & 1479 & 3652 & 1450 & 3681 & 1.0079 \\ 
\hline
307200 & 21 & 4538734 & 6912 & 1929 & 4983 & 1879 & 5033 & 1.0100 \\ 
\hline
358400 & 21 & 5235821 & 8802 & 2385 & 6417 & 2329 & 6473 & 1.0087 \\ 
\hline
409600 & 21 & 5917489 & 10934 & 2876 & 8058 & 2814 & 8120 & 1.0077 \\ 
\hline
102400 & 25 & 1202962 & 588 & 152 & 436 & 147 & 441 & 1.0115 \\ 
\hline
153600 & 25 & 1788533 & 1139 & 281 & 858 & 274 & 865 & 1.0082 \\ 
\hline
204800 & 25 & 2362981 & 1788 & 438 & 1350 & 429 & 1359 & 1.0067 \\ 
\hline
256000 & 25 & 2927656 & 2592 & 619 & 1973 & 601 & 1991 & 1.0091 \\ 
\hline
307200 & 25 & 3484849 & 3517 & 809 & 2708 & 783 & 2734 & 1.0096 \\ 
\hline
358400 & 25 & 4032490 & 4485 & 995 & 3490 & 966 & 3519 & 1.0083 \\ 
\hline
409600 & 25 & 4571554 & 5613 & 1220 & 4393 & 1175 & 4438 & 1.0102 \\ 
\hline
\multicolumn{9}{|c|}{}\\ 
 \multicolumn{9}{|c|}{ :  95\% C.I for average  is } \\ 
\hline
\end{tabular}
 \end{center}
\caption{Approximation ratio of the {\sf GDY} heuristic}
\label{tab:appx}
\end{table}


\noindent {\bf Acknowledgements.} This work has been supported in
part by the following grants: NSF 0326155, NSF 0829916 and NIH
1R01GM079689-01A1.

\section*{Appendix}

\begin{table}
\begin{center}
\begin{tabular}{|llllll|}
\hline
\multicolumn{6}{|c|}{\text{iteration: , }}\\
\hline
& & & & & \\
\hline 
& & & & & \\
\hline
\hline
\multicolumn{6}{|c|}{\text{iteration: , }}\\
\hline
& & & & & \\
\hline 
& & & & & \\
\hline
\hline
\multicolumn{6}{|c|}{\text{iteration: , }}\\
\hline
& & & & & \\
\hline 
& & & & & \\
\hline
\hline
\multicolumn{6}{|c|}{\text{iteration: , }}\\
\hline
& & & & & \\
\hline 
& & & & & \\
\hline
\hline
\multicolumn{6}{|c|}{\text{iteration: , }}\\
\hline
& & & & & \\
\hline 
& & & & & \\
\hline
\hline
\multicolumn{6}{|c|}{\text{iteration: , }}\\
\hline
& & & & & \\
\hline 
& & & & & \\
\hline
\hline
\multicolumn{6}{|c|}{\text{iteration: , }}\\
\hline
& & & & & \\
\hline 
& & & & & \\
\hline
\hline
\multicolumn{6}{|c|}{\text{\sf The shortest path from node  to node  is of length }} \\
\hline
\end{tabular}
 \end{center}
\caption{Changes in the  and  labels in every iteration of Algorithm~\ref{algo:shortest_path} on
the bi-directed graph in Figure~\ref{fig:bi-walk-example}}
\label{tab:bi-example}
\end{table}

\bibliographystyle{splncs}
\begin{thebibliography}{1}

\bibitem{hgp_project}
Lander, E.S., Linton, L.M., Birren, B., Nusbaum, C., Zody, M.C.e.a.:
\newblock Initial sequencing and analysis of the human genome.
\newblock Nature \textbf{409} (2001)  860--921

\bibitem{celera_project}
Craig~Venter, J., Adams, M.D., Myers, E.W., Li, P.W., Mural, R.J.e.:
\newblock The sequence of the human genome.
\newblock Science \textbf{291} (2001)  1304--1351

\bibitem{velvet08}
Zerbino, D.R., Birney, E.:
\newblock Velvet: Algorithms for de novo short read assembly using de bruijn
  graphs.
\newblock Genome research \textbf{18} (2008)  821--829

\bibitem{pevzner01}
Pevzner, P.A., Tang, H., Waterman, M.S.:
\newblock An eulerian path approach to dna fragment assembly.
\newblock Proceedings of the National Academy of Sciences of the United States
  of America \textbf{98} (2001)  9748--9753

\bibitem{myers05}
Myers, E.W.:
\newblock The fragment assembly string graph.
\newblock Bioinformatics \textbf{21} (2005)  ii79--ii85

\bibitem{bidirected_graph}
Medvedev, P., Georgiou, K., Myers, G., Brudno, M.:
\newblock Computability of models for sequence assembly. Volume 4645 LNBI.
\newblock (2007)

\end{thebibliography}

\end{document}
