

\typeout{IJCAI--22 Instructions for Authors}



\documentclass{article}
\pdfpagewidth=8.5in
\pdfpageheight=11in
\usepackage{ijcai22}

\usepackage{times}
\usepackage{soul}
\usepackage{url}
\usepackage[hidelinks]{hyperref}
\usepackage[small]{caption}
\usepackage{graphicx}
\usepackage{amsmath}
\usepackage{amsthm}
\usepackage{booktabs}
\usepackage{algorithm}
\usepackage{algorithmic}
\usepackage{multirow}
\usepackage{graphicx}
\usepackage{diagbox}
\usepackage{amssymb}



\usepackage{float}                  \usepackage{subfig}                 \usepackage{overpic}                \usepackage[T1]{fontenc} 

\usepackage{color}

\urlstyle{same}














\newtheorem{example}{Example}
\newtheorem{theorem}{Theorem}










\title{TransEDRP: Dual Transformer model with Edge Emdedded for Drug Respond Prediction}





\author{
Kun Li
\and
Wenbin Hu\
\affiliations
School of Computer Science, Wuhan University
\emails
\{li\_\_kun, hwb\}@whu.edu.cn
}


\begin{document}

\maketitle

\begin{abstract}


GNN-based methods have achieved excellent results as a mainstream task in drug response prediction tasks in recent years. Traditional GNN methods use only the atoms in a drug molecule as nodes to obtain the representation of the molecular graph through node information passing, whereas the method using the transformer can only extract information about the nodes. However, the covalent bonding and chirality of a drug molecule have a great influence on the pharmacological properties of the molecule, and these information are implied in the chemical bonds formed by the edges between the atoms. In addition, CNN methods for modelling cell lines genomics sequences can only perceive local rather than global information about the sequence. In order to solve the above problems, we propose the decoupled dual  transformer structure with edge embedded for drug respond prediction (TransEDRP), which is used for the representation of cell line genomics and drug respectively. For the drug branch, we encoded the chemical bond information within the molecule as the embedding of the edge in the molecular graph, extracted the global structural and biochemical information of the drug molecule using graph transformer. For the branch of cell lines genomics, we use the multi-headed attention mechanism to globally represent the genomics sequence. Finally, the drug and genomics branches are fused to predict IC50 values through the transformer layer and the fully connected layer, which two branches are different modalities. Extensive experiments have shown that our method is better than the current mainstream approach in all evaluation indicators.















\end{abstract}

\section{Introduction}













With the development of deep learning and big data, several scholars have applied graph computation theory to the task of drug response prediction(DRP). Compared with methods based on non-graph computational deep learning \cite{tcnns,MOLI,CDRscan}, GNNs methods have achieved better performance in the task of drug molecule representation \cite{DeepCDR,18bMultipleOmic,GraphDTA,110bGCNforDRP,GraTransDRP}.Usually, the DRP method based on graph neural networks is to convert the drug molecule into a graph structure  where nodes are the atoms and the edges are the bonding of atoms. The atom-based graph representation constructs a defined graph structure for each drug molecule, which can more completely and stably represent the atomic features and the chemical bonds between atoms, and helps the model to accurately predict the biochemical properties of the molecule.
In particular, GraTransDRP \cite{GraTransDRP} has achieved state-of-the-art results to date, which combined with Graph and Transformer for predicting cancer drug response.





















\begin{figure}[htpb]
    \centering
    \includegraphics[width=.95\linewidth]{img/introduction1.jpg}
    \caption{The process of diagrammatic characterisation of a drug molecule, using the chemical bonds between atoms as edges. The information on the edges includes chirality, covalent bond, aromatic, etc.}
   	\label{fig:intro1}
\end{figure}







	






However, most existing methods based on GNNs assume that different types of intramolecular chemical bonds are the same type. Therefore, they apply permutation-invariant local aggregation of neighbors to learn a representation for each node without edge information, which omits important information about chemical bonds.

In fact, most atoms in chemical drug molecules have different labels or features, the types of chemical bonds in a molecule are different  (see Figure \ref{fig:intro1} b and c). In addition, molecules with atoms in the same order of interconnection but with double bonds and ring structures lead to the creation of chiral compounds. Most of the drugs used today are small organic molecules consisting of less than 50 atoms, a large proportion of which are chiral, and their pharmacological effects are realized through strict chiral matching and molecular recognition with large molecules in vivo. The enantiomers of chemical drugs containing chiral factors differ significantly in their pharmacological activity, metabolic processes and toxicity in the human body. 

As a result, often only one of the two isomers is effective and the other is ineffective or even harmful. For example (see Figure \ref{fig:intro1} a), thalidomide could be used to treat multiple myeloma, but due to technical problems the phenomenon of molecular chirality was ignored, leading to a failure to distinguish between the R-configuration of the compound, which has an inhibitory and sedative effect on pregnancy, and the S-configuration, which has teratogenic properties. Ultimately, tens of thousands of babies were malformed or died in the 1960s.



\begin{figure}[htbp]
    \centering
	\includegraphics[width=.985\linewidth]{img/introduction2.jpg}
	\caption{Receptive field of 2-layer CNN (a) vs. 1-layer multi-head attention (b), where we set the kernel size to 1, stride to 1 and dilation to 0.}
	\label{fig:intro2}
\end{figure}







1D-CNN is widely used for extraction of features in cell genomics  \cite{tcnns,110bGCNforDRP,DeepCDR,GraTransDRP,GraphDTA}.
Although CNN has  time and space complexity and can process long sequences, it is mainly used to extract local features of sequences, which a single operation of its convolution kernel can only compute a small range in the sequence(see Figure \ref{fig:intro2}). 
Since the receptive field expands faster in the middle of the sequence and slower in the sides, the CNN can not make fair use of the information at all positions, resulting in the sequence position having an impact on the output of the convolution (sequences  of cell genomics  are position-independent).








 We propose TransEDRP  to solve the above problems mentioned above separately, which is the dual transformer structure.



\begin{itemize}
  \item [1)] 
For the representation of drug molecules, we have creatively introduced molecular chirality and covalent bond types to encode information on intramolecular chemical bonds, and designed a graph transformer with edge embedded network to effectively combine intramolecular chemical bond information to represent molecules.
  \item [2)] 
For the representation of genomics of cell lines, we designed a multi-head attention feature extraction module  that allows similarity to be calculated between any two gene expression terms in genomics sequence and is used to improve the ability of the model to express complex dependencies in genomics sequences.
\end{itemize}





We fused the outputs of the two branches in a base concatenation way and then predicted  by MLP and activation function. Noting that the two branches are decoupled, allowing the flexibility to modify either branch without affecting the other, which improves the practicality and scalability of the model.

Extensive comparative experiments have been done on mainstream datasets Genomicsof Drug Sensitivity in Cancer (GDSC) \cite{GDSC} and Cancer Cell Line Encyclopedia (CCLE) \cite{CCLE} to show that TransEDRP achieved state-of-the-art performance in all evaluation metrics, where the comparative methods are included tCNNs, GraphDRP, GraphTransDRP, etc. In addition, we have performed reliable ablation experiments and parametric sensitivity analyses to demonstrate the validity and robustness of our model. 








\section{Related Work}

With the completion of large-scale projects on drug response in cell lines, such as GDSC \cite{GDSC}, CCLE \cite{CCLE}, there is data support for drug response prediction.

The expression of drug molecules can utilize the Pubchem ID recorded in GDSC to get the line notation of simplified molecularinput line entry system (SMILES) from Pubchem, which was initially used to extract the structural and chemical features of each drug.

In addition, the release of high-throughput sequencing data for cancer cell lines has made it possible to study cancer phenotypes using multi-omics data such as genomics, transcriptomics and epigenomics. For example, the CCLE has published approximately 1,000 cancer cell lines and their corresponding holographic profiles. The half maximal inhibitory concentration () is the half inhibitory concentration of an antagonist being measured. It indicates the half amount of a drug or substance (inhibitor) that inhibits a biological process (or a substance included in such a process, such as an enzyme, cell receptor or microorganism).





Based on these data, many researchers have conducted studies on drug sensitivity prediction tasks. These methods can be broadly classified as graph based and no-graph based.

The no-graph based methods generally use neural network models of CNN and MLP for information extraction,e.g. CDRscan \cite{CDRscan}, tCNNs \cite{tcnns},  MOLI \cite{MOLI}, etc.
CDRscan \cite{CDRscan} is the first  deep learning method for drug response prediction, which used genomics mutation as a cancer cell profile. tCNNs \cite{tcnns} adopt two convolution networks to learn the representations of drug and cancer cell lines,which they using  SMILES sequences of drugs and genomics mutation data of cell lines. MOLI \cite{MOLI} is a multi-omics late integration method based on deep neural networks, which takes somatic mutation, copy number aberration and gene expression data as input.

The graph-based methods focuses on converting drug molecules into graph structures and then using graph neural networks (GNNs) models for graph representation, e.g. GraphDRP  \cite{110bGCNforDRP}, DeepCDR \cite{DeepCDR}, GraTransDRP \cite{GraTransDRP}, etc. Considering that GNNs can extract  structural features of molecular, GraphDRP \cite{110bGCNforDRP} used GNNs to drug response prediction for the first time, which achieved superior performance compared to traditional machine learning methods.
Subsequently, \cite{18bMultipleOmic} proposed a deep learning model, GraOmicDRP, to learn the features of drug from graph representation and integrate multiple genomics data of cell lines, which is similar to MOLI \cite{MOLI}. DeepCDR  \cite{DeepCDR} is a hybrid graph convolutional network that integrated multi-omics profiles and explored intrinsic chemical structures of drugs for drug response prediction.
GraTransDRP \cite{GraTransDRP} is combined with GNNs and Transformer for predicting cancer drug response, which is primarily used to reduce the error associated with the approximation of heterogeneous graphs to homogeneous graphs. DeepTTA \cite{DeepTTC} decomposes drug molecules into fragments, and encodes drug molecules using fragment probabilities,  then uses Transformer methods for drug response prediction.


























\section{Method}

\subsection{Framework of TransEDRP }

\begin{figure*}[!ht]
    \centering
    \includegraphics[width=0.85\textwidth]{img/model2.jpg}
    \caption{The framework illustration for TransEDRP. The inputs of the model are the drug SMILES string and the genomics sequence of cell lines, respectively. Our model therefore consists of three components: an encoder module for the drug, an encoder module for the genomics, and a prediction module for the interactions between the drugs and the genomics.}
    \label{fig:model}
\end{figure*}


We proposed an end-to-end deep learning model to predict the anti-cancer drug response, named TransEDRP (see Figure \ref{fig:model}). Due to the specificity of  cancer drug response task, there are usually two main parts for feature extraction of anti-cancer drug and genomics of cell lines from the cancer patients, respectively. For the extensibility of the model, two separate branches based on the transformer are proposed for the two parts. Specifically,  we utilize the graph transformer with edge embedded module to extract the higher-order semantics of the drug molecule graph,  and utilize the multi-head attention module to extract information from genomics of cell line, then the two types of features are are fused to predict the  through the fully connected neural network.






















\subsection{Graph transformer with edge embedding}





We use the  mapping the edges features   into the  (see Table \ref{tab:bondE}).


\begin{table}[htbp]
\caption{The isomers caused by double bonds or single bonds of ring-forming carbon atoms that are not free to rotate are geometric isomers, and for this property of the molecule we obtain information on the five types of bond. In the 3D diagram of a molecule, we refer to a bond between two carbon atoms, except for the third example, which is a chemical bond between a carbon and an oxygen atom (oxygen atoms are indicated by red nodes).}
\renewcommand\arraystretch{1.2}
\centering
\resizebox{\columnwidth}{!}{\begin{tabular}{cccccc}
 \bottomrule 
Example & Type as Double & IsAromatic & IsConjugated & IsInRing & BondDir \\ \hline

\begin{minipage}[b]{0.2\columnwidth}
	\raisebox{-.5\height}
	{\includegraphics[width=\linewidth]{img/mol1.png}}
\end{minipage}    
	
& 1.0(SINGLE) & False & True & False & NONE \\

\begin{minipage}[b]{0.2\columnwidth}
	\raisebox{-.5\height}
	{\includegraphics[width=\linewidth]{img/mol2.png}}
\end{minipage}   

& 1.5(AROMATIC) & True & True & True & NONE \\

\begin{minipage}[b]{0.2\columnwidth}
	\raisebox{-.5\height}
	{\includegraphics[width=\linewidth]{img/mol3.png}}
\end{minipage}   

& 2.0(DOUBLE) & False & True & False & \begin{tabular}[c]{@{}c@{}}ENDDOWNRIGHT\\ / ENDUPRIGHT\end{tabular} \\

\begin{minipage}[b]{0.2\columnwidth}
	\raisebox{-.5\height}
	{\includegraphics[width=\linewidth]{img/mol4.png}}
\end{minipage}  

& 3.0(TRIPLE) & False & True & False & NONE \\  \bottomrule 
\end{tabular}}

\label{tab:bondE}
\end{table}


We initialize the graph for the drug molecule SMILES string to obtain the adjacency matrix    and the feature matrix of the atomic nodes ,  is the number of nodes in a graph, also known as the number of atoms in the drug molecule.

To extract the global information of the graph, we introduced the multi-head attention encoder layer to encode representation of each node. In this paper, the multi-head attention encoder layer is made up of several self-attention to extract the  information of the graph: 










\noindent where , and  is the index of the multi-head. The representation of the node  is obtained after multi-head attention extraction (we set the number of the head is 1).


Then we adopt a graph attention network (GAT) \cite{GAT} in our model for the first massage passing.
Take  as the node input to the GAT and representation the dependencies of the node neighbours by using the edges features  as the  weight-share  , where . The attention score between two node  and  can be calculate as fellows:






\noindent where  is the two-layer feed forward neural network. Then  we use softmax function normalizes the correlation coefficient of node .







\noindent where  is a neighborhood of node  in the graph.The aggregated features from each head are concatenated to obtain :




\noindent where  is the LeakyReLU non-linear activation functions,  is the index of the multi-heads and  is total number of multi-head,    means concatenation of vectors.

Denote the  as the node features output from GAT. Then we use Equation \ref{alg: attention} again to obtain the representation of node  .




The message aggregation of the node representation  of the graph is performed by GCN with the following equation:



\noindent where  is the linear transformation matrix,  is also a nonlinear activation function.

We denote the  as the node features output from GCN.





For graph pooling, we compute the global mean pooling and global max pooling of all nodes   respectively  and denote them as .




Finally, we use the fully-connected layers and dropout layer as fusion neural network  to combine two vectors and convert them to , and get representation of each drug molecule graph .

























\subsection{ Multi-head attention genomics extraction block}

We obtained gene mutation and pan-cancer gene from CCLE \cite{CCLE} and  encoded them into one-hot format to obtain   as an anti-cancer cell line for gene expression.
 The genomics of cell lines  is processed by the attention extraction module
, of which is made up of multiple attention-based layers. One layer is given as an example:





\noindent where  means layer normalization,  function is referenced to Equation \ref{alg: attention},  is the fully connected network. We add the result of attention to  and  as the residual connection, and then perform layer normalization.



\noindent where  is the fully connected network, . Then the genomics of cell lines can be expressed as . We map the genomics information into the 128-dimensional feature space consistent with the size of the feature space of the drug molecule for multimodal fusion.











\subsection{Multimodal fusion block for drug response prediction}






Finally, for the prediction of drug response,  ,  are  concatenated and put through two fully-connected layers   and ReLU activation function  to predict the response .




\noindent where  is the true label of  score. We use mean squared loss  as the loss function of our model, which are consistent with other methods.







\section{Experimental Setting}


\subsection{Dataset}


Drug sensitivity screening projects such as CCLE \cite{CCLE}
 and GDSC \cite{GDSC} can provide genomics information and anti-cancer drug response data.
Drug molecules can be downloaded from Pubchem in SMILES standard format.
We parse the SMILES of the drug through rdkit and DeepChem to obtain the features of all the atoms and bonds within the molecule. Each chemical bond includes two features: double bond stereo information, the type of bond.
Fellow GraphDRP \cite{110bGCNforDRP}, each atom includes five features: atom symbol, atom degree calculated by the number of bonded neighbors and Hydrogen, the total number of Hydrogen, implicit value of the atom,(possible formal charge list) and whether the atom is aromatic.
If there exists a bond among a pair of atoms, an edge is set.
The genomics data of cell lines include mutation and copy number aberration , which is converted to one-hot format with a vector of 735 dimensions.
The response values  score is normalized in a range (0,1).



As a result, we obtained  drugs,  cell line gene expressions, and  drug response data after coding the drug and genomics sequences. Each drug was transformed into an undirected graph G , where the dimensions of the atom feature in a drug molecule is 78. Each cell line is converted to one-hot format with a vector of 735 dimensions.










\begin{table*}[htbp]
\caption{The best performance is in bold, the second best experimental result is indicated  in italics}
\renewcommand\arraystretch{1.2}
\resizebox{\textwidth}{!}{\begin{tabular}{ccccccc}
\toprule[1pt]
\multicolumn{2}{c}{\multirow{2}{*}{Models}}          & \multicolumn{2}{c}{Pearson }         & \multicolumn{2}{c}{RMSE }            & Spearman                 \\ \cline{3-7} 
\multicolumn{2}{c}{}             & Paper show & Our test        & Paper show & Our test        & Our test        \\ \hline
\multicolumn{2}{c}{tCNNs}        & 0.91600    &  0.88307  0.00213             & 0.02840    & 0.03291  0.00089             & 0.85547  0.00073          \\
\multicolumn{1}{c|}{\multirow{4}{*}{GraphDRP}} & GCN & 0.92160   & 0.91794 0.00027          & 0.02590   & 0.02664 0.00006          & 0.89770 0.00017          \\
\multicolumn{1}{c|}{} & GIN      & 0.93100    & 0.93005 0.00060 & 0.02440    & 0.02477 0.00015 & 0.90842 0.00071 \\
\multicolumn{1}{c|}{} & GAT      & 0.92700    & 0.92646 0.00055 & 0.02500    & 0.02532 0.00012 & 0.90309 0.00045 \\
\multicolumn{1}{c|}{} & GCN\_GAT & 0.93080    & 0.93060 0.00016 & 0.02430    & 0.02452 0.00003 & 0.90944 0.00019 \\
\multicolumn{2}{c}{GraTransDRP}  & \textit{0.93170}    & \textit{0.93481}  \textit{0.00019} & \textit{0.02380}    & \textit{0.02377}  \textit{0.00003} & \textit{0.91364}  \textit{0.00047} \\
\multicolumn{2}{c}{\textbf{Ours}}    &  & \textbf{0.93807}   \textbf{0.00017} &  & \textbf{0.02319}   \textbf{0.00007} & \textbf{0.91796}   \textbf{0.00014} \\ \toprule[1pt]
\end{tabular}}





\label{table: comparative experiment}
\end{table*}

\subsection{Experimental Design}

Drug sensitivity data for cancer from GDSC database, we split it into 80\% as the training set, 10\% as the validation, and 10\% as the test set. The validation set is used to tune the hyperparameters, and the testing set is used to evaluate the model.  


In the training phase we set the batch size to 256 , the learning rate to 0.0005, the random seeds to 42. For our model, the number of attention heads for drug representation is 1, the number of GAT heads is 10, the number of  multi-head attention blocks for cell lines genomics is 3 in total, each with 15 attention heads. By using an Adam optimizer, all the experiments are performed five times and then averaged. In the comparative experiment we show the variance of the evaluation results from different methods as a stability reference. Our equipment information is: Intel(R) Xeon(R) Silver 4210R, GeForce RTX 2080 Ti.



\subsection{Performance Evaluation}

We use   Pearson's correlation coefficient (),  Spearman's correlation coefficient (), and root means square error () as metrics.
 
 RMSE measures the difference between the predicted value and the true value:




\noindent where  is the number of data ,  is a ground-truth of -th sample,  is the predicted value of -th sample. Pearson correlation coefficient measures how strong a relationship is between two variables.  is defined as:


\noindent where the standard deviation of  and  be   respectively.

The Spearman rank correlation coefficient is a non-parametric (distribution-independent) rank statistical parameter, which is usually considered as the Pearson linear correlation coefficient between the ranked variables. Then the Spearman rank correlation coefficient is:




 The lower  is, the better the model is. Meanwhile, the higher  and  are, the better the model is. Typically, model performance is evaluated by these three metrics.




\section{Results and Discussion}

\subsection{Sensitivity Analysis of the hyperparameter}




Mapping sequences in a lower dimensional feature space is easier to learn rich feature information without increasing the time complexity \cite{attention}. The length of the genomics sequence of the cell lines encoded is 735, and the multi-head mechanism could reduce the difficulty of feature extraction. Therefore, we investigated the trend of model performance as a result of variation in the number of attention blocks  and list of head number in each multi-head attention-based gene sequence feature extraction module .



\begin{table}[htbp]
\caption{We show the model evaluation metrics Pearson at different variations of n and h  h when n is determined from 1 to 5.}
\centering
\renewcommand\arraystretch{1.2}
\resizebox{\columnwidth}{!}{\begin{tabular}{cccccc}
\toprule[1pt]
\diagbox{h}{Pearson}{n} & 1 & 2 & 3 & 4 & 5 \\ \hline
1 & {\underline{0.93649} } & 0.93751 & 0.93738 & { \textbf{0.93794}} & 0.93763 \\
7 & {\underline{0.93642} } & { 0.93781} & 0.93747 & 0.937379 & \textbf{0.93796} \\
15 & {\underline{0.93624} } & 0.93797 & { \textbf{0.93810}} & 0.937252 & 0.93744 \\ \toprule[1pt]
\end{tabular}}

\label{tab:nh}
\end{table}


\begin{figure}[htbp]
	\centering
    \includegraphics[width=.95\linewidth]{img/infer/hpyer.png}\label{fig:3b}
	
	\caption{The line shows the trend in model performance with n when  is determined to be 1.
	}
	\label{fig:SAH}
\end{figure}

As shown in Table \ref{tab:nh}, in contrast to n greater than 1, the performance of the model drops significantly for n equal to 1. In Figure \ref{fig:SAH}, With  all 1,  the performance of the model increases slowly to decreases rapidly as  changes from 1 to 9.
The time complexity of multi-head attention is , and the single-layer CNN is , where  is the sequence length,  is the representation dimension,  is the kernel size of convolutions.

Furthermore, comparing CNNs, our attention with only 1 layer and the head is 1 also outperforms state-of-the-art, fully demonstrating the problem of receptive field in Figure \ref{fig:intro2}.










 
\subsection{Comparative Experiment}




For our comparative drug response prediction experiments, we selected five methods that are representative of current publicly available methods: GCN, GIN, GAT, concatenate of GCN and GAT under GraphDRP \cite{110bGCNforDRP} structure , GraTransDRP \cite{GraTransDRP}, tCNNs \cite{tcnns} with excellent performance among non-graph based methods.

The Comparative Experiments were done on the GDSC drug response datasets and measured in the ,  and  metrics. Since the experimental environment and data processing of the tCNNs method are different from our method, this paper shows the experimental results from \cite{110bGCNforDRP}.



Five randomized experiments on each model to obtain the mean and standard deviation as the model performance.
In the evaluation of Pearson, RMSE and Spearman metrics, ours model achieved the best results, improving over the current SOTA method GraTransDRP by 0.35\% (Pearson), 2.34\% (RMSE) and 0.47\% (Spearman) respectively. 

In addition the graph-based methods  are better than the CNN-based  method  tCNNs due to the introduction of molecular  structural information. 




\subsection{Ablation Experiments}





\begin{table}[htbp]
\caption{Ablation experiments for TransEDRP. }
\renewcommand\arraystretch{1.2}
\centering
\resizebox{\columnwidth}{!}{\begin{tabular}{cccccc}
\toprule[1pt]
\begin{tabular}[c]{@{}c@{}}Graph\\ Transformer\end{tabular} & \begin{tabular}[c]{@{}c@{}}Edge\\ Embedding\end{tabular} & \begin{tabular}[c]{@{}c@{}}Cell Line\\ Extract\end{tabular} & RMSE & Pearson & Spearman \\ \hline
\multicolumn{1}{c}{\multirow{4}{*}{No}} & No & Conv1d & 0.02461 & 0.93004 & 0.90965 \\
\multicolumn{1}{c}{} & Yes & Conv1d & 0.02442 & 0.93112 & 0.91047 \\
\multicolumn{1}{c}{} & No & Attention & 0.02539 & 0.92533 & 0.90370 \\
\multicolumn{1}{c}{} & Yes & Attention & 0.02523 & 0.92635 & 0.90548 \\ \hline
\multicolumn{1}{c}{\multirow{4}{*}{Yes}} & No & Conv1d & 0.02378 & 0.93480 & 0.91360 \\
\multicolumn{1}{c}{} & Yes & Conv1d & 0.02370 & 0.93530 & 0.91467 \\
\multicolumn{1}{c}{} & No & Attention & 0.02329 & 0.93774 & 0.91738 \\
\multicolumn{1}{c}{} & \textbf{Yes} & \textbf{Attention} & \textbf{0.02329} & \textbf{0.93808} & \textbf{0.91796} \\ \toprule[1pt]
\end{tabular}}

\label{tab:result}
\end{table}

Our models for the ablation experiments in drug molecular representation  and genomics representation methods are shown in Table \ref{tab:result}. 

In the ablation experiments, we had three points to verify: 


\begin{itemize}
\item The validity of using transformer to extract node features for the representation of molecular graphs;
\item The validity of using edge embedding in the graph representation;
\item The validity of using multi-head attention for the extraction of cell line genomics.
\end{itemize} 


By comparing the performance of graph transformer with edge embedding and multi-head attention in RMSE, Pearson and Spearman metrics, the effectiveness of the two parts of the method can be verified. The best experimental result is achieved by using Graph Transformer with edge embedding for the representation of drug and multi-head attention for the representation of cell line genomics.



In response to the third point, we found that the effectiveness of using the multi-head attention to extract cell lines genomics is only manifested when a graph transformer was used in the other branch. Although the structure of the dual transformer shows that the features of the two branches are not coupled before fusion, the results show that there is a clear interaction between them, which may imply that the extent of drug molecule representation is a determining factor in the performance of drug response prediction models.







\subsection{External validation}













\begin{figure}  \centering
\includegraphics[width=0.5\textwidth]{img/infer/show.jpg} 
\caption{The x-axis indicates the true value and the y-axis indicates the predicted results, the sample size is 19,105. }
\label{fig:showtest}
\end{figure}
The visualisation is shown in Figure \ref{fig:showtest}. The red line represents the linear fit to the data we predicted.
The red dashed line indicates that the prediction is the same as the true value in the ideal case.
We visualise the distribution of true and predicted values through scatter plots and our model shows good predictive performance for a large number of random samples.














\section{Summary and Future Work}

\subsection{Summary}
In this study, we propose a method for drug response prediction called TransEDRP. Our model used graph transformer with edge embedding network to extract a better graph representation of drug and used the multi-head attention to extract  genomics representation of cell line, which is a dual-branch attention architecture.

Our approach was compared with tCNNs \cite{tcnns}, GraphDRP \cite{110bGCNforDRP}, GraphTransDRP \cite{GraTransDRP}, etc. The reproducible experimental results showed that TransEDRP achieved SOTA in all experiments. In this study, we focused on the representation of the graph structure data of the drug and the representation of the cell lines genomics separately to improve the prediction accuracy of , and we conducted ablation experiments to demonstrate the validity of the individual improvements of our model.



\subsection{Future Work}

In the area of drug molecule representation, future experiments could be made with 3d drug molecule modelling and more robust and explainable  analyses based on drug effect groups, for example.
Currently, it is more difficult to compare the performance of different models in the same situation due to the limitations of the data set and the different data selection, feature design and processing of drug molecules and  genomics of cell line. by different drug response prediction methods. Perhaps, in the future, the performance of different models should be tested in the same data set  split case (an unbiased test set need to be published).



















































































































































































































\section*{Ethical Statement}

There are no ethical issues.

\section*{Acknowledgments}




\bibliographystyle{named}
\bibliography{ijcai22}

\end{document}
