\documentclass{cai}



\def\bs#1{{\tt\char`\\#1}}


\usepackage{graphics}
\usepackage{epstopdf}
\usepackage{epsfig}




\usepackage{color}
\usepackage{epic,eepic}

\usepackage{amsfonts}
\usepackage{latexsym}
\usepackage{amssymb}
\usepackage{amsmath}

\usepackage{float}
\usepackage{xspace}

\usepackage{hyperref}

\usepackage{graphics}
\usepackage{epstopdf}
\usepackage{epsfig}

\def\NULL{\textrm{\tt null}}

\def\argmax{\ensuremath{{\textrm{argmax}}}}

\newcommand{\umod}{{\bf~\%~}\xspace}
\newcommand{\uassign}{\leftarrow}
\newcommand{\uif}{{\bf if}\xspace}
\newcommand{\uthen}{{\bf then}\xspace}
\newcommand{\uelse}{{\bf else}\xspace}
\newcommand{\uand}{{\bf~\&~}\xspace}
\newcommand{\uor}{{\bf~|~}\xspace}
\newcommand{\uxor}{{\bf~^\wedge~}\xspace}
\newcommand{\unot}{{\bf\sim}\xspace}
\newcommand{\uwhile}{{\bf while}\xspace}
\newcommand{\urepeat}{{\bf repeat}\xspace}
\newcommand{\uuntil}{{\bf until}\xspace}
\newcommand{\ufor}{{\bf for}\xspace}
\newcommand{\uto}{{\bf to}\xspace}
\newcommand{\uforeach}{{\bf for~each}\xspace}
\newcommand{\udo}{{\bf do}\xspace}
\newcommand{\ureturn}{{\bf return}\xspace}
\newcommand{\ubreak}{{\bf break}\xspace}
\newcommand{\ucontinue}{{\bf continue}\xspace}
\newcommand{\urem}[1]{{\put(250,0){\sc\footnotesize// #1}}\noindent}
\newcounter{lineno}
\newcommand{\uln}{\stepcounter{lineno}\>\hspace{-5mm}\thelineno}
\newcommand{\utab}{\qquad}

\newcommand{\startindent}{\hspace{0.8em}}

\newenvironment{code}{\setcounter{lineno}{0}\begin{tabbing}
\utab\=\utab\=\utab\=\utab\=\utab\=\utab\=\utab\=\utab\=\utab\=\utab\=\utab\=\utab\=\utab\= \kill
}
{
\end{tabbing}\vspace{-2mm}
}

\floatstyle{ruled}
\newfloat{algorithm}{tbp}{flt}
\floatname{algorithm}{\small\bf Alg.}

\newcommand{\ceil}[1]{{\lceil #1 \rceil}}
\newcommand{\floor}[1]{{\lfloor #1 \rfloor}}
\newcommand{\abs}[1]{{\left | #1 \right |}}

\newcommand{\exclude}[1]{}


\newcommand{\avg}[1]{{\textrm{avg}(#1)}}

\def\qed{\hfill \ensuremath \Box}

\def\mmod{~\textrm{mod}~}

\sloppy



\begin{document}
\label{firstpage}

\title[Two simple full-text indexes based on the suffix array]
      {Two Simple Full-Text Indexes\\ based on the Suffix Array}

\author[Sz.~Grabowski, M.~Raniszewski]
       {Szymon \surname{Grabowski}, Marcin \surname{Raniszewski}}

\affiliation{Institute of Applied Computer Science\\
Lodz University of Technology\\
Al.\ Politechniki 11\\
90--924 {\L}\'od\'z, Poland}

\email{sgrabow@kis.p.lodz.pl, mranisz@kis.p.lodz.pl}



\noreceived{} \nocommunicated{}

\maketitle

\begin{abstract}
We propose two suffix array inspired full-text indexes.
One, called SA-hash, augments the suffix array with a hash table 
to speed up pattern searches due to significantly narrowed search 
interval before the binary search phase.
The other, called FBCSA, is a compact data structure, similar to
M{\"a}kinen's compact suffix array, but working on fixed sized blocks.
Experiments 
on the Pizza~\&~Chili 200\,MB datasets 
show that SA-hash is about 2--3 times faster in pattern searches (counts) 
than the standard suffix array, for the price of requiring  
bytes of extra space, where  is the text length, and setting a 
minimum pattern length.
FBCSA is relatively fast in single cell accesses (a few times faster 
than related indexes at about the same or better compression), 
but not competitive if many consecutive cells are to be extracted. 
Still, for the task of extracting, e.g., 10 successive cells its time-space 
relation remains attractive.
\end{abstract}

\begin{keywords}
Suffix array, compressed indexes, compact indexes, hashing
\end{keywords}



\section{Introduction}

The field of text-oriented data structures continues to bloom.
Curiously, in many cases several years after ingenious theoretical solutions
their more practical (which means: faster and/or simpler) counterparts are presented, 
to mention only recent advances in rank/select implementations~\cite{GP14} 
or the FM-index reaching the compression ratio bounded by -th order
entropy with very simple means~\cite{DBLP:conf/spire/KarkkainenP11}.

Despite the great interest in compact or compressed\footnote{By the latter we mean indexes with space use bounded by  or even 
 bits,  where  is the text length,
and  and  respectively the order-0 and the order- entropy.
The former term, compact full-text indexes, is less definite, 
and, roughly speaking, may fit any structure with less than  
bits of space, at least for ``typical'' texts.}
full-text indexes in recent years~\cite{NMacmcs06}, we believe that in 
some applications search speed is more important than memory savings, thus 
different space-time tradeoffs are worth being explored.
The classic suffix array (SA)~\cite{MM90}, combining speed, simplicity and often 
reasonable memory use, may be a good starting point for such research.

In this paper we present two SA-based full-text indexes, combining effectiveness 
and simplicity.
One augments the standard SA with a hash table to speed up searches, 
for a moderate overhead in the memory use, 
the other is a byte-aligned variant of M{\"a}kinen's compact 
suffix array~\cite{DBLP:conf/cpm/Makinen00,DBLP:journals/fuin/Makinen03}.

A preliminary version of this article appeared in Proc. PSC 2014~\cite{GR14}.

\section{Preliminaries}
\label{sec:prelim}

We use 0-based sequence notation, that is, a sequence  of length  
is written as , or equivalently as .

One may define a {\em full-text index} over text  of length  
as a data structure supporting at least two basic types of queries, 
both with respect to a pattern  of length , 
where  and  share an integer alphabet of size .
One query type is {\em count}: return the number  of occurrences 
of  in .
The other query type is {\em locate}: for each pattern occurrence 
report its position in , that is, such  that .

The {\em suffix array}  for text  is a permutation 
of the indexes  such that 
 for all , 
where the ``'' relation is the lexicographical order.
The inverse suffix array  is the inverse permutation of , 
that is, .

If not stated otherwise, all logarithms throughout the paper are in base 2.


\section{Related work}
\label{sec:rwork}

The full-text indexing history starts with the {\em suffix tree} 
(ST)~\cite{Wei73}, a trie whose string collection is the set of all the 
suffixes of a given text, with an additional requirement that all 
non-branching paths of edges are converted into single edges. 

Typically, each ST path is truncated as soon as it points to a unique suffix. 
A leaf in the suffix tree holds a pointer to the text location where the 
corresponding suffix starts.
As there are  leaves, no non-branching nodes and edge labels represented 
with pointers to the text, the suffix tree takes  words of space, 
which is in turn  bits.

Although initially the suffix tree was known to be constructible in 
linear time only for constant alphabets, 
later an ingenious -time algorithm for integer 
alphabets was found~\cite{Far97}.
Moreover, the linear-time construction algorithms can 
be fast not only in theory, but also in practice~\cite{Gri07}.
Assuming constant-time access to any child of a given node, 
the search in the ST takes only  time in the worst case.
In practice, this is cumbersome for a large alphabet, 
of size , 
as it requires using 
perfect hashing, which also makes the construction time linear only in expectation.
A small alphabet is easier to handle, which goes in line with the wide 
use of the suffix tree in bioinformatics.

The main problem with the suffix tree 
is its large space requirement.
Even in the most economical version~\cite{KB00} 
the ST space use 
reaches almost  bytes on average and  in the worst case, 
plus the text, for , and even more for large alphabets.
Most implementations need  bytes or more.

An important alternative to the suffix tree is the {\em suffix array} (SA)~\cite{MM90}.
It is an array of  pointers to all text suffixes 
sorted according to the lexicographic order of these suffixes.
The SA needs  bits for its  suffix pointers (indexes), 
plus  bits for the text, which typically translates to  
bytes in total.
The pattern search time is  in the worst case 
and  on average,
which can be
improved to  in the worst case 
using the longest common prefix (lcp) table.
Alternatively, the  time can be reached even without the lcp,
in a more theoretical solution with a specific suffix permutation~\cite{DBLP:journals/talg/FranceschiniG08}.
Yet Manber and Myers in their seminal paper~\cite{MM90} presented a nice 
trick saving several first steps in the binary search:
if we know the SA intervals for all the possible first  symbols of the
pattern, we can immediately start the binary search in a corresponding interval.
We can set  close to , with  extra bits of
space, but constant expected size of the interval, which leads to  average
search time and only 
 
cache misses on average, 
where 

is the cache line length expressed in symbols,  
typically 64 symbols/bytes in a modern CPU.
Unfortunately, real texts are far from random, hence in practice, 
if text symbols are bytes, 
we can use  up to 3, which offers a limited (yet, non-negligible)
benefit.
This idea, later denoted as using a lookup table (LUT), is fairly well known, 
see e.g. its impact in the search over a suffix array on words~\cite{DBLP:conf/cpm/FerraginaF07}.

The suffix array can be built from the suffix tree by visiting its leaves 
in order (hence preserving  construction time), yet this approach is 
impractical.
Only in 2003 several algorithms building the SA directly in linear time 
were presented, e.g.,~\cite{KS03}, and currently the fastest -time 
construction algorithm is the one given by Nong~\cite{NZC11}.

A number of suffix tree or suffix array inspired indexes have been proposed 
as well, including the suffix cactus~\cite{karkkainen1995suffix} and the 
enhanced suffix array (ESA)~\cite{abouelhoda2002enhanced}, with space use usually 
between SA and ST, 
but according to our knowledge they generally are not faster than their 
famous predecessors in the count or locate queries.

On a theoretical front, the suffix tray by Cole et al.~\cite{cole2006suffix} 
allows to achieve  search time, with  worst-case 
time construction and  bits of space, which was recently improved 
by Fischer and Gawrychowski~\cite{FG15} 
to  deterministic time, with preserved construction 
cost complexities.

The common wisdom about the practical performance of ST and SA 
is that they are comparable, but Grimsmo in his interesting experimental 
work~\cite{Gri07} showed that a careful ST implementation 
may be up to about 50\% faster than SA if the number of matches is very small 
(in particular, one hit), but if the number of hits grows, the SA becomes 
more competitive, sometimes being even about an order of magnitude faster.
Another conclusion from Grimsmo's experiments is that the ESA may also be 
moderately faster than SA if the alphabet is small (say, up to 8 symbols) 
but SA easily wins for a large alphabet.

Since around 2000 we can witness a great interest in succinct data structures, 
in particular, text indexes.
Two main ideas that deserve being mentioned are 
the compressed suffix array (CSA)~\cite{GV00,DBLP:conf/soda/Sadakane02} 
and the FM-index~\cite{FM00}; the reader is referred to the survey~\cite{NMacmcs06} 
for an 
thorough
coverage of the area.

It was noticed in extensive experimental comparisons~\cite{FGNVjea08,GP14} 
that compressed indexes are not much slower, and sometimes comparable, 
to the suffix array in count queries, but locate is 2--3 orders of magnitude 
slower if the number of matches is large.
This instigated researchers to follow one of two paths in order to 
mitigate the locate cost for succinct indexes.
One, pioneered by M{\"a}kinen~\cite{DBLP:conf/cpm/Makinen00,DBLP:journals/fuin/Makinen03} 
and addressed in a different way by Gonz{\'a}lez et al.~\cite{DBLP:conf/cpm/GonzalezN07,GNFjea14},
exploits repetitions in the suffix array (the idea is explained in Section~\ref{sec:fbsa}).
The other approach is to build semi-external data structures 
(see~\cite{GM13,GMCTW14} and references therein).


\section{Suffix array with deep buckets}

The mentioned idea of Manber and Myers with precomputed interval (bucket) 
boundaries for  starting symbols tends to bring more gain with growing , 
but also precomputing costs grow exponentially. 
Obviously,  integers are needed to be kept in the lookup table.
Our proposal is to apply hashing on relatively long strings, with an extra trick 
to reduce the number of 
unnecessary references to the text.

\begin{figure}
\begin{small}
\begin{code}
HT\_build(, , , , ) \\
Precondition:  \\
\rule{\textwidth}{0.3mm} \\
\vspace{-20mm}(01) \startindent allocate  \\
(02) \startindent \ufor  \uto  \udo  \\
(03) \startindent  \\
(04) \startindent  \\
(05) \startindent ;  \\
(06) \startindent \ufor  \uto  \udo \\
(07) \startindent \>\>\uif  \uthen \ \ucontinue \\
(08) \startindent \>\>\uif  \uthen \\
(09) \startindent \>\>\>\uif  \uthen \\
(10) \startindent \>\>\>\> \\
(11) \startindent \>\>\>\> \\
(12) \startindent \>\>\> \\
(13) \startindent \>\>\> \\
(14) \startindent \>\>\>\urepeat \\
(15) \startindent \>\>\>\>\uif  \uthen \\
(16) \startindent \>\>\>\>\> \\
(17) \startindent \>\>\>\>\>\ubreak \\
(18) \startindent \>\>\>\>\uelse  \\
(19) \startindent \>\>\>\uuntil false \\
(20) \startindent  \\
(21) \startindent \ureturn  \\
\end{code}
\caption{Building the hash table of a given size }
\label{fig:HT_build}
\end{small}
\end{figure}


We start with building the hash table HT (Fig.~\ref{fig:HT_build}).
The hash function is calculated for the {\em distinct} -symbol () prefixes of 
suffixes from the (previously built) suffix array.
That is, we process the suffixes in their SA order and if the current suffix 
shares its -long prefix with its predecessor, it is skipped (line~08).
The value written to HT (line~11) is a pair: 
(the position in the SA of the first suffix with the 
given prefix, the position in the SA of the last suffix with the given prefix).
Linear probing is used as the collision resolution method.
As for the hash function, we used 
xxhash (\url{https://code.google.com/p/xxhash/}). 
We tested also a few alternatives: MurmurHash 
(\url{http://en.wikipedia.org/wiki/MurmurHash}) is practically as good  
as xxhash, 
CRC (\url{http://rosettacode.org/wiki/CRC-32})
is slightly slower overall (with up to about 3\% slower searches), 
while the loss of sdbm (\url{http://www.cse.yorku.ca/~oz/hash.html}) 
is greater, often exceeding 10\%.

Fig.~\ref{fig:Pattern_search} presents the pattern search (locate) procedure.
It is assumed that the pattern length  is not less than .
First the range of rows in the suffix array corresponding to the first two 
symbols of the pattern is found in a ``standard'' lookup table (line~1); 
an empty range immediately terminates the search with no matches returned (line~2).
Then, the hash function over the pattern prefix is calculated and a scan over the 
hash table performed until no extra collisions (line~5; return no matches) 
or found a match over the pattern prefix, which give us information about the range 
of suffixes starting with the current prefix (line~6).
In this case, the binary search strategy is applied to narrow down the SA interval 
to contain exactly the suffixes starting with the whole pattern.
(As an implementation note: the binary search could be modified to ignore the first  
symbols in the comparisons, but it did not help in our experiments, 
due to specifics of the used A\_strcmp function from the asmlib library\footnote{http://www.agner.org/optimize/asmlib.zip, v2.34, by Agner Fog.}).


\begin{figure}
\begin{small}
\begin{code}
Pattern\_search(, , , , , \\
\>\>\>) \\
Precondition:  \\
\rule{\textwidth}{0.3mm} \\
(1) \startindent  \\
(2) \startindent if  \uthen report no matches; \ureturn \\
(3) \startindent  \\
(4) \startindent \urepeat \\
(5) \startindent \>\>\uif  \uthen report no matches; \ureturn \\
(6) \startindent \>\>\uif () \textbf{and} \\
\>\>\>() \\
(7) \startindent \>\>\>\uthen binSearch(); \ureturn \\
(8) \startindent \>\> \\
(9) \startindent \uuntil false \\
\end{code}
\caption{Pattern search with SA-hash}
\label{fig:Pattern_search}
\end{small}
\end{figure}


\subsection{Reducing the memory for the hash table}

Each slot in the hash table (HT) contains two 32-bit integers, 
for the start and the end position of the range of suffixes starting 
with the corresponding prefix of length .
Yet, it is possible in practice to reduce the second value to 16 bits.
To this end, we make use of a lookup table over 
pairs of symbols (LUT2)
to initially narrow down the interval related to which the range in the HT 
will be encoded.
Then the actual range will be written approximately, with quantized
right boundary of the range.

For clarity, let us denote the new hash table with . 
The code for building  
is shown in Fig.~\ref{fig:HT_approx_build}.

Let us explain why a similar saving cannot be applied also to the 
start position of the range.
This is because a collision which (unluckily) points to a subrange of 
the actual HT range that we are looking for could not be detected.
Here is an example.
Let us assume that we have two -long prefixes: 
``somethin'' and ``once in'', which have the same hash value (collision).
The SA range for ``something'' is  and 
LUT2 stores (for ``so'') the range .
The SA range for ``once in'' is  and 
LUT2 table stores (for ``on'') the range .
Now we are decoding the range of ``somethin'' suffixes and 
there is a collision with ``once in''. 
Hence we obtained the SA range  which is a subrange of 
 and we cannot detect a collision. 
We are searching in narrower range, so the results may be wrong.
Quantizing only the right boundary of the range does not imply 
a similar problem.


\begin{figure}
\begin{small}
\begin{code}
HT\_approx\_build(, , , , ) \\
Precondition:  \\
\rule{\textwidth}{0.3mm} \\
\vspace{-20mm}(01) \startindent allocate  \\
(02) \startindent \ufor  \uto  \udo  \\
(03) \startindent  \\
(04) \startindent  \\
(05) \startindent ; ; ;  \\
(06) \startindent \ufor  \uto  \udo \\
(07) \startindent \>\>\uif  \uthen \ \ucontinue \\
(08) \startindent \>\>\uif  \uthen \\
(09) \startindent \>\>\>\uif  \uthen \\
(10) \startindent \>\>\>\> \\
(11) \startindent \>\>\>\> \\
(12) \startindent \>\>\> \\
(13) \startindent \>\>\> \\
(14) \startindent \>\>\>\urepeat \\
(15) \startindent \>\>\>\>\uif  \uthen \\
(16) \startindent \>\>\>\>\> \\
(17) \startindent \>\>\>\>\> \\
(18) \startindent \>\>\>\>\> \\
(19) \startindent \>\>\>\>\>\ubreak \\
(20) \startindent \>\>\>\>\uelse  \\
(21) \startindent \>\>\>\uuntil false \\
(22) \startindent  \\
(23) \startindent \ureturn  \\
\end{code}
\caption{Building the hash table with reduced memory}
\label{fig:HT_approx_build}
\end{small}
\end{figure}



\section{Fixed Block based Compact Suffix Array} \label{sec:fbsa}

M{\"a}kinen's compact suffix
array~\cite{DBLP:conf/cpm/Makinen00,DBLP:journals/fuin/Makinen03} 
finds and succinctly represents repeating suffix areas.
We propose a variant of this index
whose key feature is 
finding {\em approximate} repetitions of suffix areas of predefined size.
Choosing the fixed area size 
allows to maintain a byte-aligned data layout, 
beneficial for speed and simplicity.
Even more, by setting a natural restriction on one of the key parameters 
we force the structure's building bricks to be multiples of 32 bits, 
which prevents misaligned access to data.

M{\"a}kinen's index was the first {\em opportunistic} scheme for 
compressing a suffix array, that is such that uses less space on compressible 
texts.
The key idea was to exploit runs in the SA, that is, maximal 
segments  for which there exists another 
segment , such that  
for all .
This structure still allows for binary search, only the accesses to SA cells 
require local decompression.
In our approach we take suffix areas of fixed size, e.g., 32 bytes: 
, and find for them other suffix array segments 
 such that for each  there exists  such that . 
Moreover, the sequence of chosen values of  is ascending. 


\begin{figure}
\begin{small}
\begin{code}
FBCSA\_build(, , , ) \\
\rule{\textwidth}{0.3mm} \\
/* assume  is a multiple of  */ \\
(01) \startindent ;  \\
(02) \startindent  \\
(03) \startindent \urepeat \\
\>\>/* current block of the suffix array is  */ \\
(04) \startindent \>\>find 3 most frequent symbols in  \\
\>\>\> and store them in  \\ 
\>\>\> /* if there are less than 3 distinct symbols in , \\ 
\>\>\>\> the trailing cells of  are set to ) */ \\
(05) \startindent \>\>\ufor  \uto  \udo \\
(06) \startindent \>\>\>\uif  \uthen .append(00) \\
(07) \startindent \>\>\>\uelse \ \uif  \uthen .append(01) \\
(08) \startindent \>\>\>\>\uelse \ \uif  \uthen .append(10) \\
(09) \startindent \>\>\>\>\>\uelse .append(11) \\
(10) \startindent \>\>.pos() \\
(11) \startindent \>\>.pos() /* set  if  */ \\
(12) \startindent \>\>.pos() /* set  if  */ \\
(13) \startindent \>\> \\
(14) \startindent \>\>.append() \\
(15) \startindent \>\>.append() /* append  if  */ \\
(16) \startindent \>\>.append() /* append  if  */ \\
(17) \startindent \>\>\ufor  \uto  \udo \\
(18) \startindent \>\>\>\uif () \textbf{or} () \\ 
(19) \startindent \>\>\>\>\uthen .append(1); .append() \\
(20) \startindent \>\>\>\uelse .append(0) \\
(21) \startindent \>\>.append() \\
(22) \startindent \>\> \\
(23) \startindent \>\>\uif  \uthen \ \ubreak \\
(24) \startindent \uuntil false \\
(25) \startindent \ureturn  \\
\end{code}
\caption{Building the fixed block based compact suffix array (FBCSA)}
\label{fig:FBSA_build}
\end{small}
\end{figure}


The construction algorithm for our structure, called {\em fixed block based 
compact suffix array} (FBCSA), is presented in Fig.~\ref{fig:FBSA_build}.
As a result, we obtain two arrays,  and , which are empty 
at the beginning, and their elements are always appended at the end 
during the construction.
The elements appended to  are single bits or pairs of bits 
while  stores suffix array indexes (32-bit integers).

The construction makes use of the suffix array  of text ,
the inverse suffix array  and  (which can be obtained 
from  and , that is, ).

Additionally, there are two construction-time parameters: 
block size  and sampling step .
The block size tells how many successive  indexes are encoded together 
and is assumed to be a multiple of 32, for int32 alignment of the structure 
layout.
The parameter  means that every -th  index will be 
represented verbatim.
This sampling parameter is a time-space tradeoff; 
using larger  reduces the overall space but decoding a particular SA index 
typically involves more recursive invocations.

Let us describe the encoding procedure for one block,
, where  is a multiple of .

First we find the three most frequent symbols in  and store them 
(in arbitrary order) in a small helper array  (line~04).
If the current block of  does not contain three different symbols, 
the  value will be written in the last one or two cell(s) of .
Then we write information about the symbols from  in the current block of 
into : we append 2-bit combination (00, 01 or 10) if a given symbol is from  
and the remaining combination (11) otherwise (lines~05--09).
We also store the positions of the first occurrences of the symbols from  
in the current block of , using the variables , ,  
(lines~10--12); 
again  values are used if needed.
These positions allow to use links to runs of suffixes 
preceding subsets of the current ones marked by the respective symbols from .

We believe that a small example will be useful here.
Let  and the current block be  (note this is a toy example 
and in the real implementation  must be a multiple of 32).
The  block contains the indexes: 1000, 522, 801, 303, 906, 477, 52, 610.
Let their preceding symbols (from ) be: , , , , , , , .
The three most frequent symbols, written to , are thus: , , .
The first occurrences of these symbols are at positions: 
,  and . 
We conclude that SA has the following groups of suffix offsets: 
 (as there are three symbols  in the current block of ), 
 and  and they start at positions: 
,  and .


We come back to the pseudocode.
The described (up to three) links are obtained thanks to 
 (lines~14--16) and are written to .
Finally, the offsets of the suffixes preceded with a symbol not from  
(if any) have to be written to  explicitly.
Additionally, the sampled suffixes (i.e., those whose offset modulo  is 0) 
are handled in the same way (line~18).
To distinguish between referrentially encoded and explicitly written 
suffix offsets, we spent a bit per suffix and append them to  
(lines~19--20).
To allow for easy synchronization between the portions of data in  
and , the size of  (in bytes) as it was before processing the 
current block is written to  (line~21).


\begin{figure}
\begin{code}
Find(, , , ) \\
\rule{\textwidth}{0.3mm} \\
/* assume  is a multiple of 32 */ \\
(01) \startindent  \\
(02) \startindent  \\
(03) \startindent  \\
(04) \startindent  \\
(05) \startindent  \\
(06) \startindent  \\
(07) \startindent  \\
(08) \startindent \uif  \uthen \\
(09) \startindent \>\>\ureturn  \\
(10) \startindent \uelse \\
(11) \startindent \>\> \\
(12) \startindent \>\> \\
(13) \startindent \>\> \\
(14) \startindent \>\>\ureturn Find(, , , \\
  \hspace{5.5em}  \\
\end{code}
\caption{Find() extracts  from the FBCSA structure}
\label{fig:FBSA_access}
\end{figure}


Fig.~\ref{fig:FBSA_access} presents the function Find(), 
which returns .
The helper arrays  and  contain respectively bits and 
pairs of bits (extracted from one or several integers).
The function  (popcount) returns the number of occurrences 
of symbol (integer)  in the given array of symbols (integers).
In modern CPUs  for a bit-vector of size e.g. 64 is usually 
available as a single op-code.


\section{Experimental results}

All experiments were run on a machine equipped with a 6-core Intel i7 CPU
(4930K) clocked at 3.4\,GHz, with 64\,GB of RAM, 
running Ubuntu 14.04 LTS 64-bit.
The RAM modules were \,GB DDR3-1600 with the timings 11-11-11 
(Kingston KVR16R11D4K4/64).
The CPU cache sizes were:
\,KB (data) and \,KB (instructions) in the L1 level,
\,KB in L2 and 12\,MB in L3.
One CPU core was used for the computations.
All codes were written in C++ and compiled with 64-bit gcc 4.8.2, 
with \texttt{-O3} option
(and for the FBCSA search algorithms with the additional \texttt{-mpopcnt} option).
The source codes for the FBCSA algorithm can be downloaded from \url{http://ranisz.iis.p.lodz.pl/indexes/fbcsa/}.

The test datasets were taken from the popular Pizza~\&~Chili 
site (\url{http://pizzachili.dcc.uchile.cl/}).
For most experiments we used the 200-megabyte versions of the files \texttt{dna},
\texttt{english}, \texttt{proteins}, \texttt{sources} and \texttt{xml}.
Only to compare search times of FBCSA variants against M{\"a}kinen's CSA
we used 50-megabyte datasets, due to text size limitations of the 
CSA implementation.

In order to test the search algorithms, we generated 500 thousand patterns 
for each used pattern length; the patterns were extracted randomly from the 
corresponding datasets (i.e., each pattern returns at least one match).

In the first experiment we compared pattern search (count) speed 
using the following indexes:
\begin{itemize}
\item plain suffix array (SA),
\item suffix array with a lookup table over the first 2 symbols (SA-LUT2),
\item suffix array with a lookup table over the first 3 symbols (SA-LUT3),
\item the proposed suffix array with deep buckets, 
      with hashing the prefixes of length  
      (only for \texttt{dna}  and for \texttt{proteins}  is used); 
      the load factor  in the hash table was set to 90\% (SA-hash),
\item a more compact variant of SA-hash, with 6 bytes rather than 8 bytes per 
      entry in the hash table (SA-hash-dense),
\item the proposed fixed block based compact suffix array with parameters  
      and  (FBCSA),
\item FBCSA (parameters as before) with a lookup table over the first 2 symbols 
      (FBCSA-LUT2),
\item FBCSA (parameters as before) with a lookup table over the first 3 symbols 
      (FBCSA-LUT3),
\item FBCSA (parameters as before) with a hash of prefixes of length  
      (only for \texttt{dna}  and for \texttt{proteins}  is used);
      the load factor in the hash table was set to 90\% (FBCSA-hash),
\item a more compact variant of FBCSA-hash, with 6 bytes rather than 8 bytes per 
      entry in the hash table (FBCSA-hash-dense).
\end{itemize}


\begin{figure}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{dna200_sa.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{english200_sa.eps}
}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{proteins200_sa.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{sources200_sa.eps}
}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{xml200_sa.eps}
}
\caption[Results]
{Pattern search time (count query). 
All times are averages over 500K random patterns of the same length 
, where  is 8 for most datasets 
except for \texttt{dna} (12) and \texttt{proteins} (5). 
The patterns were extracted from the respective texts.}
\label{fig:times1a}
\end{figure}


\begin{figure}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{dna200_fbcsa.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{english200_fbcsa.eps}
}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{proteins200_fbcsa.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{sources200_fbcsa.eps}
}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{xml200_fbcsa.eps}
}
\caption[Results]
{Pattern search time (count query). 
All times are averages over 500K random patterns of the same length 
, where  is 8 for most datasets 
except for \texttt{dna} (12) and \texttt{proteins} (5). 
The patterns were extracted from the respective texts.}
\label{fig:times1b}
\end{figure}


The results are presented in Fig.~\ref{fig:times1a} (faster indexes) 
and Fig.~\ref{fig:times1b} (FBCSA variants).
As expected, SA-hash is the fastest index among the tested ones.
The reader may also look at Table~\ref{table:speedups} with a rundown 
of the achieved speedups, where the plain suffix array is 
the baseline index and its speed is denoted with 1.00.


\begin{table}
\centering
\begin{tabular}{lrrrrr}
\hline
    &~~~~~~~~\texttt{dna}~&~~\texttt{english}~&~\texttt{proteins}~&~~\texttt{sources}~&~~~~~~\texttt{xml}~\\
\hline
 & & & & & \\
\hline
SA    & 1.00~~& 1.00~~& 1.00~~& 1.00~~& 1.00~~\\
SA-LUT2 & 1.21~~& 1.38~~& 1.37~~& 1.44~~& 1.39~~\\
SA-LUT3 & 1.28~~& 1.51~~& 1.59~~& 1.62~~& 1.52~~\\
SA-hash & 3.26~~& 2.79~~& 2.74~~& 2.76~~& 2.14~~\\
SA-hash-dense & 2.63~~& 2.46~~& 2.44~~& 2.47~~& 1.95~~\\
\hline
 & & & & & \\
\hline
SA    & 1.00~~& 1.00~~& 1.00~~& 1.00~~& 1.00~~\\
SA-LUT2 & 1.21~~& 1.37~~& 1.37~~& 1.42~~& 1.37~~\\
SA-LUT3 & 1.28~~& 1.49~~& 1.56~~& 1.60~~& 1.46~~\\
SA-hash & 3.36~~& 2.78~~& 2.74~~& 2.77~~& 1.81~~\\
SA-hash-dense & 2.64~~& 2.45~~& 2.44~~& 2.47~~& 1.69~~\\


\hline
\end{tabular}
\vspace{4mm}
\caption{Speedups with regard to the search speed of the plain suffix array, 
for the five 200\,MB datasets and pattern lengths  and }
\label{table:speedups}
\end{table}
 

The SA-hash index has two drawbacks: it requires significantly more space 
than the standard SA and we assume (at construction time) a minimal 
pattern length .
The latter issue may be eliminated, but for the price of even more space use; 
namely, we can build one hash table for each pattern length from 1 to  
(counting queries for those short patterns do not ever need to perform 
binary search over the suffix array).
For the shortest lengths ( or ) lookup tables may be 
alternatively used.

We have not implemented this ``all-HT'' variant, but it is easy to estimate 
the memory use for each dataset.
To this end, one needs to know the number of distinct -grams for 
 (Table~\ref{table:qgrams}).
Note that the alphabet size, i.e., the number of 1-grams,
for the DNA and proteins datasets is 16 and 25, respectively.
These surprisingly large values are explained by the content of the files 
in the corpus, ``polluted'' slightly with textual headers, End-of-Line symbols, etc.


\begin{table}
\centering
\begin{tabular}{lrrrrr}
\hline
~~   &~~~~~~~\texttt{dna}~~&~~~~\texttt{english}~~&~~\texttt{proteins}~~&~~~\texttt{sources}~~&~~~~~~~~~~\texttt{xml}~~\\
\hline
~~1 & 16~~~& 225~~~& 25~~~& 230~~~& 96~~~\\
~~2 & 152~~~& 10,829~~~& 607~~~& 9,525~~~& 7,054~~~\\
~~3 & 683~~~& 102,666~~~& 11,607~~~& 253,831~~~& 141,783~~~\\
~~4 & 2,222~~~& 589,230~~~& 224,132~~~& 1,719,387~~~& 908,131~~~\\
~~5 & 5,892~~~& 2,150,525~~~& 3,623,281~~~& 5,252,826~~~& 2,716,438~~~\\
~~6 & 12,804~~~& 5,566,993~~~& 36,525,895~~~& 10,669,627~~~& 5,555,190~~~\\
~~7 & 28,473~~~& 11,599,445~~~& 94,488,651~~~& 17,826,241~~~& 8,957,209~~~\\
~~8 & 80,397~~~& 20,782,043~~~& 112,880,347~~~& 26,325,724~~~& 12,534,152~~~\\
~~9 & 279,680~~~& 33,143,032~~~& 117,199,335~~~& 35,666,486~~~& 16,212,609~~~\\
~10 & 1,065,613~~~& 48,061,001~~~& 119,518,691~~~& 45,354,280~~~& 20,018,262~~~\\
\hline
\end{tabular}
\vspace{4mm}
\caption{The number of distinct -grams () in the 200\,MB datasets.
The number of distinct 12-grams for \texttt{dna} is 13,752,341.}
\label{table:qgrams}
\end{table}


An obvious space-time factor in a hash table with open addressing 
is its load factor . 
We checked several values of  on two datasets 
(Table~\ref{table:HT_lf}) 
to conclude that using  
is a reasonable alternative 
to , as the pattern search times grow by only about 10\% or less.

The number of bytes for one hash table with  entries 
and  load factor is, 
in our implementation of SA-hash, , 
since each entry contains two 4-byte integers.
For example, in our experiments the hash table for \texttt{english} 
with 
needed 20,782,043  184,729,272 bytes, i.e.,  88.1\%
of the size of the text itself.
Note that the overhead in the SA-hash-dense variant with the same  
is 20,782,043  138,546,954 bytes, i.e.,  66.1\% 
of the text size.


\begin{table}
\centering
\begin{tabular}{lccccccc}
\hline
                    & \multicolumn{7}{c}{HT load factor (\%)}  \\
\cline{2-8}
     &~~25~~&~~50~~&~~60~~&~~70~~&~~80~~&~~90~~&~~95~~ \\
\hline
\texttt{dna},  & 0.625 & 0.636 & 0.643 & 0.655 & 0.672 & 0.717 & 0.789 \\
\texttt{dna},  & 0.807 & 0.817 & 0.825 & 0.836 & 0.855 & 0.903 & 0.978 \\
\texttt{dna},  & 0.781 & 0.792 & 0.806 & 0.809 & 0.829 & 0.877 & 0.953 \\
\texttt{dna},  & 0.791 & 0.802 & 0.814 & 0.819 & 0.837 & 0.883 & 0.966 \\
\texttt{english},  & 0.734 & 0.740 & 0.744 & 0.749 & 0.754 & 0.769 & 0.782 \\
\texttt{english},  & 1.024 & 1.026 & 1.034 & 1.034 & 1.039 & 1.053 & 1.064 \\
\texttt{english},  & 1.013 & 1.019 & 1.023 & 1.027 & 1.036 & 1.042 & 1.057 \\
\texttt{english},  & 1.035 & 1.040 & 1.043 & 1.047 & 1.053 & 1.063 & 1.079 \\
\hline
\end{tabular}
\vspace{4mm}
\caption{Average pattern search times (in s) in function of the HT load factor 
 for the SA-hash algorithm (xxhash function used). Each 200-megabyte  dataset name followed with the pattern length ().}
\label{table:HT_lf}
\end{table}


Finally, in Table~\ref{table:space} we present the overall space use 
for the five non-compact SA variants: 
plain SA, SA-LUT2, SA-LUT3, SA-hash and SA-hash-dense, plus SA-allHT(-dense), 
which is a (not implemented) structure comprising a suffix array, a LUT2 and 
one hash table for each .
The space is expressed as a multiple of the text length  (including the text), 
which is for example 5.000 for the plain suffix array.
We note that the lookup table structures become a relatively smaller fraction 
when larger texts are indexed.
For the variants with hash tables we take two load factors: 50\% and 90\%.


\begin{table}
\centering
\begin{tabular}{lrrrrr}
\hline
   &~\texttt{dna}~&~\texttt{english}~&\texttt{proteins}~&~\texttt{sources}~&~~~~~\texttt{xml}~~\\
\hline
SA    & 5.000~~& 5.000~~& 5.000~~& 5.000~~& 5.000~~\\
SA-LUT2 & 5.001~~& 5.001~~& 5.001~~& 5.001~~& 5.001~~\\
SA-LUT3 & 5.320~~& 5.320~~& 5.320~~& 5.320~~& 5.320~~\\
SA-hash-50 & 6.050~~& 6.587~~& 5.278~~& 7.010~~& 5.958~~\\
SA-hash-90 & 5.583~~& 5.882~~& 5.154~~& 6.117~~& 5.532~~\\
SA-hash-dense-90 & 5.438~~& 5.661~~& 5.116~~& 5.838~~& 5.399~~\\
SA-allHT-50 & 6.472~~& 8.114~~& 5.296~~& 9.736~~& 7.353~~\\
SA-allHT-90 & 5.818~~& 6.730~~& 5.164~~& 7.631~~& 6.307~~\\
SA-allHT-dense-90 & 5.613~~& 6.298~~& 5.123~~& 6.973~~& 5.980~~\\
\hline
\end{tabular}
\vspace{4mm}
\caption{Space use for the non-compact data structures as a multiple 
of the indexed text size (including the text), with the assumption 
that text symbols are represented in 1 byte each and SA offsets are 
represented in 4 bytes. The datasets have 200\,MB in size. 
The value of  
for SA-hash-50 and SA-hash-90, used in the construction of these structures 
and affecting their size, is like in the experiments from Fig.~\ref{fig:times1a}.
The index SA-allHT-* contains LUT2 and one hash table for each 
, when  depends on the current dataset, 
as explained.
The -50 and -90 suffixes in the structure names denote the hash load factors 
(in percent).
}
\label{table:space}
\end{table}


In the next set of experiments we evaluated the FBCSA index. 
Its properties of interest, for various block size () and 
sampling step () parameters, are: the space use, 
pattern search times, times to access (extract) one random SA cell, 
times to access (extract) multiple consecutive SA cells. 
For  we set the values 32 and 64.
The  was tested in a wider range ().
Using  results in better compression but decoding a cell 
is also slightly slower (see Fig.~\ref{fig:times2}).

We tried to compare FBCSA against its competitors. 
Unfortunately, we were unable to run LCSA~/~LCSA-Psi~\cite{GNFjea14}
(in spite of contacting its authors) 
and MakCSA~\cite{DBLP:journals/fuin/Makinen03} cannot (directly) 
access single SA cells.
From the comparison with the results presented in~\cite[Sect.~4]{GNFjea14} 
we conclude that FBCSA is a few times faster in single cell access 
than the other related algorithms, MakCSA~\cite{DBLP:journals/fuin/Makinen03} 
(augmented with a compressed bitmap from~\cite{RamanRR02} to extract arbitrary 
ranges of the suffix array) and LCSA~/~LCSA-Psi~\cite{GNFjea14}, 
at similar or better compression.
Extracting  consecutive cells is not however an efficient operation for FBCSA 
(as opposed to MakCSA and LCSA~/~LCSA-Psi, see Figs~5--7 in ~\cite{GNFjea14}), 
yet for small  the time growth is slower than linear, due to a few sampled 
(and thus written explicitly) SA offsets in a typical block (Fig.~\ref{fig:times3}).
Therefore, in extracting only 5 or 10 successive cells our index is still 
competitive.



We also compared FBCSA variants against MakCSA in search (count) queries.
Alas, it was possible to use MakCSA only for 50-megabyte datasets.
The results of our comparison are shown in Fig.~\ref{fig:fb_mak}.
MakCSA wins on \texttt{proteins50} and \texttt{english50}, 
is comparable to our variants on \texttt{dna50}, and 
clearly loses on \texttt{sources50} and \texttt{xml50} 
(note the logarithmic scale for the last dataset).
Also, we can add two remarks.
First, the relative overhead of the lookup tables (LUT2 and LUT3) is roughly 
4 times smaller for 200-megabyte datasets, yet (as mentioned) 
MakCSA does not support such large datasets.
Second, the hash component of the index may be optimized for the FBCSA indexes, 
with hopefully more competitive space-time tradeoffs.


\begin{figure}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{dna50_m16.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{english50_m16.eps}
}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{proteins50_m16.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{sources50_m16.eps}
}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{xml50_m16.eps}
}
\caption[Results]
{Pattern search time (count query) for FBCSA variants () and 
M{\"a}kinen's CSA.
The different results in a series are obtained from varying the sampling 
parameter  in .
All times are averages over 500K random patterns of the same length 
.
The patterns were extracted from the respective texts.
Note the logarithmic scale for the \texttt{xml50} dataset.}
\label{fig:fb_mak}
\end{figure}


\begin{figure}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{32_run1.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{64_run1.eps}
}
\caption[Results]
{FBCSA index sizes and cell access times with varying  parameter 
().
The parameter  was set to 32 (left figure) or 64 (right figure).
The times are averages over 10M random cell accesses.}
\label{fig:times2}
\end{figure}


\begin{figure}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{32_run5.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{64_run5.eps}
}
\centerline{
\includegraphics[width=0.49\textwidth,scale=1.0]{32_run10.eps}
\includegraphics[width=0.49\textwidth,scale=1.0]{64_run10.eps}
}
\caption[Results]
{FBCSA, extraction time for  (top figures) and  (bottom figures) 
consecutive cells, with varying  parameter 
().
The parameter  was set to 32 (left figures) or 64 (right figures).
The times are averages over 1M random cell run extractions.}
\label{fig:times3}
\end{figure}


\section{Conclusions}

We presented two simple full-text indexes.
One, called SA-hash, speeds up standard suffix array searches with 
reducing significantly the initial search range, thanks to a hash table 
storing range boundaries of all intervals sharing a prefix of a specified 
length.
Despite its simplicity, we are not aware of such use of hashing in 
exact pattern matching, and the approximately 3-fold speedups compared 
to a standard SA may be worth the extra space in many applications.

The other presented data structure is a compact variant of the suffix array, 
related to M{\"a}kinen's compact SA~\cite{DBLP:journals/fuin/Makinen03}.
Our solution works on blocks of fixed size, which provides int32 alignment 
of the layout.
This index is rather fast in single cell access, but not competitive 
if many (e.g., 100) consecutive cells are to be extracted.

Several aspects of the presented indexes require further study.
In the SA-hash scheme collisions in the HT may be eliminated 
with 
perfect hashing.
This should also reduce the overall space use.
In case of plain text, the standard suffix array component may be replaced 
with a suffix array on words~\cite{DBLP:conf/cpm/FerraginaF07}, with possibly 
new interesting space-time tradeoffs.
The idea of deep buckets may be incorporated into some compressed indexes, 
e.g., to save on the several first LF-mapping steps in the FM-index.


\section{Acknowledgement}

The work was supported by the Polish National Science Centre under the project DEC-2013/09/B/ST6/03117 (both authors).



\begin{thebibliography}{99}

\bibitem{GP14}
Gog, S.---Petri, M.: Optimized succinct data structures for massive data.
  Software--Practice and Experience, Vol.~44, 2014, No.~11, pp.~1287--1314.

\bibitem{DBLP:conf/spire/KarkkainenP11}
K{\"a}rkk{\"a}inen, J.---Puglisi, S.~J.: Fixed block compression boosting in {FM}-indexes. 
  In: R.~Grossi, F.~Sebastiani and F.~Silvestri (Eds.): 
  String Processing and Information Retrieval, 
  Proceeding of the 18th International Symposium, SPIRE'11, Pisa, Italy, October 2011,
  Vol. 7024 of Lecture Notes in Computer Science, Springer, pp. 174--184.

\bibitem{NMacmcs06}
Navarro, G.---M{\"a}kinen, V.: Compressed full-text indexes. ACM Computing Surveys, Vol.~39, 2007, No.~1, article 2.

\bibitem{MM90}
Manber, U.---Myers, G.: Suffix arrays: a new method for on-line string searches.
  In: D.~S.~Johnson (Ed.):
  Discrete Algorithms,
  Proceedings of the First Annual ACM-SIAM Symposium, SODA'90, San Francisco, CA, USA, January 1990,
  ACM/SIAM, pp. 319--327.

\bibitem{DBLP:conf/cpm/Makinen00}
M{\"a}kinen, V.: Compact suffix array. 
  In: R.~Giancarlo and D.~Sankoff (Eds.): 
  Combinatorial Pattern Matching,
  Proceeding of the 11th Annual Symposium, CPM'00, Montreal, Canada, June 2000,
  Vol. 1848 of Lecture Notes in Computer Science, Springer, pp. 305--319.

\bibitem{DBLP:journals/fuin/Makinen03}
M{\"a}kinen, V.: Compact suffix array -- a space-efficient full-text index.
  Fundamenta Informaticae, Vol.~56, 2003, No.~1--2, pp.~191--210.

\bibitem{GR14}
Grabowski, S.---Raniszewski, M.: Two simple full-text indexes based on the suffix
  array. 
  In: J.~Holub and J.~{\v{Z}}{\v{d}}{\'{a}}rek (Eds.): Proceedings of the
  Prague Stringology Conference (PSC) 2014, Czech Technical University in
  Prague, Czech Republic, 2014, pp. 179--191.

\bibitem{Wei73}
Weiner, P.: Linear pattern matching algorithm.
  In: Proceedings of the 14th
  Annual IEEE Symposium on Switching and Automata Theory, Washington, DC, 1973,
  pp. 1--11.

\bibitem{Far97}
Farach, M.: Optimal suffix tree construction with large alphabets. 
  In: Foundations of Computer Science,
  Proceedings of the 38th IEEE Annual Symposium, FOCS'97, Miami Beach, FL, USA, October 1997,
  IEEE Computer Society, pp. 137--143.

\bibitem{Gri07}
Grimsmo, N.: On performance and cache effects in substring indexes. Tech. Rep.
  IDI-TR-2007-04, NTNU, Department of Computer and Information Science, Sem
  Salands vei 7-9, NO-7491 Trondheim, Norway (2007).

\bibitem{KB00}
Kurtz, S.---Balkenhol, B.: Space efficient linear time computation of the
  {Burrows} and {Wheeler} transformation. In: Numbers, Information and
  Complexity, Kluwer Academic Publishers, 2000, pp. 375--383.

\bibitem{DBLP:journals/talg/FranceschiniG08}
Franceschini, G.---Grossi, R.: No sorting? {B}etter searching! ACM Transactions
  on Algorithms, Vol.~4, 2008, No.~1, pp.~2:1--2:13.

\bibitem{DBLP:conf/cpm/FerraginaF07}
Ferragina, P.---Fischer, J.: Suffix arrays on words. 
  In: B.~Ma and K.~Zhang (Eds.):
  Combinatorial Pattern Matching,
  Proceeding of the 18th Annual Symposium, CPM'07, London, Canada, July 2007,
  Vol. 4580 of Lecture Notes in Computer Science, Springer, pp. 328--339.

\bibitem{KS03}
K{\"a}rkk{\"a}inen, J.---Sanders, P.: Simple linear work suffix array
  construction. 
  In: J.~C.~M. Baeten, J.~K.~Lenstra, J.~Parrow and G.~J.~Woeginger (Eds.): 
  Automata, Languages and Programming,
  Proceeding of the 30th International Colloquium, ICALP'03,
  Eindhoven, The Netherlands, June 2003,
  Vol. 2719 of Lecture Notes in Computer Science, Springer, pp. 943--955.

\bibitem{NZC11}
Nong, G.---Zhang, S.---Chan, W.~H.: Two efficient algorithms for linear time suffix
  array construction. IEEE Transactions on Computers, Vol.~60, 2011, No.~10, 
  pp.~1471--1484.

\bibitem{karkkainen1995suffix}
K{\"a}rkk{\"a}inen, J.: Suffix cactus: A cross between suffix tree and suffix
  array. In: Z.~Galil and E.~Ukkonen (Eds.): 
  Combinatorial Pattern Matching,
  Proceedings of the Sixth Annual Symposium, CPM'95, Espoo, Finland, July 1995,
  Vol. 937 of Lecture Notes in Computer Science, Springer, pp. 191--204.

\bibitem{abouelhoda2002enhanced}
Abouelhoda, M.~I.---Kurtz, S.---Ohlebusch, E.: The enhanced suffix array and its
  applications to genome analysis.
  In: Algorithms in Bioinformatics, 
  Proceedings of the Second International Workshop, WABI'02, Rome, Italy, September 2002,
  Vol. 2452 of Lecture Notes in Computer Science, Springer, pp. 449--463.

\bibitem{cole2006suffix}
Cole, R.---Kopelowitz, T.---Lewenstein, M.: Suffix trays and suffix trists:
  structures for faster text indexing. 
  In: M. Bugliesi, B. Preneel, V. Sassone and I. Wegener (Eds.):
  Automata, Languages and Programming,
  Proceeding of the 30th International Colloquium, Part I, ICALP'06,
  Vol. 4051 of Lecture Notes in Computer Science,
  Springer, 2006, pp. 358--369.

\bibitem{FG15}
Fischer, J.---Gawrychowski, P.: Alphabet-dependent string searching with
  wexponential search trees. In: Combinatorial Pattern Matching (CPM), 2015, to appear.

\bibitem{GV00}
Grossi, R.---Vitter, J.~S.: Compressed suffix arrays and suffix trees with
  applications to text indexing and string matching. 
  In: F.~F. Yao and E.~M. Luks (Eds.):
  Theory of Computing,
  Proceedings of the 32nd ACM Symposium, STOC'00, Portland, OR, USA, May 2000,
  ACM Press, pp. 397--406.

\bibitem{DBLP:conf/soda/Sadakane02}
Sadakane, K.: Succinct representations of lcp information and improvements in
  the compressed suffix arrays. 
  In: D. Eppstein (Ed.):
  Discrete Algorithms,
  In: Proceedings of the 13th Annual ACM-SIAM Annual Symposium, SODA'02, San Francisco, CA, USA, January 2002,
  ACM/SIAM, pp. 225--232.

\bibitem{FM00}
Ferragina, P.---Manzini, G.: Opportunistic data structures with applications. 
  In: Foundations of Computer Science,
  Proceedings of the 41st IEEE Annual Symposium, FOCS'00, Redondo Beach, CA, USA, November 2000,
  IEEE Computer Society, pp. 390--398.

\bibitem{FGNVjea08}
Ferragina, P.---Gonz{\'a}lez, R.---Navarro, G.---Venturini, R.: Compressed text
  indexes: From theory to practice. Journal of Experimental Algorithms, Vol.~13, 2009, pp.~12:1.12--12:1.31.

\bibitem{DBLP:conf/cpm/GonzalezN07}
Gonz{\'a}lez, R.---Navarro, G.: Compressed text indexes with fast locate. 
  In: B.~Ma and K.~Zhang (Eds.):
  Combinatorial Pattern Matching,
  Proceeding of the 18th Annual Symposium, CPM'07, London, Canada, July 2007,
  Vol. 4580 of Lecture Notes in Computer Science, Springer, pp. 216--227.

\bibitem{GNFjea14}
Gonz{\'a}lez, R.---Navarro, G.---Ferrada, H.: Locally compressed suffix arrays. 
  ACM Journal of Experimental Algorithmics, Vol.~19, 2014, No.~1, article 1.

\bibitem{GM13}
Gog, S.---Moffat, A.: Adding compression and blended search to a compact two-level
  suffix array.
  In: O. Kurland, M. Lewenstein and E. Porat (Eds.):  
  String Processing and Information Retrieval, 
  Proceeding of the 20th International Symposium, SPIRE'13, Jerusalem, Israel, October 2013,
  Vol. 8214 of Lecture Notes in Computer Science, Springer, pp. 141--152.

\bibitem{GMCTW14}
Gog, S.---Moffat, A.---Culpepper, J.~S.---Turpin, A.---Wirth, A.: Large-scale pattern search using reduced-space on-disk suffix arrays. IEEE Transactions on
  Knowledge and Data Engineering, Vol.~26, 2014, No.~8, pp.~1918--1931.

\bibitem{RamanRR02}
Raman, R.---Raman, V.---Rao, S.~S.: Succinct indexable dictionaries with
  applications to encoding -ary trees and multisets. 
  In: D. Eppstein (Ed.):
  Discrete Algorithms,
  Proceedings of the 13th Annual ACM-SIAM Symposium, SODA'02, San Francisco, CA, USA, January 2002,
  ACM/SIAM, pp. 233--242.

\end{thebibliography}



\label{lastpage}
\end{document}
