\documentclass{eptcs}
\providecommand{\anno}{2012} \providecommand{\firstpage}{27}
\providecommand{\lastpage}{47}
\usepackage{breakurl}
\usepackage[strings]{underscore}
\usepackage[english]{babel}
\usepackage{color}
\usepackage{MnSymbol}
\usepackage{bussproofs}
\usepackage{tikz}
\usepackage{tkz-graph}
\usepackage{lscape}
\usepackage{amsthm}
\usepackage{rotating}
\usetikzlibrary{arrows,shapes}
\usepackage{ifthen}
\usepackage{calc}
\usepackage{multido}

\newcommand{\tuple}[1]{{\langle #1 \rangle}}
\newtheorem{definition}{Definition}
\newtheorem{theorem}{Theorem}
\newtheorem{lemma}{Lemma}
\newtheorem{proposition}{Proposition}
\newcommand{\partset}[1]{\ensuremath{\textbf{2}^{#1}}}
\newcommand{\eqdef}[0]{\; \triangleq \; } \newcommand{\abbrev}[1]{#1, \relax}
\newcommand{\ie}[0]{\abbrev{\textit{i.e.}}}
\newcommand{\eg}[0]{\abbrev{\textit{e.g.}}}
\newcommand{\cf}[0]{\abbrev{\textit{cf.}}}
\newcommand{\sem}[1]{\left \lsem {#1} \right \rsem}
\newcommand{\A}[1]{[{#1}]}
\newcommand{\E}[1]{\langle{#1}\rangle}
\newcommand{\hybridset}[0]{\textsf{F}_{\mathcal H}}
\newcommand{\restrict}[2]{{#1}|_{#2}}
\newcommand{\dom}[1]{\text{Dom} \; #1} \newcommand{\cause}[0]{\mathrel{\medcircle\kern-3.25pt \to}} \newcommand{\pcause}[0]{\mathrel{\odot\kern-3.25pt \to}} \newcommand{\remcause}[0]{\mathrel{\oplus\kern-3.25pt \to}} \newcommand{\allcause}[0]{\mathrel{\ostar\kern-3.25pt \to}} \newcommand{\reg}[0]{\longrightarrow}
\newcommand{\activate}[0]{\stackrel{+}{\reg}}
\newcommand{\inhibit}[0]{\stackrel{-}{\reg}}
\newcommand{\htrace}[1]{\setcounter{ti}{\theti+\thetinterv}
\draw (\theti,\thehauteur+0.2) -- (\theti,-0.1) node[below] {\tiny }; }
\newcommand{\htracei}[2]{\setcounter{ti}{\theti+\thetinterv}
\pgfmathparse{\theti+#1-1}\let\z\pgfmathresult
\foreach \x in {\theti,...,\z} {
	\draw (\x,\thehauteur+0.2) -- (\x,-0.1) node[below] {\tiny }; }
\setcounter{ti}{\theti+#1-1} }

\newcommand{\hist}[2]{\pgfmathparse{\thetimeg+(#1-\thetimeg)/2}\let\z\pgfmathresult
\setcounter{timeg}{#1}
\draw [very thick](\thetimeg,\thehauteur+0.2) -- (\thetimeg,-0.1);
\draw [very thick](\thetimeg,-0.6) -- (\thetimeg,-1.3);
\node at (\z,-1){\small };
}
\newcommand{\hcmax}[2]{\draw [dashed] (-0.1,#1) -- (11,#1);
\node at (-0.3,#1 ) {\textbf{#2}};
}
\newcounter{timeg}
\setcounter{timeg}{0}
\newcounter{hauteur}
\setcounter{hauteur}{0}
\newcounter{longueur}
\setcounter{longueur}{0}
\newcounter{tinterv}
\setcounter{tinterv}{0}
\newcounter{ti}
\setcounter{ti}{0}
\newenvironment{hgraph}[3]{\setcounter{longueur}{#1}
\setcounter{hauteur}{#2}
\setcounter{tinterv}{#3}
\setcounter{timeg}{0}
\setcounter{ti}{0}
\begin{tikzpicture}[xscale=0.6, yscale=0.7]
\draw[->,very thick] (0,0) -- (#1+0.3,0);
\node at (#1-0.5,0.5){\footnotesize \textit{time}};
\draw[->,very thick] (0,-1.3) -- (0,#2+0.3);
\node at (0.4,#2){\footnotesize \textit{q.}};
\node at (-0.27,-0.4) {\small T};
\draw [semithick](0,-0.7) -- (#1,-0.7);
\node at (-0.27,-1) {\small H};
\draw [semithick](0,-1.3) -- (#1,-1.3);
}{\end{tikzpicture}}

\title{ \textsc{gubs}, a Behavior-based Language for Open System \\
Dedicated to Synthetic Biology}
\author{Adrien Basso-Blandin
\institute{IBISC lab. }
 \institute{Evry University}
\email{abasso@ibisc.univ-evry.fr}
\and 
 Franck Delaplace
\institute{IBISC lab.}
 \institute{Evry University}
\email{franck.delaplace@ibisc.univ-evry.fr}}

\def\titlerunning{\textsc{gubs}, a Behavioral Language for Synthetic Biology}
\def\authorrunning{A.~Basso-Blandin and F.~Delaplace}

\begin{document}
\maketitle

\begin{abstract}
In this article, we propose a domain specific language, GUBS (Genomic Unified Behavior Specification), dedicated to the behavioural specification of synthetic biological devices, viewed as discrete open dynamical systems. \textsc{gubs} is a rule-based declarative language. By contrast to a closed system, a program is always a partial description of the behaviour of the system. The semantics of the language accounts the existence of some hidden non-specified actions that possibly alter the behaviour of the programmed devices.
The compilation framework follows a scheme similar to automatic theorem proving, aiming at improving synthetic biological design safety.



\end{abstract}
\section{Introduction}


Synthetic biology is an emerging scientific field combining the investigative nature of biology with the constructive nature of engineering~\cite{Purnick2009} to design synthetic biological systems. The issue is to devise new functionality/behaviour that does not exist in nature.
Then, the field of synthetic biology is looking forward principles and tools to make the biological devices inter-operable and programmable~\cite{Lu2009}. Synthetic biology projects were first focusing on the design and the improvement of small genetic devices comparable to logical gates for electronic circuits~\cite{Regot2010,Clancy2010}. Recently, projects have attempted to develop large bio-systems integrating different devices with as a long-term goal, the design of \textit{de-novo} synthetic genome~\cite{Gibson2010}. In this endeavour, the computer-aided-design (CAD) environments play a central role by providing the required features to engineer systems: specification, analysis, and tuning~\cite{Bilitchenko2011,Pedersen2009a, Umesh2010, Czar2009}. Pioneer applications show the valuable potential of such environments in \textsc{igem} competition. 


Currently, the design specifies the structural assembly of \textsc{dna} sequences (biobrick) as in \textsc{genocad}~\cite{Cai2009}. Although this description is indispensable to provide a finalized specification of devices, the abstraction level seems inappropriate for tackling large bio-systems. The required size of programs for sequence description likely makes the task error-prone and un-come-at-able. In the same way as large softwares cannot be programmed in binary, large biological systems cannot be described as a\textsc{dna} sequence assembly. Then, scaling up the complexity of the synthetic biological systems needs to complete the structural description by an additional abstract programming layout based on a high-level programming language and harness the automatic conversion of the design specification into a DNA sequence, like compilers. High level programming language for synthetic biology is announced as a key milestone for the second wave of synthetic biology to overcome the complexity of large synthetic system design~\cite{Purnick2009}. Nonetheless, in this domain, language technology is still in its infancy and transforming this vision into a concrete reality remains a daunting challenge. 

Such high-level language should describe the devices in term of functionalities, offering the ability to program the behaviour directly instead of the structure supporting this behaviour.
Indeed, behaviour specification contributes to accurately document the device by adding its behavioural description, to assess its functionality automatically and formally, notably by generating test-benches from this specification, and to get a relative independence to technology because different biological structures can carry out the same functionality. In this framework, the components are selected and organized automatically or semi-automatically to generate a structural description of the device at compile phase whose behaviour complies with the specified function.
A such approach has been already achieved in hardware by using languages as \textsc{vhdl}~\cite{Ashenden2002} or \textsc{verilog}~\cite{Thomas1998} to overcome the growing complexity of electronic circuits. However, the major difference in synthetic biology relates to the openness of biological system. Thereby, the issue is to propose a behavioural language for open systems.
More precisely, \textsc{gubs} is a rule-based declarative language dedicated to the behavioural specification of \emph{discrete open dynamical systems }for synthetic biology interacting with its environment. \textsc{gubs} symbolically defines
the behaviours to provide a relative independence from structures by postponing the biological component selection at compile phase. Within this framework, the compiler translates the behavioural specification to a structural description of a device whose behaviour carries the functional features defined by a program. The proposed compilation method is inspired by automated theorem proving. 

After introducing the main features of \textsc{gubs} language (Section~\ref{sec:gubs}), we define the semantics of \textsc{gubs} based on hybrid logic. Then, we detail the proof-based principles governing the compilation (Section~\ref{sec:compil}) illustrated with a complete example (Section~\ref{sec:example}). After a survey of the related works, Section~\ref{sec:related}, we conclude (Section~\ref{sec:conclusion}). 


\section{\textsc{gubs} language}
\label{sec:gubs}
\newcommand{\context}[1]{[#1]}

In this section, we describe the main features of \textsc{gubs}. 
\paragraph{Constant and variables.} In \textsc{gubs}, two kinds of objects are distinguished: the \emph{constants} and the \emph{variables}. The constants designate the pre-defined objects in a corpus of knowledge. In biology, the constants may refer to proteins or genes of interest. For example, the agent  refers to LacZ protein or gene. By convention, their name starts with a capital letter. The variables refer to an abstraction of these pre-defined objects and can be potentially replaced (substituted) by any constant. By convention, the variable names start with a minuscule letter.
\paragraph{Agents, attributes and states.}
The \emph{agents} represent the biological objects. Their different observable \emph{states} characterize their different \emph{behaviours}. The behaviours actually define the different capacities for actions on the state of the other agents. They are characterized symbolically by a set of \emph{attributes} categorizing these different capacities. The real significance of the attributes is a matter of convention depending on the targeted realization (\eg protein pathways, gene network) and will be addressed through examples. 
For instance, the regulatory activity of a gene is observationally related to thresholds of \textsc{rna} transcripts concentration. At a given threshold, a gene regulates a given set of genes whereas at another one the regulation applies to another set of genes (See Figure~\ref{fig:causes}). The different thresholds define the levels of gene activities leading to different regulatory activities.
For a gene , if we identify three different kinds of regulatory activities, the state of this gene will be defined by three different attributes  that characterize symbolically three possible behaviours. 
For example,  expresses the fact that agent  is in state  and then ready for the action corresponding to this attribute. In some cases, a single state is sufficient to qualify the capacity for the action of the agent. Hence, the agent is identified to its capacity. Then,  means that agent  is available. 

By contrast,  signifies that the state of the agent differs from  ( when an agent has a single capacity). It is worth to point out that, not being in a state defined by an attribute, does not necessarily means that the agent state is in another attribute. Indeed, for open systems the state of the agents could be of any sort that does not necessarily belong to the pre-defined attributes. 

 Two kinds of relations on attributes are defined: an order, , meaning ``less capacity than'' and an inequality, , meaning ``different capacity than''. Then  implies that the capacity for the action of  includes the capacity related to . Usually, in gene regulatory model~\cite{Delaplace2010}, the set of genes regulated at a given level will also be regulated at a higher level. By contrast, in signalling pathways, the phosphorylation of a protein induces a conformational change of the structure leading to a specific signalling potentiality not occurring in the unphosphosrylated conformation. Assuming that  and  respectively represents the phosphorylated and the unphosphorylated conformations of protein , we have . Then,  implies  implicitly. The attributes and the relation between attributes will be declared as follows:

A simple set of attributes replaces the relations if unknown and no specific relation is set between attributes.

Finally, the description of the agent state is extended to a collection of agent states as follows: , meaning that all the agent states, , are observed concomitantly.

\paragraph{Trace, event, and history.} 
A \textsc{gubs} program describes a behaviour, its interpretation is based on the observations of designed systems. Then, the issue is to formalize the notion of behaviour observation. To this end, we focus on the notion of \emph{trace} that symbolically represents the evolution of some quantities related to the agents of interest by the evolution of these agent states. A trace can be obtained from experiments by establishing a correspondence between measurements of some quantities (\eg \textsc{rna} transcript concentration) and attributes of agents. Formally, a trace, , is a finite sequence of agent state sets where each set contains the agent states at a given instant. For instance, the evolution of a concentration evolving from  to  for  may be described by the following trace of  instants: 

However, all the events in a trace are not necessarily relevant with regards to the behaviour description. For example, if we focus on the evolution from  to  for  , only three events are relevant for the behaviour description: ; without accounting the intermediary evolution stages occurring between. Then, the behaviour recognition always emphasizes the key events in a trace entailing its contraction to show their succession. Such a contracted series is called a \emph{consistent history} of the expected behaviour. Generally speaking, an history is related to a \emph{chronological division} of a trace into periods where the events of a period represent all the agent states occurring at each instant.
Then, an history is a sequence of these event sets.
Given a trace , and a chronological division, , corresponding to a sequence of the starting dates for each period, the history is a sequence of agent states occurring in each period, , such that each . Hence, a consistent history is purposely made to point the characteristic event steps of a behaviour description out.

 In the previous example, a chronological division\footnote{Step  is inserted as an extra step to comply with the definition of the chronological division.} of the trace leading to an history consistent with the expected evolution from  to  for  is  which corresponds to following discrete time-intervals ).
The resulting history is: . Notice that  also fits. However, the chronological division  leads to an inconsistent history because the level \textit{Mid} is not seen as an intermediary event in the history (See also Figure~\ref{fig:causes} depicting the trace and consistent history of the dependences). The formal definition of the consistency in the scope of the semantics will be given in Section~\ref{sec:semantics}.

\paragraph{Behavioral dependence and observation spot.}
\begin{figure}[ht]
\begin{center}
\begin{hgraph}{11}{3}{1}
\htracei{2}{\emptyset}
\htracei{2}{\{c\}}
\htracei{2}{\emptyset}
\htracei{2}{\{e\}}
\htracei{3}{\emptyset}
\draw(5.5,3)node[fill=white] {};
\begin{scope}
\clip (0,0) rectangle (11,3.2);
\draw [color=blue, thick](0,0.1) -- (1,0.1);
\draw[domain=-1:2,samples=100,color=blue,thick] plot ({\x+2},{1.7/(1+exp(-(\x*4.5)))+0.1});
\draw[domain=-2:1,samples=100,color=blue,thick] plot ({\x+5},{1.7/(1+exp(-(-(\x)*4.5)))+0.1});
\draw [color=blue,thick](6,0.1) -- (11,0.1);
\draw [color=red,thick](0,0.2) -- (4,0.2);
\draw[domain=-2:2,samples=100,color=red,thick] plot ({\x+6},{2.3/(1+exp(-(\x*4.5)))+0.2});
\draw[domain=-2:2,samples=100,color=red,thick] plot ({\x+9},{2.3/(1+exp(-(-(\x)*4.5)))+0.2});
\end{scope}
\hcmax{1.8}{c}
\hcmax{2.5}{e}
\hist{3}{\emptyset}
\hist{7}{\{c\}}
\hist{9}{\{e\}}
\hist{11}{\emptyset}
\end{hgraph}

\bigskip
\begin{hgraph}{11}{3}{1}
\htracei{2}{\emptyset}
\htrace{\{c\}}
\htracei{2}{\{c,e\}}
\htrace{\{e\}}
\htracei{5}{\emptyset}
\draw(5.5,3)node[fill=white] {};
\begin{scope}
\clip (0,0) rectangle (11,3.2);
\draw [color=blue,thick](0,0.1) -- (1,0.1);
\draw[domain=-1:2,samples=100,color=blue,thick] plot ({\x+2},{1.7/(1+exp(-(\x*4.5)))+0.1});
\draw[domain=-2:1,samples=100,color=blue,thick] plot ({\x+6},{1.7/(1+exp(-(-(\x)*4.5)))+0.1});
\draw [color=blue,thick](7,0.1) -- (11,0.1);
\draw [color=red,thick](0,0.2) -- (1,0.2);
\draw[domain=-2:2,samples=100,color=red,thick] plot ({\x+3},{2.3/(1+exp(-(\x*4.5)))+0.2});
\draw[domain=-2:2,samples=100,color=red,thick] plot ({\x+7},{2.3/(1+exp(-(-(\x)*4.5)))+0.2});
\draw [color=red,thick](9,0.2) -- (11,0.2);
\end{scope}
\hcmax{1.8}{c}
\hcmax{2.5}{e}
\hist{3}{\emptyset}
\hist{4}{\{c\}}
\hist{8}{\{c,e\}}
\hist{11}{\emptyset}
\end{hgraph}
\begin{hgraph}{11}{3}{1}
\htracei{2}{\emptyset}
\htracei{2}{\{c\}}
\htracei{3}{\emptyset}
\htracei{4}{\{e\}}
\draw(5.5,3)node[fill=white]{};
\begin{scope}
\clip (0,0) rectangle (11,3.2);
\draw[domain=-2:2,samples=100,color=blue, thick] plot ({\x+2},{1.7/(1+exp(-(\x*4.5)))+0.1});
\draw[domain=-2:2,samples=100,color=blue, thick] plot ({\x+5},{1.7/(1+exp(-(-(\x)*4.5)))+0.1});
\draw [color=blue,thick](7,0.1) -- (11,0.1);
\draw [color=red,thick](0,0.2) -- (5,0.2);
\draw[domain=0:4,samples=100,color=red,thick] plot ({\x+5},{2.3/(1+exp(-((\x-2)*4.5)))+0.2});
\draw [color=red,thick](8,2.5) -- (11,2.5);
\end{scope}
\hcmax{1.8}{c}
\hcmax{2.5}{e}
\hist{3}{\emptyset}
\hist{8}{\{c\}}
\hist{10}{\{e\}}
\hist{11}{\{e\}}
\end{hgraph}
\end{center}
\caption{The curves represent the typical behaviours of the causal dependences based on the time evolution of a quantity () related to agents  and  (\eg \textsc{rna} transcript for gene regulation). The symbolic agent states  and  are here both associated to the maximal threshold of the quantity. The symbolic trace () is issued from a periodic sampling of the evolution by identifying whether  or  occur. A consistent history () complying to a causal dependence definition is represented below the trace.
The first graphic illustrates the normal causality: , the second the persistent:  and the third the remanent one: .}
\label{fig:causes}
\end{figure}
 A behavioural dependence identifies a relation between behaviours as a causal relation on events. Basically, the dependences should define the control of agents on another. However, the definition of the causality also needs to tackle the openness of a system by adapting it to this context. A seminal definition of the causality, proposed by Hume~\cite{Hume1739}, is formulated in terms of regularity on events: ``[we may define] a cause to be an object, followed by another, and where all the objects similar to the first are followed by objects similar to the second''. Although this definition appropriately characterizes the notion of control, the openness of the system implies to account the environment actions that possibly alter the causal dependence chain. For example, a programmed activation  may be contradicted by an existing inhibition  addressing the same target gene . Hence, while  is active, it may appear that  will not be active because the regulatory strength of  is greater than the regulatory strength of , contradicting the expected activation by a hidden inhibition. Hence, pushed to the limit, this consideration prevents the ability to describe any behaviour causally because any programmed action can be unexpectedly preempted by an external one. 

However, by assuming that the design always describes a new functionality which is not observed naturally, the effect becomes the event indicating the effectiveness of a causal relation. As no cause external to the description can trigger the effect, the over-determination by unknown causes is prevented, then insuring that the program is the sole device entailing the expected effect in the biological system.
Hence, the definition of the causal dependence will be governed by the effect leading to the following definition of the dependence: \emph{``if effect  would occur then  occurs''}. Moreover, the scope of future (resp. past) is narrowed to a \emph{closest future (resp. past) period}, representing the fact that a response is always expected in a given delay. Notice that, the proposed definition circumvents the afore mentioned problem illustrated by the hidden inhibition because if the effect does not occur the question of the existence of a cause is meaningless. This definition is somehow equivalent to the causal claims proposed by Lewis~\cite{Lewis2000} in terms of counter-factual conditionals, \ie ``If  had not occurred,  would not have occurred''.
 
Three behavioural dependences are defined in \textsc{gubs}: the \emph{normal} denoted by , \emph{persistent} by , and \emph{remanent} by . Informally, for normal dependence the cause precedes the effect providing the effect is observed; for persistent dependence the cause still precedes the effect but it is maintained while the effect is observed; and for remanent dependence, the effect is maintained despite the cause has disappeared. These dependences symbolize common biological interactions. For instance,
in genetic engineering, the recombination enables the emergence of a regulated gene or an hereditary trait permanently. A such mechanism typifies the remanent dependence in biology. The relations between gene expression at steady state are symbolized by persistent dependence. The behavioural dependences are defined as follows (see Section~\ref{sec:semantics} for their formalization): 
\begin{itemize}
\item : if  occurs then  occurs in the closest past.
\item : if  occurs then  occurs in the closest past and also currently.
\item : if  occurs then, either  occurs in the closest past or the dependence complies to the property of the normal dependence.
\end{itemize}
Figure~\ref{fig:causes} exemplifies the correspondence between experimental traces, symbolic traces and the history for the causal dependences. 
 All the dependences are extended to a set of causes and a set of consequences, \ie . 
For example, let us define the activation and the inhibition as follows:
 and , 
the program depicting a negative regulatory circuit with two genes, \ie , is:


The \emph{observation spots} describe the set of observations expected in a trace. For instance, observing that gene  is at level high is written . As the activation of a dependence lies on the observation of the effect, the observation spot is used to determine which effects must be necessarily observed. For example, in the negative regulatory circuit, the characteristic observation spots are: .

\paragraph{Compartment \& Context.} A compartment encloses a set of dependences making them local to the compartment. For instance,  describes a normal dependence occurring in compartment . The compartments are hierarchically organized and all the compartments are included in another except for the outermost one.
Although the compartments directly refer to the compartmentalized cellular organization (\eg nucleus, mitochondria), they are also used to emphasize the isolation of some interactions by syntactically enclosing the dependences into a compartment.  refers to an agent state in compartment . 

A context refers to a stimulus acting on the system, as environmental conditions or external signalling. The application of a context  to a set of dependences  is written  where  is either a variable or a constant. This means that dependences of  are triggered when the context  is present. For instance, recently Ye et al.~\cite{Ye2011} explore the opto-genetics signalling to control the expression of target transgenes. The blue-light induces the expression of transgene () via a signalling cascade leading to the binding of \textsc{nfat} transcription factor to a specific promoter (\textsc{pnfat}). The following program using a context summarizes the process: . A context can be decomposed to several contexts, , meaning that all the conditions must be met to trigger the dependences of . The interpretation is equivalent to a context cascading, . Moreover, the observation spots and the attribute definition are context insensitive.

\subsection{Semantics of \textsc{gubs}}
\label{sec:semantics}
The interpretation of \textsc{gubs} is a formula such that the set of all the 
models validating it defines all the possible histories complying to the programmed behaviour.
The interpretation is based on multi-modal hybrid logic with the ``Always'' operator, .

\paragraph{Hybrid logic.} In what follows, we recall the formal syntax and semantics of hybrid logic.
 The hybdrid logic~\cite{Blackburn2006,Brauner2010} offers the possibility to denominate worlds by new symbols called \textit{nominals}. They will be used in satisfaction modal operators ; the formula  asserts that  is satisfied at the unique point named by the nominal  identifying a particular truth values of a formula at this point. 
Given a set of propositional symbol, , a set of relational symbol , and a set of nominal  disjoint to 
 , a set of well formed formula in the signature of  is defined as follows:
 
with  and .
 Moreover, the syntax is extended to other logical operators classically \footnote{ .}:
 . 


The interpretation is carried out using the Kripke model satisfaction definition (Table~\ref{tab:hybrid}).
 is interpreted as the satisfaction of a formula  by a model  at world  where  stands for the realizability relation (\ie ``is a model of''). A model \emph{validates} a formula, denoted by , if and only if it is satisfied for all the worlds of the model (\ie ).
\begin{definition}[Kripke model]
\label{def:kripke}
A Kripke model is a structure  where  is a non-empty set of \emph{worlds},  a subset of relational symbols denoting the \emph{modalities},  a relation of accessibility,  an interpretation attributing to each nominal and propositional variable a set of worlds such that any nominal addresses one world at most (\ie ).

\medskip
\noindent
By convention,  stands for the union of the accessibility relation, .
\end{definition}
A \emph{modal theory} of a model  regarding to a set of formulas , , is the set of formulas of  validated by , \ie . 
 denotes the set of all models validating , \ie .

\begin{table}[ht]
\small

\label{tab:hybrid}
\caption{Hybrid logic interpretation.}
\end{table}
\paragraph{Semantics.}
A \textsc{gubs} program is interpreted by a hybrid logic formula where the modal operators characterize here the temporal observations on an history:
  means \emph{``observed in all the closest futures''} and  means \emph{``observed in a possible closest future at least''} (resp.  for the closest past). Moreover, we assume that the accessibility relations, , are indexed by the non empty parts of the set of all the contexts of a program , denoted by  (\ie ). Then, a non-empty set of contexts ,, is a modality, \ie  with  by convention.
Let  be the set of words  with the concatenation operation and the neutral element, the empty word  and  the set of well-formed formulas of , the semantics is defined by four functions:
, where  respectively stand for the set of \textsc{gubs} programs, the set of agent state set and the set of relations on attributes.  initiates the interpretation. Table~\ref{tab:semantics} defines these functions. 
\begin{table}[ht]
\footnotesize

\label{tab:semantics}
\caption{Semantics of \textsc{gubs}. In the definition,  represents an attribute,  a behaviour,  an agent,  a set of agent states or an agent state,  a relation on attributes,  a compartment,  a set of contexts and  a set of behaviours (\ie contexts, compartments, dependences, attributes, observation spots).
}
\end{table}
For instance, the program of the negative regulatory network,
, is translated into the following formula:


\paragraph{Consistent history.} Now, we formally define the consistency of the history with regards to models. 
An history is assimilated to a path in a model ending by a world labelled with an observation spot label. 
The set of Kripke-models validating the interpretation of a program , , not only contains all the consistent histories, but also the possible histories corresponding to behavioural alterations due to external perturbations. Thus, the compilation generates a device such that all the models validating its interpretation integrate all the observations related to the program, including the consistent and the inconsistent ones.

More precisely, the consistency lies on the identification of the largest number of ``relevant'' events characterizing a complete causal chain described in a program. As an history is also a model, a consistent history should validate the interpretation of the complete causal chain.
 The dependence formula set  of a program  corresponds to a set of formulas where each formula is the interpretation of a dependence taken separately with the attributes related to the involved agents.
 By definition of the semantics, any model validating the interpretation of a program also validates each formula of this set. The consistency of an history is then based on the validated formulas of this set by this history. \emph{An history  is consistent for  if and only if no other modal theory of histories based on  (\ie  with  as an history), ending with the same labelled world includes the modal theory of this history (\ie ).}

\section{Compilation}
\label{sec:compil}
At compile phase, a program is transformed to a structure (\eg a \textsc{dna} sequence) while inserted in a vector cell, should behave according to the programmed specification. The structure will result to an assembly of several devices stored in a library of components (\eg parts registry). As the design relates here to a behavioural/functional description, we need to bridge the gap between structural and functional description. This stage is called the \emph{functional synthesis}. The issue is to select a set of components whose assembly preserves the behaviour of the program. To achieve this goal, a \textsc{gubs} program is associated to each component to describe its behaviour. Thereby, the component assembly corresponds to a program assembly preserving the behaviour of the compiled program.
Preserving a behaviour is laid on a property called the \emph{behavioural inclusion} formalizing the fact that the characteristic observational traits of the specified function must be recognized in traces related to the device experiments. In other words, we can exhibit histories consistent with the programmed behaviour from histories consistent with the device behaviour description. 
The behavioural inclusion is defined from the interpretation of the programs, as a logical consequence (Definition~\ref{def:behinc}). 
\begin{definition}[Behavioral inclusion]
\label{def:behinc}
A program  \emph{behaviourally includes} another program , if and only if the interpretation of the latter is a logical consequence of the interpretation of the former:

\end{definition}
The behavioural inclusion is a pre-order\footnote{A reflexive and transitive relation.} such that the empty program, denoted by , is a minimum, meaning that a program with no behaviour can be observed in all traces. And a program whose interpretation equals , is a maximum. 
 Figure~\ref{fig:subModel} illustrates the behavioural inclusion on a particular model.
\begin{figure}[t]
	\centering
	\footnotesize \begin{tabular}{l r}
	
	& \begin{tikzpicture}[baseline={([yshift={-\ht\strutbox}]current bounding box.north)},scale=0.8]
\small
\GraphInit[vstyle=Normal]
\SetVertexNormal[Shape=circle,LineWidth=0.5pt,LineColor=black]


\Vertex[x=1,y=2,LabelOut=true,Lpos=180,Ldist=1,L={\textbf{a :}},style={minimum size=15}]{P}
\Vertex[x=3,y=2,LabelOut=true,Lpos=180,Ldist=1,L={\textbf{b :}},style={minimum size=15}]{Q}
\Vertex[x=3,y=8,L={},style={minimum size=15}]{P0}
\Vertex[x=1,y=6,L={},style={fill=black!15,minimum size=15}]{P1}
\Vertex[x=3,y=5,L={},style={minimum size=15}]{P2}
\Vertex[x=1,y=4,L={},style={fill=black!15,minimum size=15}]{P3}
\Vertex[x=1,y=2,L={},style={fill=black!15,minimum size=15}]{P4}
\Vertex[x=3,y=2,L={},,style={fill=black!15,minimum size=15}]{P5}


\Vertex[x=5,y=8,L={},style={minimum size=15}]{Q0}
\Vertex[x=4,y=6,L={},style={minimum size=15}]{Q1}
\Vertex[x=6,y=2,LabelOut=true,Lpos=180,Ldist=1,L={\textbf{c :}},style={minimum size=15}]{R}
\Vertex[x=6,y=6,L={},style={fill=black!15,minimum size=15}]{Q2}
\Vertex[x=4,y=4,L={},style={fill=black!15,minimum size=15}]{Q3}
\Vertex[x=6,y=4,L={},style={fill=black!15,minimum size=15}]{Q4}
\Vertex[x=6,y=2,L={},style={fill=black!15,minimum size=15}]{Q5}
\tikzstyle{EdgeStyle}=[post]
\Edge[label={}](P0)(P1)
\Edge[label={}](P0)(P2)
\Edge[label={}](P1)(P3)
\Edge[label={}](P2)(P5)
\Edge[label={}](P3)(P4)
\Edge[label={}](P3)(P5)
\Edge[label={}](Q0)(Q1)
\Edge[label={}](Q0)(Q2)
\Edge[label={}](Q2)(Q3)
\Edge[label={}](Q2)(Q4)
\Edge[label={}](Q3)(Q5)
\Edge[label={}](Q4)(Q5)
\end{tikzpicture}
\end{tabular}
\caption{Behavioral inclusion example. Consistent histories of  necessary contains worlds coloured in gray. }
	\label{fig:subModel}
\end{figure}
\paragraph{Observability.}
\newcommand{\obs}[0]{\textbf{obs}\,}
It may arise that no history will be consistent with a programmed behaviour. For example, the program  is not observable in a trace. Indeed, its interpretation yields to the following formula: , false in all models because world  must both satisfies  and  by definition of the persistent dependence. A \textsc{gubs} program is said \emph{observable} if and only if the formula resulting from its interpretation is validated by one model at least. Hence, the interpretation of an unobservable program is an antilogy. An unobservable program can be assimilated to a programming error. The detection of such errors can be carried out at compile-phase by using tableaux method~\cite{Cerrito2011}
that automatically determines whether a formula is satisfiable in a model. Indeed \textsc{gubs} uses fragment of \textit{HL(@)} logic which is decidable. Notice that an observable program always behaviourally includes an observable program (Proposition~\ref{prop:comp1}).
\begin{proposition} A program behaviourally included in an observable program is observable: 
\label{prop:comp1}

\end{proposition}
\subsection{Functional synthesis}
The functional synthesis is the operation whereby biological components of a library are selected and assembled to generate a device behaviourally including the designed function. The behaviour of each component is described by a \textsc{gubs} program. At its simplest, the functional synthesis could be considered as a proper substitution of variables by constants. For example, in the following activation ,  will be substituted by gene , providing that component  describes the activation . However, more complex situations may arise during component selection. For example, if the activation  occurs with another regulation only \ie  then the selection of  adds a supplementary regulation.

Formally, a finite substitution is a set of mappings, , on variables and constants such that a variable can be substituted by a variable or a constant, and a constant can only substituted by itself\footnote{  or  represents its application on program  and identity substitutions are omitted.}. For instance, we have:
.

\paragraph{Functional synthesis rules.} The functional synthesis is defined by rules (Table~\ref{tab:rules}) governing the component assembly. Only the dependences and the attributes will be functionally synthesize. 
The observation spots are considered as annotations used for the compilation process. 
To insure the correctness, each transform must preserved the seminal behaviour. Hence, 
each program resulting from the application of a rule must behaviourally includes the previous one. 
Formally, the functional synthesis is modelled by a relation on programs denoted by , \ie  where  is the initial program and  the transformed one, such that each rule insures that:
\emph{ is correct with regards to a substitution , that is 
 and  is observable.}
Also notice that the behavioural inclusion is preserved by substitution (Proposition~\ref{prop:subst-behinc}). 

 \begin{proposition} For all substitutions , we have: . 
\label{prop:subst-behinc}
\end{proposition}

\noindent
Table~\ref{tab:rules} describes the functional synthesis rules\footnote{Rules are of the form: \AxiomC{hypothesis}
\UnaryInfC{conclusion}
\DisplayProof. }. 
  is a set of components representing the library. 
 denotes the fact that program  corresponds to an assembly including  \ie  where  or  may be an empty program.
\begin{table}[ht]
\begin{center}
\footnotesize
\renewcommand{\arraystretch}{2}
\setlength{\doublerulesep}{\arrayrulewidth}
\begin{tabular}{l r} 
\multicolumn{2}{c}{ \textsc{- Instantiation -}} \\
\multicolumn{2}{c}{
\AxiomC{}
\AxiomC{}
\AxiomC{}
\RightLabel{(Inst.)}
\TrinaryInfC{}
\DisplayProof}
\\ \multicolumn{2}{c}{ \textsc{- Commutativity, Contraction -}} \\
\AxiomC{}
\RightLabel{(Com.)}
\UnaryInfC{}
\DisplayProof 
&
 \AxiomC{}
\RightLabel{(Cont.)}
\UnaryInfC{}
\DisplayProof 
\\ 
\multicolumn{2}{c}{ \textsc{- Assembly -}} \\
\multicolumn{2}{c}{ 
\AxiomC{}
\AxiomC{}
\AxiomC{} 
\RightLabel{(Asm.)}
\TrinaryInfC{}
\DisplayProof 
} \end{tabular}
\end{center}
\caption{ Functional synthesis rules}
\label{tab:rules}
\end{table}
Rule~(Inst.) describes the fact that an observable instance of a part of a component in the library is functionally synthesized. Rule~(Com.) expresses the commutativity of the assembly. Rule (Cont.) contracts the redundant formulation of programs. Finally, Rule~(Asm.) details the conditions for an assembly of two components, each representing a functional synthesis of a part of the designed function. 
A detailed example of their use on a real case is given in Section~\ref{sec:example}.
\begin{theorem} The functional synthesis rules (Table~\ref{tab:rules}) are correct.
\label{the:fsr}
\end{theorem}
\begin{table}[ht]
\begin{center}
\footnotesize
\textsc{- dependences - }
\renewcommand{\arraystretch}{2}
\setlength{\doublerulesep}{\arrayrulewidth}
\begin{tabular}{c c c} 
\AxiomC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\DisplayProof 
&
\AxiomC{} 
\RightLabel{(N2P.)}
\UnaryInfC{}
\DisplayProof 
&
\AxiomC{} 
\RightLabel{(R2N.)}
\UnaryInfC{}
\DisplayProof 
\end{tabular}

\bigskip
\textsc{- agent states - }

\begin{tabular}{l c r} 
\AxiomC{} 
\RightLabel{(SCom.)}
\UnaryInfC{}
\DisplayProof 
& 
\AxiomC{}
\RightLabel{(SCont.)}
\UnaryInfC{}
\DisplayProof
&
\AxiomC{}
\RightLabel{(Incl.)}
\UnaryInfC{}
\DisplayProof
\end{tabular}
\end{center}
\caption{Rules for the dependences and the agent states.  stands for a collection, , of agent states, including negation, and  stands for the rest of the program. }
\label{tab:causes}
\end{table}
Another set of rules, more specifically devoted to dependences (Table~\ref{tab:causes}), defines the alternate possibilities to express similar behaviours. The table also includes the rules for agent sets. 
Rule (Trans.) expands the chain of the persistent dependences by adding intermediary dependence to refine a pathway. Rule (N2P.) transforms a normal dependence to a persistent one since the latter is a normal dependence with an additional property. And Rule (R2N.) transforms a remanent dependence to a normal dependence, since normal dependence is also remanent dependence with a repetition of the effect restricted to one step.
According to these rules, all the dependence chains can be implemented with persistent dependences.

A possible algorithm for the assembly could be based on a combinatorial application of the rules. However, such algorithm may reveal inefficient in practice. The conditions for an efficient algorithm of compilation should be based on an internal representation of a program, as a set of contextualized dependences with attributes, , such that  are respectively: a set of attributes specification related to the agent involved in the dependency, a set of contexts and sets of agent states. Any program can be encoded under this representation from a normal form of the program (not detailed here). Accordingly, the problem solved by the compilation algorithm can be defined as follows (Definition~\ref{def:probasm}):
\begin{definition}[Functional Synthesis Problem]
\label{def:probasm}
Let  be set where each  is a set of contextualized dependences with attributes and  a set of contextualized dependences with attribute, can we find the smallest observable subset of components  such that there exists a substitution  so that its application on the components of  form a cover of ,\ie

\end{definition}
As the set cover problem is reducible to this problem, the problem is NP-complete. Then, the resolution is oriented towards a heuristic algorithm. 

\section{Example}
\label{sec:example}
\begin{figure}[ht]
\newcommand{\labelstyle}[1]{\textit{\scriptsize #1}}
\centering
	\begin{tikzpicture}[scale=0.7]
	\small
\GraphInit[vstyle=Normal]
\SetVertexNormal[Shape=circle,LineWidth=0.5pt,LineColor=black]
\SetUpEdge[lw =1.5pt, labelcolor = white, labeltext = black, labelstyle = {sloped, text= black,above}]
\Vertex[x=-2.5 ,y=6,L={\labelstyle{\scriptsize Tetr}},style={minimum size=30pt}]{P0}
\Vertex[x=-0 ,y=6,L={\labelstyle{LuxR}},style={minimum size=30pt}]{P1}
\Vertex[x=2.5 ,y=5,NoLabel=true, style={minimum size=3pt, inner sep=0pt, outer sep=4pt, shape=circle,blue}]{A3}
\Vertex[x=2.2 ,y=6.,NoLabel=true, style={minimum size=3pt, inner sep=0pt, shape=circle,blue}]{P3}
\Vertex[x=2 ,y=5.5,NoLabel=true, style={minimum size=3pt, inner sep=0pt, outer sep=4pt, shape=circle,blue}]{A4}
\Vertex[x=2.7,y=6.2,NoLabel=true, style={minimum size=3pt, inner sep=0pt, shape=circle,blue}]{P5}
\Vertex[x=3.2 ,y=5.3,NoLabel=true,style={minimum size=3pt, inner sep=0pt, shape=circle,blue}]{A0}
\Vertex[x=2.8 ,y=5.7,NoLabel=true,style={minimum size=3pt, inner sep=0pt, outer sep=5pt,shape=circle,blue}]{A2}

\Vertex[x=2.5 ,y=5,LabelOut=true,Lpos=270,Ldist=1,L={},style={minimum size=3pt, inner sep=0pt, shape=circle,blue}]{P8}
\Vertex[x=5 ,y=6,L={\labelstyle{LuxR}},style={minimum size=30pt}]{Q0}
\Vertex[x=7.5 ,y=6,L=\labelstyle{},style={minimum size=30pt}]{Q1}
\Vertex[x=6 ,y=4,L={\labelstyle{Cl}},style={minimum size=30pt}]{Q2}
\Vertex[x=8.5 ,y=4,L={\labelstyle{Lacl}},style={minimum size=30pt}]{Q3}
\Vertex[x=10 ,y=6,L={\labelstyle{GFP}},style={minimum size=30pt}]{Q4}
\tikzstyle{EdgeStyle}=[color=black!70,post]
\Edge[label={}](P0)(P1)
\Edge[label={}](P1)(A4)
\Edge[label={}](A2)(Q0)
\Edge[label={}](Q0)(Q1)
\Edge[label={}](Q0)(Q2)
\Edge[label={}](Q2)(Q3)
\Edge[label={}](Q1)(Q4)
\Edge[label={}](Q3)(Q4)
\end{tikzpicture}
\caption{The band detector regulatory circuit.}
	\label{fig:bandDetect}
\end{figure}

The compilation process is here exemplified in a real case by the design of the Band Detector proposed in \cite{Basu2005}.
This example explains how from a simple abstract definition of the functionality a complex design can be synthesized.
Accordingly, \textsc{gubs} may be used to describe a behaviour with a high-level of programming well as a low-level, detailing the components
involved in the process. Although, the functional synthesis is not yet performed automatically, it is worth to point out that the different transforms of the high-level program to obtain the final design complies to rules of Tables~\ref{tab:rules},~\ref{tab:causes}, insuring its correctness and so, its functional safety in the context of open system. 
 
 The design aims at forming patterns of different colours in a population of bacteria exploiting the quorum sensing phenomenon by staining with fluorescent protein (GFP).
The amount of molecules of interest that receives a cell depends on its relative position to the cell diffusing the molecule of interest controlled by an external event: more the cell is far from the source, the fewer is the amount of molecules received. 
The activation or inhibition of the fluorescent protein due to the concentration will distinguish the bands surrounding the source. In the original design, the protein does not fluoresce in an intermediary band. 

From a computing standpoint, we can assimilate the design to a message transmission coupled to a sensor/actuator responsible for fluorescence, then leading to a concise \textsc{gubs} program presented below: the diffusive molecule is \textit{AHL} which production is controlled by a context and the observation is applied on \textit{GFP}. Two categories of cells are defined: the \emph{Sender} and the \emph{Receiver}. Therefore, two \textsc{gubs} programs identify the two cell types.
 

Figure ~\ref{fig:bandDetect} describes the original genetic circuit used in the article. The diffusible molecule is the constant \emph{AHL}.
The gene \textit{LuxR} has three activation thresholds: at Level , it activates both \textit{LaclM1} and \textit{Cl}, at level , the amount of \textit{AHL} only allows activation of \textit{Cl}, and finally, at level , none are activated.
\begin{table}[ht]
\footnotesize
\centering

\caption{Part of the database dedicated to the Band Detector.}
\label{tab:bddWeiss}
\end{table}
We show that from the sender-receiver program, we obtain the original design by applying the afore mentioned rules
 with an appropriate selection of components. The regulations of Figure~\ref{fig:bandDetect} are described in \textsc{gubs} program
(Table~\ref{tab:bddWeiss}) translating in term of dependences and relations on their attributes their regulatory action. We focus here on some illustrative steps of the sender program compilation. The complete functional synthesis is given in Appendix. The compilation consists in finding the appropriate components whose assembly behaviourally includes the sender-receiver program, with the particularity that the diffusive molecule must be the same in both programs. To ease compilation follow-up, we label each dependency of the sender-receiver program (Table~\ref{tab:reecCode}).
\begin{table}[t]
\scriptsize
\centering

with  as attributes of \textit{AHL}.
\caption{Separation of the dependences.}
\label{tab:reecCode}
\end{table}
Let us consider  whose compilation is closed to  and . Notice that  cannot be directly instantiated with any component because, in the one hand, the component  contains a context like  but applied on gene  instead of \textit{AHL}, and on the other hand  has the \textit{AHL} molecule but no context is defined. So, to fit  with the components ,  and , first, the normal dependence is converted to persistent one (Rule~(N2P.)).

Thereby, the resulting dependence can be separated to match the assembly  by applying (Trans.) rule twice.  and  are fresh variables.

Finally, we obtain a new program program  compatible with , and
each variable is substituted by a constant (biological element) with the application of Rule~(Inst.). For  we have:

By following this scheme for  and , we respectively obtain  and .
The final assembly corresponds to the functional synthesis of \textit{Sender} program. 

In conclusion, the functional synthesis generates the original genetic circuit (Figure~\ref{fig:bandDetect}) from the sender program. A similar approach can be also applied to obtain the receiver program (see the complete proof in Appendix~\ref{sec:proofEx}).


\section{Related works}
\label{sec:related}
Several domain specific languages have been developped to model and simulate biological systems. 
Based on process-calculus, seminally used to model process concurrency, several rule-based languages model protein interactions~\cite{Priami2001,Danos07,Ciocchetta2009}. Another approach is based on logic, such as \textsc{biocham}~\cite{Calzone2006} that formalizes the temporal properties of a biological system. As these languages are dedicated to simulation, the objective is to close the systems because the simulations need to integrate all the characteristics of the analysed systems. By comparison, the purpose of \textsc{gubs} is different since the issue is to represent the behaviour of a synthetic device in an organism, leading to translate the notion of the openness of biological systems by the semantics of the language.

In synthetic biology, the structural description languages ~\cite{Czar2009,Pedersen2009a,Bilitchenko2011} allow to specify 
 well-formed genome sequences by grammars modularly and hierarchically. Although the sequence description is necessary, the programmer must previously anticipate the behaviour of the device to conceive. Besides, the behavioural design is not included in the program while it initially motivates it. In \textsc{gubs}, the design is driven by a behaviour description and sequence selection is postponed at compile phase. Moreover, the size of the structural description is also subject to a combinatorial explosion when the complexity of programmed systems increases.

Amorphous programming language has been also investigated to specify the biological devices at the scale of cell colony, here considered as a possible computing medium for amorphous program. J.~Beal~\cite{Beal2011} demonstrates the proof of concept of this approach in \textsc{proto}, showing the feasibility of an automatic compile chain. In \textsc{gubs}, the compile chain is based on rewriting rules whose correctness have been formally proved with regards to a semantics describing the constraints of an open system.

Developing a language for biological systems actually involves to consider several unknown due to their openness: lack of knowledge on all the interactions in biological circuits and imprecise definition of initial conditions. We only know the result of a chain of effects. Then, the major constraint for programming open system seems to be: how to provide an expressive language to describe the dynamics of such systems, but simple enough to capture the essence of the biological questions in a small program in order to allow programming of large biological systems with a program humanly achievable.

In the future, the design in synthetic biology will certainly require different programming layouts based on different paradigms addressing the integration levels of biological systems. In a tower of languages, starting from a language with collective operations on cell colony, using an amorphous programming language as Proto~\cite{Beal2011} or a language for dynamical systems with dynamical structures as MGS~\cite{Spicher2005}, and ending by a structural description programmed in a grammar based language, \textsc{gubs} language occupies the intermediary level dedicated to cell entity behavioural programming. 


\section{Conclusion}
\label{sec:conclusion}
In \textsc{gubs} language, we propose to characterize a programming paradigm abstracting the molecular interactions in the context of open system, that differs to an approach dedicated to biological system modeling. Accordingly, the interactions are symbolized by causal dependences whose interpretation is driven by effect. We have demonstrated the proof-of-concept of the compilation based on rewriting rules, and illustrated it on a realistic example. The perspective of this work is to find an efficient compilation algorithm. 
Identifying the biological parameters guiding the component selection should be a key issue in this undertaking.

\paragraph{Acknowledgements.}
The funding for most of this work is granted by the \textsc{anr synbiotic} (\textsc{anr blan} 0307 01) and we would like to thank the colleagues of this project for their fruitful discussions. 
\bibliographystyle{eptcs}
\bibliography{gub}	

\vfill 
\section*{Appendix}
\begin{table}[ht]

\label{tab:syntax}
\caption{Syntax of \textsc{gubs} program}
\end{table}
\subsection*{Proofs}
\begin{proof}[Proposition~\ref{prop:comp1}]
By contradiction, assume that  is unobservable, then there does not exist a model satisfying the formula. As  is observable, we deduce that there exists models satisfying , but no restricted model must satisfy , that contradicts the definition of the behavioural consequence.
\end{proof}

\begin{proposition}
\label{prop:correct1}
Let  be a formula, let  be a substitution on nominals, variables and relational symbols, let  be a model, we define the model  from  as follows:

\begin{enumerate}
\item  \label{p1}
\item  \label{p2}
\end{enumerate}
 we have: 

\begin{proof}
The proof is defined by induction on the formula:

without loss of generality, we assume that  is in Negation Normal Form where negation occurs only immediately before variables only. Recall that every formula can be set in Negation Normal Form. 
\begin{itemize}
\item . By (\ref{p1}), we have  leading to the equivalence.

\item . By definition of the realizability relation, this is equivalent to: . By (\ref{p1}), this equivalence holds.

\item . 
 By definition of the substitution, we have to prove:
 
By definition of the realizability relation we can formulate the property equivalently as follows: 
 
By induction hypothesis, we have:  and , implying the previous condition.

\item . The proof is similar to the proof of the previous item ().

\item 
By definition of the substitution we have to prove that: 

By definition of the realizability relation, this is equivalent to:

By setting , from (\ref{p1}) we have: .
By induction hypothesis, we have: 
The both last properties imply that:

which implies the initial property.

\item 
By definition of the substitution we prove that:


By definition of the realizability relation the condition is equivalent to:

 By setting , the following equivalence holds from (\ref{p2}): . By induction hypothesis, we have: .
The both last properties imply that:
 
which implies the initial property.

\item  The proof is similar to the previous item.
\item 
By definition of the substitution we prove that: 


By definition of the realizability relation, we have: 

which is directly verified by induction hypothesis.
\item  
The proof is similar to the previous item.
\end{itemize}
\end{proof}
\end{proposition}

\begin{proof}[Proposition~\ref{prop:subst-behinc}]
First, let us remark that when , the property is trivially verified.
Besides, under the assumption , if  is not observable the property is also verified because an unobservable program includes all programs behaviourally (Definition~\ref{def:behinc}).

In the rest of the proof, we assume that  is behaviourally included in  and  is observable (\ie  and ).
Hence, by definition of the observability there exists a model  such that . By proposition~\ref{prop:correct1}, we deduce that there exists a model  such that: . Moreover, as  by hypothesis, there exists  such that: . By construction of  we deduce that there exists a sub model of , denoted by , complying to the properties, (\ref{p1}) and (\ref{p2}) of Proposition~\ref{prop:correct1} which corresponds to . Moreover, we have  by Proposition~\ref{prop:correct1}. Then we conclude that: .
\end{proof}

\begin{proof}[Theorem~\ref{the:fsr}]
First, let us remark that  is true whenever  by definition of the behavioural inclusion (Definition~\ref{def:behinc}).
Hence, the proof doesn't consider the trivial verified case but rather the case where .
\begin{description}

\item[Inst.] Directly from the definition of the behavioural inclusion (Definition~\ref{def:behinc}).

\item[Com.] By definition of the semantics  with  and . Thus, for all  we have: . 
Hence, if  we conclude that: .

\item[Cont.] Similar to the proof of (Com.). 

\item[Asm.] First let us remark that  means that the substitution of the common variables are the same for  and , leading to, 
 and . Let . 
Then, we have the following property by definition of the semantics (Table~\ref{tab:semantics}) and .

Notice that the set of models, , is not empty since, by hypothesis,\\  holds.
As  and , any model validating  (resp. ) also validates , (resp. ) by definition of the functional synthesis. Then, we deduce that: 

Then, we conclude that:

\end{description}
\end{proof}
\subsection*{Complete compilation of the Band Detector}
\label{sec:proofEx}

\begin{table}[ht]
\begin{sideways}
\begin{tabular}{c}
\textsc{- Sender -} \\
{\tiny
\begin{array}[t]{c c c}
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\RightLabel{(Inst.)}
\BinaryInfC{}
\DisplayProof}
\\
&&\\
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\RightLabel{(Inst.)}
\BinaryInfC{}
\DisplayProof}
\\
&&\\
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\RightLabel{(Inst.)}
\BinaryInfC{}
\DisplayProof}
\\
&&\\
\AxiomC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(N2P.)}
\UnaryInfC{}
\DisplayProof
&
\AxiomC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(N2P.)}
\UnaryInfC{}
\DisplayProof
&
\AxiomC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(N2P.)}
\UnaryInfC{}
\DisplayProof
\\ 
&& \\
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\AxiomC{}
\RightLabel{(Asm.)}
\TrinaryInfC{}
\DisplayProof
}
\end{array} 
} \\ \\
\textsc{- Receiver -} \\
{ \tiny
\begin{array}[t]{c c c}
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\RightLabel{(Inst.)}
\BinaryInfC{}
\DisplayProof}\\
&&\\
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\RightLabel{(Inst.)}
\BinaryInfC{}
\DisplayProof}\\
&&\\
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\RightLabel{(Inst.)}
\BinaryInfC{}
\DisplayProof}\\
&&\\
\AxiomC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(N2P.)}
\UnaryInfC{}
\DisplayProof
&
\AxiomC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(N2P.)}
\UnaryInfC{}
\DisplayProof
&
\AxiomC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(Trans.)}
\UnaryInfC{}
\RightLabel{(N2P.)}
\UnaryInfC{}
\DisplayProof
\\
&&\\
\multicolumn{3}{c}{
\AxiomC{}
\AxiomC{}
\AxiomC{}
\RightLabel{(Asm.)}
\TrinaryInfC{}
\DisplayProof
}
\end{array}
} \\ \\
\textsc{- Final design -}
\\
{\tiny
\footnotesize

}
\end{tabular}
\end{sideways} 
\vspace{-7ex}\caption{Complete band detector compilation.}
\label{tab:proof2}
\end{table}
\end{document}
