

\documentclass{llncs}

\usepackage{enumerate}
\usepackage{amssymb}
\usepackage{epsfig}
\usepackage{tikz}

\usepackage[nothing]{algorithm}
\usepackage[]{algorithmic}
\usepackage{multirow}












\renewcommand{\algorithmiccomment}[1]{// #1}


\newcommand{\partitle}[1]{}                        \newcommand{\commentout}[1]{}





\newenvironment{pfs}{\noindent{\it Proof sketch.\rm}}{\hfill  \bigbreak}


\newcommand{\keywords}[1]{\par\addvspace\baselineskip
\noindent\keywordname\enspace\ignorespaces#1}




\newcommand{\bi}{\begin{itemize}}
\newcommand{\ei}{\end{itemize}}

\newcommand{\be}{\begin{enumerate}}
\newcommand{\ee}{\end{enumerate}}

\newcommand{\bd}{\begin{description}}
\newcommand{\ed}{\end{description}}




\title{Suffix Tree of Alignment:\\An Efficient Index for Similar Data}

\author{
    Joong Chae Na\inst{1}
\and
     Heejin Park\inst{2}
\and
     Maxime Crochemore\inst{3}
\and
     Jan Holub\inst{4}
\and \\
     Costas S. Iliopoulos\inst{3}
\and
     Laurent Mouchard\inst{5}
\and
     Kunsoo Park\inst{6}
\thanks{Corresponding author, E-mail: {\tt kpark@snu.ac.kr}}
}

\institute{
Sejong University,
Korea
\and
Hanyang University,
Korea
\and
King's College London,
UK
\and
Czech Technical University in Prague,
Czech Republic
\and
University of Rouen,
France
\and
Seoul National University,
Korea
}



\begin{document}

\maketitle



\begin{abstract}
We consider an index data structure for similar strings.
The generalized suffix tree can be a solution for this.
The generalized suffix tree of two strings  and 
 is a compacted trie representing all suffixes in  and .
It has  leaves and can be constructed in  time.
However, if the two strings are similar,
 the generalized suffix tree is not efficient
 because it does not exploit the similarity
 which is usually represented as an alignment of  and .

In this paper we propose a space/time-efficient
{\em suffix tree of alignment}
which wisely exploits the similarity in an alignment.
Our suffix tree for an alignment of  and  has
 leaves
where  is the sum of the lengths of {\em all} parts of  different from 
and  is the sum of the lengths of {\em some} common parts of  and .
We did not compromise the pattern search
to reduce the space.
Our suffix tree can be searched for a pattern 
in  time
where  is the number of occurrences of  in  and .
We also present an efficient algorithm
to construct the suffix tree of alignment.
When the suffix tree is constructed from scratch,
the algorithm requires  time
where  is the sum of the lengths of other common
substrings of  and .
When the suffix tree of  is already given,
it requires  time.
\keywords{Indexes for similar data, suffix trees, alignments}
\end{abstract}



\pagestyle{headings}  \pagenumbering{arabic}
\setcounter{page}{1}



\section{Introduction}




\partitle{Suffix trees}

The {\em suffix tree} of a string 
 is a compacted trie representing all suffixes of ~\cite{McCreight:76,Weiner:73}.
Over the years, the suffix tree
 has not only been a fundamental data structure in the area of string algorithms
 but also it has been used for many applications in engineering and computational biology.
The suffix tree can be constructed in  time
 for a constant alphabet~\cite{McCreight:76,Ukkonen:95}
 and an integer alphabet~\cite{Farach-Colton&Ferragina:00},
 where  denotes the length of .
The suffix tree has  leaves and requires  space.

\commentout{
\partitle{a family of ST}

Since the suffix tree was first introduced as a position tree by Weiner~\cite{Weiner:73} in 1973,
 many families and variants have been designed for various applications and circumstances.
The suffix array~\cite{KarkkainenSB06,Manber&Myers:93} is a sorted array of all suffixes
 and  the suffix automaton~\cite{Blumer&Blumer:85,Crochemore&Hancart:07},
 also known as DAWG (Directed Acyclic Word Graph),
 is an automaton representing suffixes of a given string,
 which can be seen as another representations of the suffix tree.
Moreover, compressed representations of the suffix tree
 such as the compressed suffix tree~\cite{Grossi&Vitter:05,Sadakane07TCS}
 and extensions to two-dimensional data
 such as two-dimensional suffix trees~\cite{Giancarlo:95,Kim&Na:11}
 have also been developed.
Other families and variants are
 the suffix tree of a tree~\cite{focs/Kosaraju89a},
 the generalized suffix tree~\cite{Amir&Farach:94,Gusfield:97},
 the suffix cactus~\cite{Karkkainen:95},
 the sparse suffix tree~\cite{Karkkainen&Ukkonen:96},
 the lazy suffix tree~\cite{GiegerichKS03},
 the truncated suffix tree~\cite{Na&Apostolico:03},
 the suffix binary search tree~\cite{Irving&Love:03},
 the suffix tray \& the suffix trist~\cite{ColeKL06},
 the linearized suffix tree~\cite{Kim&Kim:08}
 the geometric suffix tree~\cite{jacm/Shibuya10}, and so on.
}

\partitle{Similar data}

We consider storing
 and indexing multiple data which are very similar.
Nowadays, tons of new data are created every day.
Some data are totally original and substantially different from existing data.
Others are, however, created by modifying some existing data
 and thus they are similar to the existing data.
For example, a  new version of a source code is a modification of its previous version.
Today's backup is almost the same as yesterday's backup.
An individual Genome is more than 99\% identical to the Human reference Genome
 (the 1000 genome project~\cite{nature/1000Genomes10}).
Thus, storing and indexing similar data in an efficient way is becoming more and  more important.


Similar data are usually stored efficiently:
When new data are created, they are aligned with the existing ones.
Then, the resulting alignment shows the common and different parts of the new data.
By only storing the different parts of the new data, the similar data can be stored efficiently.

When it comes to indexing, however,
 neither the suffix tree nor any variant of the suffix tree
 uses this similarity or alignment to index similar data efficiently.
Consider the {\em generalized suffix tree}~\cite{Amir&Farach:94,Gusfield:97}
 for two similar strings  {\tt aaatcaaa} and  {\tt aaatgaaa}.
Three common suffixes {\tt aaa}, {\tt aa}, {\tt a}
 are stored twice in the generalized suffix tree.
Moreover, two similar suffixes {\tt aaatcaaa} and {\tt aaatgaaa}
 are stored in distinct leaves
 even though they are very similar.
Thus, the generalized suffix tree has  leaves, most of which are redundant.


\partitle{Previous works}

Recently, there have been some studies concerning efficient indexes for similar strings.
M\"{a}kinen et al.~\cite{recomb/MakinenNSV09,jcb/MakinenNSV10}
 first proposed an index for similar (repetitive) strings
 using run-length encoding, a suffix array, and BWT~\cite{Burrows&Wheeler:94}.
Huang et al.~\cite{aaim/HuangLSTY10} proposed an index of size  bits
 where  is the total length of common parts in one string,
  is the total length of different parts in all strings.
Their basic approach is building separately data structures for common parts
 and ones for different parts between strings.
A self-index based on LZ77 compression~\cite{Ziv&Lempel:77}
 has been also developed due to Kreft and Navarro~\cite{Kreft&Navarro:XX}.
Another index based on Lemple-Ziv compression scheme is
 due to Do et al.~\cite{aaim/DoJSS12}.
They compressed sequences using a variant of the relative Lempel-Ziv (RLZ) compression
 scheme~\cite{spire/KuruppuPZ10}
 and used a number of auxiliary data structures to support fast pattern search.
Navaro~\cite{iwoca/Navarro12} gave a short survey on some of these indexes.


\partitle{Motivation}

Although these studies assume slightly different models on similar strings,
 most of them adopt classical compressed indexes to utilize the similarity among strings,
 that is, they focus on how to efficiently represent or encode common (repetitive) parts in strings.
However, none of them support linear-time pattern search.
Moreover, their pattern search time do not depend on only the pattern length
 but also the text length,
 and some indexes require (somewhat complicated) auxiliary data structures
 to improve pattern search time.
In short, those data structures achieve smaller indexes
by sacrificing pattern search time.


\partitle{Our contribution}
In this paper, we propose an efficient index for similar strings
without sacrifice the pattern search time.
It is  a novel data structure for similar strings,
 named {\em suffix tree of alignment}.
We assume that strings (texts) are aligned with each others, e.g.,
 two strings  and  can be represented as
  and
 , respectively,
 where 's are common chunks and
  's and 's are chunks different from the other string.
(We note that the given alignment is not required to be optimal.)
Then, our suffix tree for  and 
 has the following properties.
(It should be noted that our index and algorithms can be generalized to three or more strings,
although we only describe our contribution for two strings
for simplicity.)
\bi
\item {\bf Space reduction}:
Our suffix tree has  leaves
 where  is the sum of the lengths of {\em all} chunks of  different from 
 (i.e., )
 and  is the sum of the lengths of {\em some} common chunks of  and .
More precisely,  is 
 where  is the longest suffix of  appearing at least twice in  or in .
The value of  is  on average for random
 strings~\cite{karlin1983new}.
Furthermore, the values of  and  are very small in practice.
For instance, consider
two human genome sequences from two different individuals.
Since they are more than 99\% identical,
 is very small compared to .
We have computed  for human genome sequences
and found out  is very close to ,
even though human genome sequences are not random.
Hence, our suffix tree is space-efficient for similar strings.
Note that the space of our index can be further reduced
 in the form of compressed indexes
 such as the compressed suffix tree~\cite{Grossi&Vitter:05,Sadakane07TCS}.
Our index is an important building block (rather than a final product)
 towards the goal of efficient indexing for highly similar data.
\item {\bf Pattern search}:
Our index is achieved without compromising the linear-time pattern search.
That is, using our suffix tree, one can search a pattern  in  time,
 where  is the number of occurrences of  in  and .
In addition to the linear-time pattern search,
 we believe that our index supports the most of suffix tree functionalities,
 e.g., regular expression matchings, matching statistics, approximate matchings,
 substring range reporting,
 and so on~\cite{Baeza-Yates&Gonnet:96,Bille&Gortz:11,Gusfield:97},
 because our index is a kind of suffix trees.
\ei


\partitle{contribution 2}

We also present an efficient algorithm to construct the suffix tree of alignment.
One na\"{i}ve method to construct our suffix tree is constructing the generalized suffix tree
 and deleting unnecessary leaves.
However, it is not time/space-efficient.
\bi
\item When our suffix tree for the strings  and  is constructed from scratch,
  our construction algorithm requires  time
  where  is the sum of the lengths of other parts of common chunks of  and .
 More precisely,  is 
  where  is the longest prefix of 
  such that  appears at least twice in  and 
  ( is the character preceding  in .
Likewise with ,
 the value of  is also very small compared to  or  in practice.
\item Our algorithm is incremental,
    i.e., we construct the suffix tree of 
     and then transform it to the suffix tree of the alignment.
    Thus, when the suffix tree of  is already given,
     it requires  time.
     is the minimum time required to make our index a kind of suffix tree
     so that linear-time pattern search is possible on both  and .
    Furthermore, our algorithm can be applied to the case when some strings are newly inserted
     or deleted.
\item Our algorithm uses constant-size extra working space except for our suffix tree itself.
 Thus, it is space-efficient compared to the na\"{i}ve method.
\ei
The space/time-efficiency of our construction algorithm becomes large
 when handling many strings.
The efficiency is feasible when the alignment has been computed in advance,
 which is the case in some applications.
For instance, in the Next-Generation Sequencing,
 the reference genome sequence is given
 and the genome sequence of a new individual is obtained
 by aligning against the reference sequence.
So, when a string (a new genome sequence) is obtained,
 the alignment is readily available.
Moreover, since our index does not require that the given alignment is optimal,
 we can use a near-optimal alignment instead of the optimal alignment
 if the time to compute an alignment is an important issue.
Since the given strings are assumed to be highly similar,
 a near-optimal alignment can be computed fast from exact string matching
 instead of dynamic programming requiring much time.




\section{Preliminaries}
\label{sec:preliminaries}




\subsection{Suffix trees}
\label{subsec:ST}



\partitle{Strings}

Let  be a string over a fixed alphabet .
A substring of  beginning at the first position of 
 is called a {\it prefix} of 
 and a substring ending at the last position of 
 is called a {\it suffix} of .
We denote by  the length of .
We assume that the last character of  is a special symbol ,
 which occurs nowhere else in .



\begin{figure}[t]
\centerline{\def\h{-2em}
\begin{tikzpicture}[font=\small,
                    inode/.style={text=black,inner sep=2pt,circle,draw=black},
                    leaf/.style={draw=black,text=black,inner sep=2pt, rectangle}]
    \node[inode] (root) at (3em,0em) {};
    \node[inode] (a) at (-4em,\h) {};
    \node[inode] (b) at (8em,\h) {};

\draw[dashed] (root) to node[above,midway,sloped] {\texttt{b}} (b);
    \draw[dashed] (b) --+ (-2em,-2em) ;
    \draw[dashed] (b) --+ (+2em,-2em) ;


    \node[inode] (aa) at (-8em,2*\h) {};
    \node[leaf] (ad) at (-4em,2*\h) {};
    \node[inode] (ab) at (2em,2*\h) {};


    \node[inode] (aaab) at (-11em,3*\h) {};
    \node[inode] (aab) at (-5em,3*\h) {};
    \node[inode] (aba) at (0em,3*\h) {};
    \node[leaf] (abbaabad) at (3em,6*\h) {};




    \node[inode] (aaba) at (-7em,3.5*\h) {};
    \node[leaf] (aaabaaabbaaba) at (-12em,6*\h) {};
    \node[leaf] (aaabbaaba) at (-10em,6*\h) {};
    \node[leaf] (aabbaaba) at (-8em,6*\h) {};
    \node[leaf] (abad) at (1em,6*\h) {};
    \node[leaf] (aabad) at (-6em,6*\h) {};
    \node[leaf] (abaaabbaabad) at (-1em,6*\h) {};
    \node[leaf] (aabbaabad) at (-4em,6*\h) {};




    \draw (root) to node[above,midway,sloped] {\texttt{a}} (a) ;


    \draw (a) to node[above,midway,sloped] {\texttt{a}} (aa) ;
    \draw (a) to node[above,midway,right] {\texttt{\#}} (ad) ;
    \draw (a) to node[above,midway,sloped] {\texttt{b}} (ab);






    \draw (ab) to node[above,midway,sloped] {\texttt{baaba\#}} (abbaabad) ;
    \draw (ab) to node[above,midway,sloped] {\texttt{a}} (aba) ;

    \draw (aa) to node[above,midway,sloped] {\texttt{ab}} (aaab) ;
    \draw (aa) to node[above,midway,sloped] {\texttt{b}} (aab) ;

    \draw (aab) to node[above,midway,sloped] {\texttt{a}} (aaba) ;

    \draw (aba) to node[above,midway,right] {\texttt{\#}} (abad) ;


    \draw (aaab) to node[above,midway,sloped] {\texttt{aaabbaaba\#}} (aaabaaabbaaba);
    \draw (aaab) to node[above,midway,sloped] {\texttt{baaba\#}} (aaabbaaba);
    \draw (aaba) to node[above,midway,sloped] {\texttt{aabbaaba\#}} (aabbaaba);


    \draw (aba) to node[above,midway,sloped] {\texttt{aabbaaba\#}} (abaaabbaabad);

     \draw (aab) to node[above,midway,sloped] {\texttt{baaba\#}}(aabbaabad);

     \draw (aaba) to node[above,midway,right] {\texttt{\#}}(aabad);

\end{tikzpicture}
 }
\caption{The suffix tree of string {\tt aaabaaabbaaba\#}.
\label{fig:ST}}
\end{figure}


\partitle{Suffix trees}

The {\em suffix tree} of a string  is a compacted trie with  leaves,
 each of which represents each suffix of .
Figure~\ref{fig:ST} shows the suffix tree of a string {\tt aaabaaabbaaba\#}.
For formal descriptions,
 the readers are referred to~\cite{Crochemore&Rytter:02,Gusfield:97}.
McCreight~\cite{McCreight:76} proposed a linear-time construction algorithm
 using auxiliary links called {\em suffix links}
 and also an algorithm for an {\em incremental editing},
 which transform the suffix tree of 
 to that of 
 for some (possibly empty) string , , , .



\begin{figure}[t]
\centerline{\def\h{-2.1em}
\begin{tikzpicture}[font=\small,
                    inode/.style={text=black,inner sep=2pt,circle,draw=black},
                    gleaf/.style={draw=black,fill=gray,text=black,inner sep=2pt, rectangle},
                    leaf/.style={draw=black,text=black,inner sep=2pt, rectangle}]

\node[inode] (root) at (10em,0em) {};

\node[inode] (a) at (0em,\h) {};
    \draw (root) to node[above,midway,sloped] {\texttt{a}} (a);

\node[inode] (aa) at (-6em,2*\h) {};
    \node[leaf] (ad1) at (-1em,2*\h) {};
    \node[gleaf] (ad2) at (1em,2*\h) {};
    \node[inode] (ab) at (8em,2*\h) {};
    \draw (a) to node[above,midway,sloped] {\texttt{a}} (aa);
    \draw (a) to node[left,midway] {\texttt{\#}} (ad1);
    \draw (a) to node[midway,right] {\texttt{\}} (abad2);

    \node[leaf] (abbaabad1) at (10em,4*\h) {};
    \node[gleaf] (abbabad2) at (12em,4*\h) {};
    \draw (abba) to node[above,midway,sloped] {\texttt{aba\#}} (abbaabad1);
    \draw (abba) to node[above,midway,sloped] {\texttt{ba\}} (aaabaabaabbabad2);

    \node[inode] (aabaa) at (-5em,5*\h) {};
    \node[leaf] (aabad1) at (-3em,5*\h) {};
    \draw (aaba) to node[above,midway,sloped] {\texttt{a}} (aabaa);
    \draw (aaba) to node[right,near start] {\texttt{\#}} (aabad1);

    \node[leaf] (aabbabad1) at (-1em,5*\h) {};
    \node[gleaf] (aabbabad2) at (1em,5*\h) {};
    \draw (aabba) to node[above,midway,sloped] {\texttt{aba\#}} (aabbabad1);
    \draw (aabba) to node[above,midway,sloped] {\texttt{ba\}} (abaabaabbabad2);
    \draw (abaab) to node[above,midway,sloped] {\texttt{baba\}} (aabaabaabbabad2);
    \draw (aabaab) to node[above,midway,sloped] {\texttt{baba\A=B=}.
 Leaves denoted by white squares and gray squares
  represent suffixes of  and , respectively.
\label{fig:GST}}
\end{figure}


\partitle{Generalized ST}

The {\em generalized suffix tree} of two strings  and 
 is a suffix tree representing all suffixes of the two strings.
It can be obtained
 by constructing the suffix tree of the concatenated string 
 where it is assumed that the last characters of  and 
  are distinct~\cite{Amir&Farach:94,Gusfield:97}.
Thus, the generalized suffix tree has  leaves
 and can be constructed in  time.
Figure~\ref{fig:GST} shows the generalized suffix tree of
 two strings {\tt aaabaaabbaaba\#} and {\tt aaabaabaabbaba\ABABABABA= \alpha \beta \gammaB= \alpha \delta \gamma\alpha\beta\gamma\delta\neq \betaBA\beta\delta\alpha (\beta /\, \delta) \gammaA=\alpha_1 \beta_1 \ldots \alpha_k \beta_k \alpha_{k+1}B= \alpha_1 \delta_1 \ldots \alpha_k \delta_k \alpha_{k+1}k\ge 1\alpha_1 (\beta_1 /\, \delta_1) \ldots \alpha_k (\beta_k /\, \delta_k) \alpha_{k+1}AB\# \in \Sigma\alpha_{k+1}i=1,\ldots, k\alpha_{i+1}\alpha_1\beta_i\delta_i\beta_i\alpha_{i+1}\delta_i\alpha_{i+1}\alpha_{i+1}i=1,\ldots,k-1\beta_{i}\beta_{i+1}\delta_{i}\delta_{i+1}\alpha_{k+1}\#\alpha_{k+1}\beta_i\delta_i\alpha_{i}\alpha_{i+1}\beta_i\alpha_{i+1}\delta_i\alpha_{i+1}cc\alpha_i\beta_i\alpha_{i+1}\delta_i\alpha_{i+1}k=1\alpha\beta\gamma\delta\alpha (\beta /\, \delta)\gammaA= \alpha \beta \gammaB= \alpha \delta \gamma\alpha^{a}\alpha^{b}\alphaAB\alpha^{*}\alpha^{a}\alpha^{b}\alpha^{*}\alphaAB\gamma\alpha^{*} \beta \gamma\gamma\alpha^{*} \delta \gamma\gamma\alpha' (\beta /\, \delta) \gamma\alpha'\alpha\alpha^{*}AB\alpha^{a}\alpha^{b}\alpha^{*}\alpha^{*}\beta\delta(\beta /\, \delta)\gamma\alpha^{*}(\beta /\, \delta)\gamma\alpha^*\alpha (\beta /\,\delta) \gammaTAB\beta \gamma\delta \gamma\alpha'(\beta /\, \delta) \gamma\alpha'\alphavvABABAB\gamma(\beta / \delta)\gamma(\beta / \delta)\gamma\alpha' (\beta /\, \delta) \gamma\alpha'\alpha\alpha^{*}\alpha^{*}\alpha'AB\alpha' (\beta /\, \delta) \gamma\alpha'' (\beta /\, \delta) \gamma\alpha''\alpha^{*}\alpha''AB(\beta /\, \delta)\alpha'' \beta \gamma\alpha'' \delta \gammaxy\alpha' (\beta /\, \delta) \gamma\alpha'\beta\delta\beta \gamma\delta \gammaTT^AAT^AO(|A|)T^ATBT^AB\gamma\alpha^{*}\delta \gamma\gamma\alpha\delta\gamma\alpha^{*}\delta \gamma\gammaT^AABB\gamma\alpha^{*}\delta \gamma\alpha^{*}\delta \gamma\alpha^{a}\alpha^{a}\delta \gamma\gamma\alpha^{*}\alpha^{*} \delta \gamma\alpha^{a}\delta \gamma\alpha\delta\gamma\alpha^{*}\delta \gammaA\alpha^{a}T^A\chi\chi\chi\chiA\chiT^A\alpha^{a}\alpha'\alpha\alpha'T^A\alpha_{(j)}\alphaj\alpha_{(j)}j=1,2,4,8,\ldots, |\alpha|\alpha_{(h)}\alpha_{(j)}\alpha_{(j)}h/2 \le |\alpha^{a}| < h \alpha^{a}\alpha_{(j)}j=h-1, h-2, \ldots\alpha^{a}\alpha^{a}\delta \gammad\gammad\alpha\deltaAT'\alpha^{*}T'T'\alpha^{a}\delta\gamma\alpha^{*}\delta \gamma\alpha^{*}T'B\alpha^{a}\delta\gammaT'ABT'BABT'\alpha^{*}\alpha^{*}\alphaABB\alpha^{*}\alpha^{b}|\alpha^{b}| \le |\alpha^{a}|\alpha'\alpha\alpha^{a}\alpha'T'\alpha^{*}=\alpha^{a}\alpha\alpha^{a}\alpha'\alpha\alpha^{a}\alpha'B\alpha'T'\alpha'T'occ_1\alpha'\alpha\alphaABocc_1AB\alpha'\beta\gamma\alpha'\delta\gamma\alpha'\delta\gammaT'\alpha'\beta\gamma\alpha'f_1T'occ_1\alpha'T'occ_2\alpha'occ_1Bp_1p_2occ_1occ_2Bp_a\alpha^{a}\delta\gammaBocc_2\alpha^{a}\delta\gammap_2p_ap_1p_2p_2p_1occ_2\alpha\alpha'\alphaA\alpha^{a}\alpha'\alpha^{a}p_1p_2\alpha''\alphap_2|\alpha'| > |\alpha''| > |\alpha^{a}|\alpha''occ_2\alpha''\alpha'\delta\gamma\alpha''\alpha\alpha'\alpha'\alpha^{a}\alpha''\alpha^{a}p_2p_aocc_2\alpha^{a}\delta\gamma\alpha'f_2T'occ_2f_2f_1\etaBp_2\eta\alpha^{a}\delta\gammap_2p_aT'\alpha^{a}\delta\gamma\alpha'f_2\etaT'ABT'\gamma\alpha'\beta\gammaf_1\gammaf_1f_2\alpha'B\alpha'T'\alpha'T'\alpha'BB\alpha'\delta\gamma\alpha'\alpha'\delta\gammaT'|\alpha'\delta\gamma| > |\alpha^{a}\delta\gamma|B\alpha^{a}\delta\gammaT'\alpha'T'B\alpha'T'A\alpha'A\alpha^{a}\alpha^{a}\alphaA|\alpha'| > |\alpha^{a}|\alpha'T'\alpha'B\alpha'\alpha\alpha^{a}\alpha'AB\alpha'T'\alpha^{*}\alpha'\alpha\alpha^{a}\alpha'T'O(|\alpha^{*}|)A\alpha'\alpha\alpha^{*}\alpha' \delta \gammaB\alpha' \delta \gammaB\alpha'\beta\gammaA\alpha' \beta \gamma\alpha' (\beta /\, \delta)\gamma\beta(\beta /\, \delta)\alpha^{a}O(|\alpha^{a}|)O(|\alpha^{a}\delta\hat{\gamma}|)\hat{\gamma}\gammad\hat{\gamma}ABd\gamma\alpha^{*}O(|\alpha^{*}|)O(|\alpha^{*}\delta\hat{\gamma}|)\alpha (\beta /\, \delta) \gamma\alpha \beta \gamma\alpha (\beta /\, \delta) \gammaO(|\alpha^{*}| + |\delta| + |\hat{\gamma}|)\alpha_1 (\beta_1 /\, \delta_1) \ldots \alpha_k (\beta_k /\, \delta_k) \alpha_{k+1}A=\alpha_1 \beta_1 \ldots \alpha_k \beta_k \alpha_{k+1}B= \alpha_1 \delta_1 \ldots \alpha_k \delta_k \alpha_{k+1}1\le i \le k+1\alpha^{a}_{i}\alpha^{b}_{i}\alpha_iAB\alpha^{*}_{i}\alpha^{a}_{i}\alpha^{b}_{i}\alpha^{*}_{i}\alpha_{i}AB\hat{\alpha}_{i}\alpha_{i}d_{i}\hat{\alpha}_{i}ABd_{i}\alpha_{i}B\alpha_{k+1}\alpha^{*}_i \beta_i \alpha_{i+1}\ldots\alpha_{k+1}\alpha_{i+1}\ldots\alpha_{k+1}\alpha^{*}_i \delta_i \alpha_{i+1}\ldots\alpha_{k+1}\alpha_{i+1}\ldots\alpha_{k+1}\alpha_{i}' (\beta_i /\, \delta_i) \ldots \alpha_{k+1}\alpha_{i}'\alpha_{i}\alpha^{*}_{i}A\alpha^{a}_{i}Ai1 \le i \le k\alpha^{a}_{i} \delta_i \alpha_{i+1} \ldots \alpha_{k+1}\alpha_{i+1}\ldots\alpha_{k+1}i\alpha^{*}_{i}i\alpha^{*}_{i} \delta_i  \ldots \alpha_{k+1}\alpha^{a}_{i} \delta_i  \ldots \alpha_{k+1}i\alpha_{i} \delta_i  \ldots \alpha_{k+1}\alpha^{*}_{i} \delta_i  \ldots \alpha_{k+1}i\alpha_1 (\beta_1 /\, \delta_1) \ldots \alpha_k (\beta_k /\, \delta_k) \alpha_{k+1}\alpha_1 \beta_1 \ldots \alpha_k \beta_k \alpha_{k+1}\alpha^{*}_i\delta_i\hat{\alpha}_{i+1}1 \le i \le k\alpha (\beta /\, \delta /\,\vartheta) \gammaA=\alpha\beta\gammaB=\alpha\delta\gammaC=\alpha\vartheta\gamma\beta\gamma\delta\gamma\vartheta\gamma\alpha^{a}\alpha^{b}\alpha^{c}\alphaABC\alpha^{*}\alpha^{a}\alpha^{b}\alpha^{c}\alpha^{*} \vartheta \gamma\gamma\alpha\beta\gamma\alpha (\beta /\, \delta ) \gammaB(\beta /\, \delta /\,\vartheta) \gammaCB$ occasionally).
We omit the details.










\begin{thebibliography}{10}

\bibitem{nature/1000Genomes10}
The 1000 Genomes~Project Consortium.
\newblock A map of human genome variation from population-scale sequencing.
\newblock {\em Nature}, 467(7319):1061--1073, 2010.

\bibitem{Amir&Farach:94}
A.~Amir, M.~Farach, Z.~Galil, R.~Giancarlo, and K.~Park.
\newblock Dynamic dictionary matching.
\newblock {\em J.~Comput. Syst. Sci.}, 49:208--222, 1994.

\bibitem{Baeza-Yates&Gonnet:96}
R.~A. Baeza-Yates and G.~H. Gonnet.
\newblock Fast text searching for regular expressions or automaton searching on
  tries.
\newblock {\em J.~ACM}, 43(6):915--936, 1996.

\bibitem{Bille&Gortz:11}
P.~Bille and I.L. G{\o}rtz.
\newblock Substring range reporting.
\newblock In {\em Proc. of the 22nd CPM}, pages 299--308, 2011.

\bibitem{Burrows&Wheeler:94}
M.~Burrows and D.~J. Wheeler.
\newblock A block-sorting lossless data compression algorithm.
\newblock Technical Report 124, Digital Equipment Corporation,
  California, 1994.

\bibitem{Crochemore&Rytter:02}
M.~Crochemore and W.~Rytter.
\newblock {\em Jewels of Stringology}.
\newblock World Scientific Publishing, Singapore, 2002.

\bibitem{aaim/DoJSS12}
H.~H. Do, J.~Jansson, K.~Sadakane, and W.~K. Sung.
\newblock Fast relative lempel-ziv self-index for similar sequences.
\newblock In {\em Proc. of FAW-AAIM 2012}, pages 291--302, 2012.

\bibitem{Farach-Colton&Ferragina:00}
M.~Farach-Colton, P.~Ferragina, and S.~Muthukrishnan.
\newblock On the sorting-complexity of suffix tree construction.
\newblock {\em J.~ACM}, 47(6):987--1011, 2000.

\bibitem{Grossi&Vitter:05}
R.~Grossi and J.~S. Vitter.
\newblock Compressed suffix arrays and suffix trees with applications to text
  indexing and string matching.
\newblock {\em SIAM J. Comput.}, 35(2):378--407, 2005.

\bibitem{Gusfield:97}
D.~Gusfield.
\newblock {\em Algorithms on Strings, Tree, and Sequences}.
\newblock Cambridge University Press, Cambridge, 1997.

\bibitem{aaim/HuangLSTY10}
S.~Huang, T.~W. Lam, W.~K. Sung, S.~L. Tam, and S.~M. Yiu.
\newblock Indexing similar dna sequences.
\newblock In {\em Proc. of AAIM 2010}, pages 180--190, 2010.

\bibitem{karlin1983new}
S.~Karlin, G.~Ghandour, F.~Ost, S.~Tavare, and L.~J. Korn.
\newblock New approaches for computer analysis of nucleic acid sequences.
\newblock {\em Proc. Natl. Acad. Sci.}, 80(18):5660--5664, 1983.

\bibitem{Kreft&Navarro:XX}
S.~Kreft and G.~Navarro.
\newblock On compressing and indexing repetitive sequences.
\newblock {\em Theor. Comput. Sci.} (to appear).

\bibitem{spire/KuruppuPZ10}
S.~Kuruppu, S.~J. Puglisi, and J.~Zobel.
\newblock Relative lempel-ziv compression of genomes for large-scale storage
  and retrieval.
\newblock In {\em Proc. of the 17th SPIRE}, pages 201--206, 2010.

\bibitem{Levenshtein:66}
V.~Levenshtein.
\newblock Binary codes capable of correcting deletions, insertions and
  reversals.
\newblock {\em Soviet Physics Doklady}, 10(8):707--710, February 1966.

\bibitem{recomb/MakinenNSV09}
V.~M{\"a}kinen, G.~Navarro, J.~Sir{\'e}n, and N.~V{\"a}lim{\"a}ki.
\newblock Storage and retrieval of individual genomes.
\newblock In {\em Proc. of the 13th RECOMB}, pages 121--137, 2009.

\bibitem{jcb/MakinenNSV10}
V.~M{\"a}kinen, G.~Navarro, J.~Sir{\'e}n, and N.~V{\"a}lim{\"a}ki.
\newblock Storage and retrieval of highly repetitive sequence collections.
\newblock {\em J.~Comput. Bio.}, 17(3):281--308, 2010.

\bibitem{McCreight:76}
E.~M. McCreight.
\newblock A space-economical suffix tree construction algorithm.
\newblock {\em J.~ACM}, 23(2):262--272, 1976.

\bibitem{iwoca/Navarro12}
G.~Navarro.
\newblock Indexing highly repetitive collections.
\newblock In {\em Proc. of 23th IWOCA}, pages 274--279, 2012.

\bibitem{Sadakane07TCS}
K.~Sadakane.
\newblock Compressed suffix trees with full functionality.
\newblock {\em Theor. Comput. Sci.}, 41(4):589--607, 2007.

\bibitem{Ukkonen:95}
E.~Ukkonen.
\newblock On-line construction of suffix trees.
\newblock {\em Algorithmica}, 14(3):249--260, 1995.

\bibitem{Weiner:73}
P.~Weiner.
\newblock Linear pattern matching algorithms.
\newblock In {\em Proc. of the 14th IEEE Symp. on Switching and Automata
  Theory}, pages 1--11, 1973.

\bibitem{Ziv&Lempel:77}
J.~Ziv and A.~Lempel.
\newblock A universal algorithm for sequential data compression.
\newblock {\em IEEE Trans. on Information Theory}, 23(3):337--343, 1977.

\end{thebibliography}




\end{document}
